

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>GEM-PRO - Genes &amp; Sequences &mdash; ssbio 1.0.0a0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />
  

  
        <link rel="index" title="Index"
              href="../genindex.html"/>
        <link rel="search" title="Search" href="../search.html"/>
    <link rel="top" title="ssbio 1.0.0a0 documentation" href="../index.html"/>
        <link rel="up" title="The GEM-PRO Pipeline" href="../gempro.html"/>
        <link rel="next" title="GEM-PRO - List of Gene IDs" href="GEM-PRO - List of Gene IDs.html"/>
        <link rel="prev" title="The GEM-PRO Pipeline" href="../gempro.html"/> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> ssbio
          

          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../getting_started.html">Getting Started</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../getting_started.html#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../getting_started.html#the-basics">The basics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../getting_started.html#structural-biology">Structural biology</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../getting_started.html#protein-structures">Protein structures</a></li>
<li class="toctree-l3"><a class="reference internal" href="../getting_started.html#biopython">Biopython</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../getting_started.html#systems-biology">Systems biology</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../getting_started.html#cobra-models-cobrapy">COBRA models &amp; COBRApy</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../getting_started.html#reading">Reading</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../structure.html">The Structure Class</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../structure.html#introduction">Introduction</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../protein.html">The Protein Class</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../protein.html#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../protein.html#tutorials">Tutorials</a><ul>
<li class="toctree-l3"><a class="reference internal" href="Protein - Structure Mapping, Alignments, and Visualization.html">Protein - Structure Mapping, Alignments, and Visualization</a><ul>
<li class="toctree-l4"><a class="reference internal" href="Protein - Structure Mapping, Alignments, and Visualization.html#Imports">Imports</a></li>
<li class="toctree-l4"><a class="reference internal" href="Protein - Structure Mapping, Alignments, and Visualization.html#Initialization-of-the-project">Initialization of the project</a></li>
<li class="toctree-l4"><a class="reference internal" href="Protein - Structure Mapping, Alignments, and Visualization.html#Mapping-sequence--&gt;-structure">Mapping sequence -&gt; structure</a></li>
<li class="toctree-l4"><a class="reference internal" href="Protein - Structure Mapping, Alignments, and Visualization.html#Downloading-and-ranking-structures">Downloading and ranking structures</a></li>
<li class="toctree-l4"><a class="reference internal" href="Protein - Structure Mapping, Alignments, and Visualization.html#Loading-and-aligning-new-sequences">Loading and aligning new sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="Protein - Structure Mapping, Alignments, and Visualization.html#Viewing-structures">Viewing structures</a></li>
<li class="toctree-l4"><a class="reference internal" href="Protein - Structure Mapping, Alignments, and Visualization.html#Mapping-sequence-residue-numbers-to-structure-residue-numbers">Mapping sequence residue numbers to structure residue numbers</a></li>
<li class="toctree-l4"><a class="reference internal" href="Protein - Structure Mapping, Alignments, and Visualization.html#Saving">Saving</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="../gempro.html">The GEM-PRO Pipeline</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../gempro.html#introduction">Introduction</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../gempro.html#tutorials">Tutorials</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">GEM-PRO - Genes &amp; Sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Imports">Imports</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Initialization-of-the-project">Initialization of the project</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Mapping-sequence--&gt;-structure">Mapping sequence -&gt; structure</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Downloading-and-ranking-structures">Downloading and ranking structures</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="GEM-PRO - List of Gene IDs.html">GEM-PRO - List of Gene IDs</a><ul>
<li class="toctree-l4"><a class="reference internal" href="GEM-PRO - List of Gene IDs.html#Imports">Imports</a></li>
<li class="toctree-l4"><a class="reference internal" href="GEM-PRO - List of Gene IDs.html#Initialization-of-the-project">Initialization of the project</a></li>
<li class="toctree-l4"><a class="reference internal" href="GEM-PRO - List of Gene IDs.html#Mapping-gene-ID--&gt;-sequence">Mapping gene ID -&gt; sequence</a></li>
<li class="toctree-l4"><a class="reference internal" href="GEM-PRO - List of Gene IDs.html#Mapping-representative-sequence-to-structure">Mapping representative sequence to structure</a></li>
<li class="toctree-l4"><a class="reference internal" href="GEM-PRO - List of Gene IDs.html#Downloading-and-ranking-structures">Downloading and ranking structures</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="GEM-PRO - SBML Model (iNJ661).html">GEM-PRO - SBML Model (iNJ661)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="GEM-PRO - SBML Model (iNJ661).html#Imports">Imports</a></li>
<li class="toctree-l4"><a class="reference internal" href="GEM-PRO - SBML Model (iNJ661).html#Initialization-of-the-project">Initialization of the project</a></li>
<li class="toctree-l4"><a class="reference internal" href="GEM-PRO - SBML Model (iNJ661).html#Mapping-gene-ID--&gt;-sequence">Mapping gene ID -&gt; sequence</a></li>
<li class="toctree-l4"><a class="reference internal" href="GEM-PRO - SBML Model (iNJ661).html#Mapping-representative-sequence-to-structure">Mapping representative sequence to structure</a></li>
<li class="toctree-l4"><a class="reference internal" href="GEM-PRO - SBML Model (iNJ661).html#Downloading-and-ranking-structures">Downloading and ranking structures</a></li>
<li class="toctree-l4"><a class="reference internal" href="GEM-PRO - SBML Model (iNJ661).html#Creating-homology-models">Creating homology models</a></li>
<li class="toctree-l4"><a class="reference internal" href="GEM-PRO - SBML Model (iNJ661).html#Saving-your-GEM-PRO">Saving your GEM-PRO</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../atlas.html">The ATLAS Pipeline</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../atlas.html#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../atlas.html#tutorials">Tutorials</a><ul class="simple">
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../python_api.html">Python API</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../index.html">ssbio</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          

 



<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../index.html">Docs</a> &raquo;</li>
      
          <li><a href="../gempro.html">The GEM-PRO Pipeline</a> &raquo;</li>
      
    <li>GEM-PRO - Genes &amp; Sequences</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/notebooks/GEM-PRO - Genes &amp; Sequences.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput,
div.nbinput div.prompt,
div.nbinput div.input_area,
div.nbinput div[class*=highlight],
div.nbinput div[class*=highlight] pre,
div.nboutput,
div.nbinput div.prompt,
div.nbinput div.output_area,
div.nboutput div[class*=highlight],
div.nboutput div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput,
div.nboutput {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput,
    div.nboutput {
        flex-direction: column;
    }
}

/* input container */
div.nbinput {
    padding-top: 5px;
}

/* last container */
div.nblast {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput div.prompt pre {
    color: #303F9F;
}

/* output prompt */
div.nboutput div.prompt pre {
    color: #D84315;
}

/* all prompts */
div.nbinput div.prompt,
div.nboutput div.prompt {
    min-width: 8ex;
    padding-top: 0.4em;
    padding-right: 0.4em;
    text-align: right;
    flex: 0;
}
@media (max-width: 540px) {
    div.nbinput div.prompt,
    div.nboutput div.prompt {
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput div.prompt.empty {
        padding: 0;
    }
}

/* disable scrollbars on prompts */
div.nbinput div.prompt pre,
div.nboutput div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput div.input_area,
div.nboutput div.output_area {
    padding: 0.4em;
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput div.input_area,
    div.nboutput div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput div.input_area {
    border: 1px solid #cfcfcf;
    border-radius: 2px;
    background: #f7f7f7;
}

/* override MathJax center alignment in output cells */
div.nboutput div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.pngmath center alignment in output cells */
div.nboutput div.math p {
    text-align: left;
}

/* standard error */
div.nboutput div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-bold { font-weight: bold; }

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast,
.nboutput.nblast {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast + .nbinput {
    margin-top: -19px;
}

/* nice headers on first paragraph of info/warning boxes */
.admonition .first {
    margin: -12px;
    padding: 6px 12px;
    margin-bottom: 12px;
    color: #fff;
    line-height: 1;
    display: block;
}
.admonition.warning .first {
    background: #f0b37e;
}
.admonition.note .first {
    background: #6ab0de;
}
.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}
</style>
<div class="section" id="GEM-PRO---Genes-&amp;-Sequences">
<h1>GEM-PRO - Genes &amp; Sequences<a class="headerlink" href="#GEM-PRO---Genes-&-Sequences" title="Permalink to this headline">¶</a></h1>
<p>This notebook gives an example of how to run the GEM-PRO pipeline with a
<strong>dictionary of gene IDs and their protein sequences</strong>.</p>
<p><div class="admonition note">
<strong>Input:</strong> Dictionary of gene IDs and protein sequences</div>
<div class="admonition note">
<strong>Output:</strong> GEM-PRO model</div>
</p><div class="section" id="Imports">
<h2>Imports<a class="headerlink" href="#Imports" title="Permalink to this headline">¶</a></h2>
<div class="nbinput nblast container">
<div class="prompt highlight-none"><div class="highlight"><pre>
<span></span>In [1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="c1"># Import the GEM-PRO class</span>
<span class="kn">from</span> <span class="nn">ssbio.pipeline.gempro</span> <span class="kn">import</span> <span class="n">GEMPRO</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast container">
<div class="prompt highlight-none"><div class="highlight"><pre>
<span></span>In [2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="c1"># Printing multiple outputs per cell</span>
<span class="kn">from</span> <span class="nn">IPython.core.interactiveshell</span> <span class="kn">import</span> <span class="n">InteractiveShell</span>
<span class="n">InteractiveShell</span><span class="o">.</span><span class="n">ast_node_interactivity</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="Logging">
<h3>Logging<a class="headerlink" href="#Logging" title="Permalink to this headline">¶</a></h3>
<p>Set the logging level in <code class="docutils literal"><span class="pre">logger.setLevel(logging.&lt;LEVEL_HERE&gt;)</span></code> to
specify how verbose you want the pipeline to be. Debug is most verbose.</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">CRITICAL</span></code><ul>
<li>Only really important messages shown</li>
</ul>
</li>
<li><code class="docutils literal"><span class="pre">ERROR</span></code><ul>
<li>Major errors</li>
</ul>
</li>
<li><code class="docutils literal"><span class="pre">WARNING</span></code><ul>
<li>Warnings that don&#8217;t affect running of the pipeline</li>
</ul>
</li>
<li><code class="docutils literal"><span class="pre">INFO</span></code> (default)<ul>
<li>Info such as the number of structures mapped per gene</li>
</ul>
</li>
<li><code class="docutils literal"><span class="pre">DEBUG</span></code><ul>
<li>Really detailed information that will print out a lot of stuff</li>
</ul>
</li>
</ul>
<div class="admonition warning">
<strong>Warning:</strong> <code class="docutils literal"><span class="pre">DEBUG</span></code> mode prints out a large amount of information,
especially if you have a lot of genes. This may stall your notebook!</div>
<div class="nbinput nblast container">
<div class="prompt highlight-none"><div class="highlight"><pre>
<span></span>In [3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="c1"># Create logger</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">()</span>
<span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>  <span class="c1"># SET YOUR LOGGING LEVEL HERE #</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="Initialization-of-the-project">
<h2>Initialization of the project<a class="headerlink" href="#Initialization-of-the-project" title="Permalink to this headline">¶</a></h2>
<p>Set these three things:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">ROOT_DIR</span></code><ul>
<li>The directory where a folder named after your <code class="docutils literal"><span class="pre">PROJECT</span></code> will be
created</li>
</ul>
</li>
<li><code class="docutils literal"><span class="pre">PROJECT</span></code><ul>
<li>Your project name</li>
</ul>
</li>
<li><code class="docutils literal"><span class="pre">GENES_AND_SEQUENCES</span></code><ul>
<li>Your dictionary of gene IDs and their sequence strings</li>
</ul>
</li>
</ul>
<p>A directory will be created in <code class="docutils literal"><span class="pre">ROOT_DIR</span></code> with your <code class="docutils literal"><span class="pre">PROJECT</span></code> name.
The folders are organized like so:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>ROOT_DIR
└── PROJECT
    ├── data  # General storage for pipeline outputs
    ├── model  # SBML and GEM-PRO models are stored here
    ├── genes  # Per gene information
    │   ├── &lt;gene_id1&gt;  # Specific gene directory
    │   │   └── protein
    │   │       ├── sequences  # Protein sequence files, alignments, etc.
    │   │       └── structures  # Protein structure files, calculations, etc.
    │   └── &lt;gene_id2&gt;
    │       └── protein
    │           ├── sequences
    │           └── structures
    ├── reactions  # Per reaction information
    │   └── &lt;reaction_id1&gt;  # Specific reaction directory
    │       └── complex
    │           └── structures  # Protein complex files
    └── metabolites  # Per metabolite information
        └── &lt;metabolite_id1&gt;  # Specific metabolite directory
            └── chemical
                └── structures  # Metabolite 2D and 3D structure files
</pre></div>
</div>
<div class="admonition note">
<strong>Note:</strong> Methods for protein complexes and metabolites are still in
development.</div>
<div class="nbinput nblast container">
<div class="prompt highlight-none"><div class="highlight"><pre>
<span></span>In [4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="c1"># SET FOLDERS AND DATA HERE</span>
<span class="kn">import</span> <span class="nn">tempfile</span>
<span class="n">ROOT_DIR</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">gettempdir</span><span class="p">()</span>

<span class="n">PROJECT</span> <span class="o">=</span> <span class="s1">&#39;genes_and_sequences_GP&#39;</span>
<span class="n">GENES_AND_SEQUENCES</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;b0870&#39;</span><span class="p">:</span> <span class="s1">&#39;MIDLRSDTVTRPSRAMLEAMMAAPVGDDVYGDDPTVNALQDYAAELSGKEAAIFLPTGTQANLVALLSHCERGEEYIVGQAAHNYLFEAGGAAVLGSIQPQPIDAAADGTLPLDKVAMKIKPDDIHFARTKLLSLENTHNGKVLPREYLKEAWEFTRERNLALHVDGARIFNAVVAYGCELKEITQYCDSFTICLSKGLGTPVGSLLVGNRDYIKRAIRWRKMTGGGMRQSGILAAAGIYALKNNVARLQEDHDNAAWMAEQLREAGADVMRQDTNMLFVRVGEENAAALGEYMKARNVLINASPIVRLVTHLDVSREQLAEVAAHWRAFLAR&#39;</span><span class="p">,</span>
                       <span class="s1">&#39;b3041&#39;</span><span class="p">:</span> <span class="s1">&#39;MNQTLLSSFGTPFERVENALAALREGRGVMVLDDEDRENEGDMIFPAETMTVEQMALTIRHGSGIVCLCITEDRRKQLDLPMMVENNTSAYGTGFTVTIEAAEGVTTGVSAADRITTVRAAIADGAKPSDLNRPGHVFPLRAQAGGVLTRGGHTEATIDLMTLAGFKPAGVLCELTNDDGTMARAPECIEFANKHNMALVTIEDLVAYRQAHERKAS&#39;</span><span class="p">}</span>
</pre></div>
</div>
</div>
<div class="nbinput container">
<div class="prompt highlight-none"><div class="highlight"><pre>
<span></span>In [5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="c1"># Create the GEM-PRO project</span>
<span class="n">my_gempro</span> <span class="o">=</span> <span class="n">GEMPRO</span><span class="p">(</span><span class="n">gem_name</span><span class="o">=</span><span class="n">PROJECT</span><span class="p">,</span> <span class="n">root_dir</span><span class="o">=</span><span class="n">ROOT_DIR</span><span class="p">,</span> <span class="n">genes_and_sequences</span><span class="o">=</span><span class="n">GENES_AND_SEQUENCES</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast container">
<div class="prompt empty container">
</div>
<div class="stderr output_area container">
<div class="highlight"><pre>
INFO:ssbio.pipeline.gempro:/tmp/genes_and_sequences_GP: GEM-PRO project location
INFO:ssbio.pipeline.gempro:Loaded in 2 sequences
INFO:ssbio.pipeline.gempro:2: number of genes
</pre></div></div>
</div>
</div>
<div class="section" id="Mapping-sequence--&gt;-structure">
<h2>Mapping sequence -&gt; structure<a class="headerlink" href="#Mapping-sequence-->-structure" title="Permalink to this headline">¶</a></h2>
<p>Since the sequences have been provided, we just need to BLAST them to
the PDB.</p>
<p><div class="admonition note">
<strong>Note:</strong> These methods do not download any 3D structure files.</div>
</p><div class="section" id="Methods">
<h3>Methods<a class="headerlink" href="#Methods" title="Permalink to this headline">¶</a></h3>
<div class="section" id="blast_seqs_to_pdb">
<h4><code class="docutils literal"><span class="pre">blast_seqs_to_pdb</span></code><a class="headerlink" href="#blast_seqs_to_pdb" title="Permalink to this headline">¶</a></h4>
<p>This will BLAST the representative sequence against the entire PDB, and
return significant hits. XML files of the BLAST results are saved in the
respective sequence folders for a protein.</p>
<p><div class="admonition note">
<strong>Warning:</strong> A PDB BLAST may return hits in other organisms.</div>
</p><ul>
<li><p class="first"><code class="docutils literal"><span class="pre">seq_ident_cutoff</span></code></p>
<ul class="simple">
<li>Default: <code class="docutils literal"><span class="pre">0</span></code></li>
<li>From 0 to 1</li>
</ul>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">evalue</span></code></p>
<ul class="simple">
<li>Default: <code class="docutils literal"><span class="pre">0.0001</span></code></li>
<li>Significance of BLAST results</li>
</ul>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">all_genes</span></code></p>
<ul class="simple">
<li>Default: <code class="docutils literal"><span class="pre">False</span></code></li>
<li>Set to <code class="docutils literal"><span class="pre">True</span></code> if you want all genes and their sequences BLASTed</li>
<li>Set to <code class="docutils literal"><span class="pre">False</span></code> if you only want to BLAST sequences that did not
have any PDBs mapped to them already</li>
</ul>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">display_link</span></code></p>
<ul class="simple">
<li>Default: <code class="docutils literal"><span class="pre">False</span></code></li>
<li>Set to <code class="docutils literal"><span class="pre">True</span></code> if you want a clickable HTML link to be printed</li>
</ul>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">force_rerun</span></code></p>
<ul class="simple">
<li>Default: <code class="docutils literal"><span class="pre">False</span></code></li>
<li>Set to <code class="docutils literal"><span class="pre">True</span></code> if you want to ignore any existing XML results and
run the BLAST again</li>
</ul>
</li>
<li><p class="first">What&#8217;s saved?</p>
<ul>
<li><p class="first">Protein structures</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">my_gempro</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="s1">&#39;b0870&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">structures</span>
</pre></div>
</div>
</li>
<li><p class="first">DataFrames</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">my_gempro</span><span class="o">.</span><span class="n">df_pdb_blast</span>
</pre></div>
</div>
</li>
</ul>
</li>
</ul>
<div class="nbinput container">
<div class="prompt highlight-none"><div class="highlight"><pre>
<span></span>In [6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="c1"># Mapping using BLAST</span>
<span class="n">my_gempro</span><span class="o">.</span><span class="n">blast_seqs_to_pdb</span><span class="p">(</span><span class="n">seq_ident_cutoff</span><span class="o">=.</span><span class="mi">80</span><span class="p">,</span> <span class="n">evalue</span><span class="o">=</span><span class="mf">0.00001</span><span class="p">)</span>
<span class="n">my_gempro</span><span class="o">.</span><span class="n">df_pdb_blast</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput container">
<div class="prompt empty container">
</div>
<div class="stderr output_area container">
<div class="highlight"><pre>
INFO:ssbio.pipeline.gempro:Completed sequence --&gt; PDB BLAST. See the &#34;df_pdb_blast&#34; attribute.
INFO:ssbio.pipeline.gempro:0: number of genes with additional structures added from BLAST
</pre></div></div>
</div>
<div class="nboutput container">
<div class="prompt empty container">
</div>
<div class="output_area container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nboutput nblast container">
<div class="prompt highlight-none"><div class="highlight"><pre>
<span></span>Out[6]:
</pre></div>
</div>
<div class="output_area container">
<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>pdb_id</th>
      <th>pdb_chain_id</th>
      <th>hit_score</th>
      <th>hit_evalue</th>
      <th>hit_percent_similar</th>
      <th>hit_percent_ident</th>
      <th>hit_num_ident</th>
      <th>hit_num_similar</th>
    </tr>
    <tr>
      <th>gene</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>b3041</th>
      <td>1iez</td>
      <td>A</td>
      <td>1060.0</td>
      <td>1.442310e-115</td>
      <td>1.000000</td>
      <td>1.000000</td>
      <td>217</td>
      <td>217</td>
    </tr>
    <tr>
      <th>b3041</th>
      <td>1g58</td>
      <td>A</td>
      <td>1050.0</td>
      <td>2.014330e-114</td>
      <td>0.995392</td>
      <td>0.995392</td>
      <td>216</td>
      <td>216</td>
    </tr>
    <tr>
      <th>b3041</th>
      <td>1g58</td>
      <td>B</td>
      <td>1050.0</td>
      <td>2.014330e-114</td>
      <td>0.995392</td>
      <td>0.995392</td>
      <td>216</td>
      <td>216</td>
    </tr>
    <tr>
      <th>b3041</th>
      <td>1g57</td>
      <td>A</td>
      <td>1050.0</td>
      <td>2.014330e-114</td>
      <td>0.995392</td>
      <td>0.995392</td>
      <td>216</td>
      <td>216</td>
    </tr>
    <tr>
      <th>b3041</th>
      <td>1g57</td>
      <td>B</td>
      <td>1050.0</td>
      <td>2.014330e-114</td>
      <td>0.995392</td>
      <td>0.995392</td>
      <td>216</td>
      <td>216</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput container">
<div class="prompt highlight-none"><div class="highlight"><pre>
<span></span>In [7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="c1"># Looking at the information saved within a gene</span>
<span class="n">my_gempro</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="s1">&#39;b0870&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">structures</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast container">
<div class="prompt highlight-none"><div class="highlight"><pre>
<span></span>Out[7]:
</pre></div>
</div>
<div class="output_area highlight-none"><div class="highlight"><pre>
<span></span>[&lt;PDBProp 3wlx at 0x7f4d568e0ef0&gt;,
 &lt;PDBProp 4rjy at 0x7f4d568e07f0&gt;,
 &lt;PDBProp 4lnm at 0x7f4d57a59ac8&gt;,
 &lt;PDBProp 4lnl at 0x7f4d56839128&gt;,
 &lt;PDBProp 4lnj at 0x7f4d56839240&gt;]
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="Downloading-and-ranking-structures">
<h2>Downloading and ranking structures<a class="headerlink" href="#Downloading-and-ranking-structures" title="Permalink to this headline">¶</a></h2>
<div class="section" id="Methods">
<h3>Methods<a class="headerlink" href="#Methods" title="Permalink to this headline">¶</a></h3>
<div class="section" id="pdb_downloader_and_metadata">
<h4><code class="docutils literal"><span class="pre">pdb_downloader_and_metadata</span></code><a class="headerlink" href="#pdb_downloader_and_metadata" title="Permalink to this headline">¶</a></h4>
<p>Download <strong>all</strong> structures per protein. This also adds metadata to each
PDB object in the list of structures.</p>
<p><div class="admonition warning">
<strong>Warning:</strong> Don&#8217;t run this if you don&#8217;t need all PDB structures - just
set representative structures below if you want 1 structure per protein.</div>
</p><ul>
<li><p class="first"><code class="docutils literal"><span class="pre">outdir</span></code></p>
<ul class="simple">
<li>Default: <code class="docutils literal"><span class="pre">None</span></code></li>
<li>Set this to a custom location if you want to save PDB files
outside the GEM-PRO project folder</li>
</ul>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">pdb_file_type</span></code></p>
<ul class="simple">
<li>Default: <code class="docutils literal"><span class="pre">'cif'</span></code> (set in GEMPRO project initialization, but can
be changed here)</li>
<li><code class="docutils literal"><span class="pre">'pdb'</span></code>, <code class="docutils literal"><span class="pre">'pdb.gz'</span></code>, <code class="docutils literal"><span class="pre">'mmcif'</span></code>, <code class="docutils literal"><span class="pre">'cif'</span></code>, <code class="docutils literal"><span class="pre">'cif.gz'</span></code>,
<code class="docutils literal"><span class="pre">'xml.gz'</span></code>, <code class="docutils literal"><span class="pre">'mmtf'</span></code>, <code class="docutils literal"><span class="pre">'mmtf.gz'</span></code> - File type for files
downloaded from the PDB.</li>
</ul>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">force_rerun</span></code></p>
<ul class="simple">
<li>Default: <code class="docutils literal"><span class="pre">False</span></code></li>
<li>Set to <code class="docutils literal"><span class="pre">True</span></code> if you want to re-download PDB files.</li>
</ul>
</li>
<li><p class="first">What&#8217;s saved?</p>
<ul>
<li><p class="first">Additional metadata per structure</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">my_gempro</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="s1">&#39;b0870&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">structures</span>
</pre></div>
</div>
</li>
<li><p class="first">DataFrames</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">my_gempro</span><span class="o">.</span><span class="n">df_pdb_metadata</span>
</pre></div>
</div>
</li>
</ul>
</li>
</ul>
<div class="nbinput container">
<div class="prompt highlight-none"><div class="highlight"><pre>
<span></span>In [8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="c1"># Download all mapped PDBs and gather the metadata</span>
<span class="n">my_gempro</span><span class="o">.</span><span class="n">pdb_downloader_and_metadata</span><span class="p">()</span>
<span class="n">my_gempro</span><span class="o">.</span><span class="n">df_pdb_metadata</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput container">
<div class="prompt empty container">
</div>
<div class="stderr output_area container">
<div class="highlight"><pre>
INFO:ssbio.pipeline.gempro:Updated PDB metadata dataframe. See the &#34;df_pdb_metadata&#34; attribute.
INFO:ssbio.pipeline.gempro:Saved 11 structures total
</pre></div></div>
</div>
<div class="nboutput container">
<div class="prompt empty container">
</div>
<div class="output_area container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nboutput nblast container">
<div class="prompt highlight-none"><div class="highlight"><pre>
<span></span>Out[8]:
</pre></div>
</div>
<div class="output_area container">
<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>pdb_id</th>
      <th>pdb_title</th>
      <th>description</th>
      <th>experimental_method</th>
      <th>resolution</th>
      <th>chemicals</th>
      <th>date</th>
      <th>taxonomy_name</th>
      <th>structure_file</th>
    </tr>
    <tr>
      <th>gene</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>b3041</th>
      <td>1iez</td>
      <td>Solution Structure of 3,4-Dihydroxy-2-Butanone...</td>
      <td>3,4-Dihydroxy-2-Butanone 4-Phosphate Synthase ...</td>
      <td>SOLUTION NMR</td>
      <td>NaN</td>
      <td>None</td>
      <td>2001-11-07;2003-04-01;2009-02-24;2011-07-13</td>
      <td>Escherichia coli</td>
      <td>1iez.cif</td>
    </tr>
    <tr>
      <th>b3041</th>
      <td>1g58</td>
      <td>CRYSTAL STRUCTURE OF 3,4-DIHYDROXY-2-BUTANONE ...</td>
      <td>3,4-DIHYDROXY-2-BUTANONE 4-PHOSPHATE SYNTHASE</td>
      <td>X-RAY DIFFRACTION</td>
      <td>1.55</td>
      <td>AU</td>
      <td>2001-04-30;2003-04-01;2009-02-24</td>
      <td>Escherichia coli</td>
      <td>1g58.cif</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
<div class="section" id="set_representative_structure">
<h4><code class="docutils literal"><span class="pre">set_representative_structure</span></code><a class="headerlink" href="#set_representative_structure" title="Permalink to this headline">¶</a></h4>
<p>Rank available structures, run QC/QA, download and clean the final
structure.</p>
<p><div class="admonition note">
<strong>Note:</strong> PDBs don&#8217;t need to be downloaded before running this step.
This is useful to limit the number of structures downloaded from the
PDB.</div>
</p><ul>
<li><p class="first"><code class="docutils literal"><span class="pre">pdb_file_type</span></code></p>
<ul class="simple">
<li>Default: <code class="docutils literal"><span class="pre">'cif'</span></code> (set in GEMPRO project initialization, but can
be changed here)</li>
<li><code class="docutils literal"><span class="pre">'pdb'</span></code>, <code class="docutils literal"><span class="pre">'pdb.gz'</span></code>, <code class="docutils literal"><span class="pre">'mmcif'</span></code>, <code class="docutils literal"><span class="pre">'cif'</span></code>, <code class="docutils literal"><span class="pre">'cif.gz'</span></code>,
<code class="docutils literal"><span class="pre">'xml.gz'</span></code>, <code class="docutils literal"><span class="pre">'mmtf'</span></code>, <code class="docutils literal"><span class="pre">'mmtf.gz'</span></code> - File type for files
downloaded from the PDB.</li>
</ul>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">engine</span></code></p>
<ul class="simple">
<li>Default: <code class="docutils literal"><span class="pre">'needle'</span></code></li>
<li>Set to <code class="docutils literal"><span class="pre">'biopython'</span></code> if you want to utilize Biopython&#8217;s built-in
pairwise alignment algorithm.</li>
</ul>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">always_use_homology</span></code></p>
<ul class="simple">
<li>Default: <code class="docutils literal"><span class="pre">False</span></code></li>
<li>Set to <code class="docutils literal"><span class="pre">True</span></code> if you always want to use homology models.</li>
</ul>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">seq_ident_cutoff</span></code></p>
<ul class="simple">
<li>Default: <code class="docutils literal"><span class="pre">0.5</span></code></li>
<li>QC/QA: sets the minimum sequence identity a structure has to have
to be selected as representative.</li>
</ul>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">allow_missing_on_termini</span></code></p>
<ul class="simple">
<li>Default: <code class="docutils literal"><span class="pre">0.2</span></code></li>
<li>QC/QA: Percentage of the total length of the reference sequence
which will be ignored when checking for modifications (mutations,
deletions, insertions, or unresolved residues). Example: if
<code class="docutils literal"><span class="pre">0.1</span></code>, and reference sequence is 100 AA, then only residues 5 to
95 will be checked for modifications.</li>
</ul>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">allow_mutants</span></code></p>
<ul class="simple">
<li>Default: <code class="docutils literal"><span class="pre">True</span></code></li>
<li>QC/QA: set to <code class="docutils literal"><span class="pre">True</span></code> if point mutations within the structure
should be allowed.</li>
</ul>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">allow_deletions</span></code></p>
<ul class="simple">
<li>Default: <code class="docutils literal"><span class="pre">False</span></code></li>
<li>QC/QA: set to <code class="docutils literal"><span class="pre">True</span></code> if deletions within the structure should be
allowed.</li>
</ul>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">allow_insertions</span></code></p>
<ul class="simple">
<li>Default: <code class="docutils literal"><span class="pre">False</span></code></li>
<li>QC/QA: set to <code class="docutils literal"><span class="pre">True</span></code> if insertions within the structure should
be allowed.</li>
</ul>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">allow_unresolved</span></code></p>
<ul class="simple">
<li>Default: <code class="docutils literal"><span class="pre">True</span></code></li>
<li>QC/QA: set to <code class="docutils literal"><span class="pre">True</span></code> if unresolved regions within the structure
should be allowed.</li>
</ul>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">force_rerun</span></code></p>
<ul class="simple">
<li>Default: <code class="docutils literal"><span class="pre">False</span></code></li>
<li>QC/QA: set to <code class="docutils literal"><span class="pre">True</span></code> if pairwise alignments and structure
cleaning should be rerun even if files exist.</li>
</ul>
</li>
<li><p class="first">What&#8217;s saved?</p>
<ul>
<li><p class="first">Representative protein structures</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">my_gempro</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="s1">&#39;b0870&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_structure</span>
</pre></div>
</div>
</li>
<li><p class="first">DataFrames</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">my_gempro</span><span class="o">.</span><span class="n">df_representative_structures</span>
</pre></div>
</div>
</li>
</ul>
</li>
</ul>
<div class="nbinput container">
<div class="prompt highlight-none"><div class="highlight"><pre>
<span></span>In [9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="c1"># Set representative structures</span>
<span class="n">my_gempro</span><span class="o">.</span><span class="n">set_representative_structure</span><span class="p">()</span>
<span class="n">my_gempro</span><span class="o">.</span><span class="n">df_representative_structures</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput container">
<div class="prompt empty container">
</div>
<div class="stderr output_area container">
<div class="highlight"><pre>
INFO:ssbio.pipeline.gempro:Created representative structures dataframe. See the &#34;df_representative_structures&#34; attribute.
INFO:ssbio.pipeline.gempro:2/2: genes with a representative structure
</pre></div></div>
</div>
<div class="nboutput container">
<div class="prompt empty container">
</div>
<div class="output_area container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nboutput nblast container">
<div class="prompt highlight-none"><div class="highlight"><pre>
<span></span>Out[9]:
</pre></div>
</div>
<div class="output_area container">
<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>id</th>
      <th>is_experimental</th>
      <th>reference_seq</th>
      <th>reference_seq_top_coverage</th>
    </tr>
    <tr>
      <th>gene</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>b3041</th>
      <td>1iez-A</td>
      <td>True</td>
      <td>b3041</td>
      <td>100.0</td>
    </tr>
    <tr>
      <th>b0870</th>
      <td>3wlx-A</td>
      <td>True</td>
      <td>b0870</td>
      <td>99.4</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput container">
<div class="prompt highlight-none"><div class="highlight"><pre>
<span></span>In [10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="c1"># Looking at the information saved within a gene</span>
<span class="n">my_gempro</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="s1">&#39;b0870&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_structure</span>
<span class="n">my_gempro</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="s1">&#39;b0870&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_structure</span><span class="o">.</span><span class="n">get_dict</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput container">
<div class="prompt highlight-none"><div class="highlight"><pre>
<span></span>Out[10]:
</pre></div>
</div>
<div class="output_area highlight-none"><div class="highlight"><pre>
<span></span>&lt;StructProp 3wlx-A at 0x7f4d548bf1d0&gt;
</pre></div>
</div>
</div>
<div class="nboutput nblast container">
<div class="prompt highlight-none"><div class="highlight"><pre>
<span></span>Out[10]:
</pre></div>
</div>
<div class="output_area highlight-none"><div class="highlight"><pre>
<span></span>{&#39;_structure_dir&#39;: &#39;/tmp/genes_and_sequences_GP/genes/b0870/b0870_protein/structures&#39;,
 &#39;chains&#39;: [&lt;ChainProp A at 0x7f4d55cd24a8&gt;],
 &#39;date&#39;: &#39;2014-12-17&#39;,
 &#39;description&#39;: &#39;Low specificity L-threonine aldolase (E.C.4.1.2.48)&#39;,
 &#39;file_type&#39;: &#39;pdb&#39;,
 &#39;id&#39;: &#39;3wlx-A&#39;,
 &#39;is_experimental&#39;: True,
 &#39;mapped_chains&#39;: [&#39;A&#39;],
 &#39;original_pdb_id&#39;: &#39;3wlx&#39;,
 &#39;reference_seq&#39;: &lt;SeqProp b0870 at 0x7f4d55cd2dd8&gt;,
 &#39;reference_seq_top_coverage&#39;: 99.4,
 &#39;representative_chain&#39;: &lt;ChainProp A at 0x7f4d55cd2cf8&gt;,
 &#39;resolution&#39;: 2.51,
 &#39;structure_file&#39;: &#39;3wlx-A_clean.pdb&#39;,
 &#39;taxonomy_name&#39;: &#39;Escherichia coli&#39;}
</pre></div>
</div>
</div>
</div>
</div>
</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="GEM-PRO - List of Gene IDs.html" class="btn btn-neutral float-right" title="GEM-PRO - List of Gene IDs" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../gempro.html" class="btn btn-neutral" title="The GEM-PRO Pipeline" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Nathan Mih.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'1.0.0a0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>