

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>ssbio.pipeline.gempro &mdash; ssbio 1.0.0a0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="../../../_static/theme_overrides.css" type="text/css" />
  

  
        <link rel="index" title="Index"
              href="../../../genindex.html"/>
        <link rel="search" title="Search" href="../../../search.html"/>
    <link rel="top" title="ssbio 1.0.0a0 documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> ssbio
          

          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../getting_started.html">Getting Started</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../getting_started.html#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../getting_started.html#the-basics">The basics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../getting_started.html#structural-biology">Structural biology</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../getting_started.html#protein-structures">Protein structures</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../getting_started.html#biopython">Biopython</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../getting_started.html#systems-biology">Systems biology</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../getting_started.html#cobra-models-cobrapy">COBRA models &amp; COBRApy</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../getting_started.html#reading">Reading</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../structure.html">The Structure Class</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../structure.html#introduction">Introduction</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../protein.html">The Protein Class</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../protein.html#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../protein.html#tutorials">Tutorials</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../notebooks/Protein - Structure Mapping, Alignments, and Visualization.html">Protein - Structure Mapping, Alignments, and Visualization</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Imports">Imports</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Initialization-of-the-project">Initialization of the project</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Mapping-sequence--&gt;-structure">Mapping sequence -&gt; structure</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Downloading-and-ranking-structures">Downloading and ranking structures</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Loading-and-aligning-new-sequences">Loading and aligning new sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Viewing-structures">Viewing structures</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Mapping-sequence-residue-numbers-to-structure-residue-numbers">Mapping sequence residue numbers to structure residue numbers</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Saving">Saving</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../gempro.html">The GEM-PRO Pipeline</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../gempro.html#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../gempro.html#tutorials">Tutorials</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../notebooks/GEM-PRO - Genes &amp; Sequences.html">GEM-PRO - Genes &amp; Sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/GEM-PRO - Genes &amp; Sequences.html#Imports">Imports</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/GEM-PRO - Genes &amp; Sequences.html#Initialization-of-the-project">Initialization of the project</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/GEM-PRO - Genes &amp; Sequences.html#Mapping-sequence--&gt;-structure">Mapping sequence -&gt; structure</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/GEM-PRO - Genes &amp; Sequences.html#Downloading-and-ranking-structures">Downloading and ranking structures</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../notebooks/GEM-PRO - List of Gene IDs.html">GEM-PRO - List of Gene IDs</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/GEM-PRO - List of Gene IDs.html#Imports">Imports</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/GEM-PRO - List of Gene IDs.html#Initialization-of-the-project">Initialization of the project</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/GEM-PRO - List of Gene IDs.html#Mapping-gene-ID--&gt;-sequence">Mapping gene ID -&gt; sequence</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/GEM-PRO - List of Gene IDs.html#Mapping-representative-sequence-to-structure">Mapping representative sequence to structure</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/GEM-PRO - List of Gene IDs.html#Downloading-and-ranking-structures">Downloading and ranking structures</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../notebooks/GEM-PRO - SBML Model (iNJ661).html">GEM-PRO - SBML Model (iNJ661)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/GEM-PRO - SBML Model (iNJ661).html#Imports">Imports</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/GEM-PRO - SBML Model (iNJ661).html#Initialization-of-the-project">Initialization of the project</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/GEM-PRO - SBML Model (iNJ661).html#Mapping-gene-ID--&gt;-sequence">Mapping gene ID -&gt; sequence</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/GEM-PRO - SBML Model (iNJ661).html#Mapping-representative-sequence-to-structure">Mapping representative sequence to structure</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/GEM-PRO - SBML Model (iNJ661).html#Downloading-and-ranking-structures">Downloading and ranking structures</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/GEM-PRO - SBML Model (iNJ661).html#Creating-homology-models">Creating homology models</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../notebooks/GEM-PRO - SBML Model (iNJ661).html#Saving-your-GEM-PRO">Saving your GEM-PRO</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../atlas.html">The ATLAS Pipeline</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../atlas.html#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../atlas.html#tutorials">Tutorials</a><ul class="simple">
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../python_api.html">Python API</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../../../index.html">ssbio</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          





<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../../index.html">Docs</a> &raquo;</li>
      
          <li><a href="../../index.html">Module code</a> &raquo;</li>
      
    <li>ssbio.pipeline.gempro</li>
      <li class="wy-breadcrumbs-aside">
        
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for ssbio.pipeline.gempro</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">os.path</span> <span class="k">as</span> <span class="nn">op</span>
<span class="kn">import</span> <span class="nn">shutil</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">from</span> <span class="nn">copy</span> <span class="k">import</span> <span class="n">copy</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">requests</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">Bio.PDB.PDBExceptions</span> <span class="k">import</span> <span class="n">PDBException</span>
<span class="kn">from</span> <span class="nn">bioservices</span> <span class="k">import</span> <span class="n">KEGG</span>
<span class="kn">from</span> <span class="nn">bioservices</span> <span class="k">import</span> <span class="n">UniProt</span>
<span class="kn">from</span> <span class="nn">cobra.core</span> <span class="k">import</span> <span class="n">DictList</span>
<span class="kn">from</span> <span class="nn">ssbio.databases.kegg</span> <span class="k">import</span> <span class="n">KEGGProp</span>
<span class="kn">from</span> <span class="nn">ssbio.databases.uniprot</span> <span class="k">import</span> <span class="n">UniProtProp</span>
<span class="kn">from</span> <span class="nn">slugify</span> <span class="k">import</span> <span class="n">Slugify</span>
<span class="kn">import</span> <span class="nn">ssbio.cobra.utils</span>
<span class="kn">import</span> <span class="nn">ssbio.databases.kegg</span>
<span class="kn">import</span> <span class="nn">ssbio.databases.pdb</span>
<span class="kn">import</span> <span class="nn">ssbio.databases.uniprot</span>
<span class="kn">import</span> <span class="nn">ssbio.sequence.properties.residues</span>
<span class="kn">import</span> <span class="nn">ssbio.sequence.properties.tmhmm</span>
<span class="kn">import</span> <span class="nn">ssbio.sequence.utils.fasta</span>
<span class="kn">import</span> <span class="nn">ssbio.structure.properties.msms</span>
<span class="kn">import</span> <span class="nn">ssbio.structure.properties.quality</span>
<span class="kn">import</span> <span class="nn">ssbio.structure.properties.residues</span>
<span class="kn">from</span> <span class="nn">ssbio</span> <span class="k">import</span> <span class="n">utils</span>
<span class="kn">from</span> <span class="nn">ssbio.core.genepro</span> <span class="k">import</span> <span class="n">GenePro</span>
<span class="kn">from</span> <span class="nn">ssbio.core.object</span> <span class="k">import</span> <span class="n">Object</span>
<span class="kn">from</span> <span class="nn">ssbio.sequence.properties.scratch</span> <span class="k">import</span> <span class="n">SCRATCH</span>
<span class="kn">from</span> <span class="nn">ssbio.sequence.seqprop</span> <span class="k">import</span> <span class="n">SeqProp</span>
<span class="kn">from</span> <span class="nn">ssbio.structure.homology.itasser.itasserprep</span> <span class="k">import</span> <span class="n">ITASSERPrep</span>

<span class="k">if</span> <span class="n">utils</span><span class="o">.</span><span class="n">is_ipynb</span><span class="p">():</span>
    <span class="kn">from</span> <span class="nn">tqdm</span> <span class="k">import</span> <span class="n">tqdm_notebook</span> <span class="k">as</span> <span class="n">tqdm</span>
<span class="k">else</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">tqdm</span> <span class="k">import</span> <span class="n">tqdm</span>

<span class="n">custom_slugify</span> <span class="o">=</span> <span class="n">Slugify</span><span class="p">(</span><span class="n">safe_chars</span><span class="o">=</span><span class="s1">&#39;-_.&#39;</span><span class="p">)</span>
<span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s2">&quot;requests&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">ERROR</span><span class="p">)</span>
<span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s2">&quot;urllib3&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">ERROR</span><span class="p">)</span>
<span class="n">log</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="n">__name__</span><span class="p">)</span>
<span class="n">date</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">Date</span><span class="p">()</span>
<span class="n">bs_unip</span> <span class="o">=</span> <span class="n">UniProt</span><span class="p">()</span>
<span class="n">bs_kegg</span> <span class="o">=</span> <span class="n">KEGG</span><span class="p">()</span>


<span class="k">class</span> <span class="nc">GEMPRO</span><span class="p">(</span><span class="n">Object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Generic class to represent all information for a GEM-PRO project.</span>

<span class="sd">    Initialize the GEM-PRO project with a genome-scale model, a list of genes, or a dict of genes and sequences.</span>
<span class="sd">    Specify the name of your project, along with the root directory where a folder with that name will be created.</span>

<span class="sd">    Main methods provided are:</span>

<span class="sd">    #. Mapping of sequence IDs</span>
<span class="sd">        * With KEGG mapper</span>
<span class="sd">        * With UniProt mapper</span>
<span class="sd">        * Allowing manual gene ID --&gt; protein sequence entry</span>
<span class="sd">        * Allowing manual gene ID --&gt; UniProt ID</span>

<span class="sd">    #. Consolidating sequence IDs and setting a representative sequence</span>

<span class="sd">    #. Mapping of representative sequence --&gt; structures</span>
<span class="sd">        * With UniProt --&gt; ranking of PDB structures</span>
<span class="sd">        * BLAST representative sequence --&gt; PDB database</span>

<span class="sd">    #. Homology modeling</span>
<span class="sd">        * Mapping to existing models</span>
<span class="sd">        * Preparation for running I-TASSER</span>
<span class="sd">        * Parsing I-TASSER runs</span>

<span class="sd">    #. Running QC/QA on structures and setting a representative structure</span>

<span class="sd">    Each step may generate reports as a Pandas DataFrame and also output various logging messages.</span>

<span class="sd">    Args:</span>
<span class="sd">        gem_name (str): The name of your GEM or just your project in general.</span>
<span class="sd">            This will be the name of the main folder that is created in root_dir.</span>
<span class="sd">        pdb_file_type (str): pdb, pdb.gz, mmcif, cif, cif.gz, xml.gz, mmtf, mmtf.gz - Choose a file type for</span>
<span class="sd">            files downloaded from the PDB</span>
<span class="sd">        create_dirs (bool): If GEM-PRO directories should be created - if not,</span>
<span class="sd">            most mapping functions need an output directory specified.</span>
<span class="sd">        root_dir (str): Path to where the folder named gem_name will be created.</span>
<span class="sd">            Default is current working directory.</span>
<span class="sd">        genome_path (str): Simple reference link to the genome FASTA file (CDS)</span>
<span class="sd">        gem (Model): COBRApy Model object</span>
<span class="sd">        gem_file_path (str): Path to GEM file</span>
<span class="sd">        gem_file_type (str): GEM model type - &#39;sbml&#39; (or &#39;xml&#39;), &#39;mat&#39;, or &#39;json&#39; formats</span>
<span class="sd">        genes_list (list): List of gene IDs that you want to map</span>
<span class="sd">        genes_and_sequences (dict): Dictionary of gene IDs and their amino acid sequence strings</span>
<span class="sd">        description (str): Optional string to describe your project</span>

<span class="sd">    Attributes:</span>
<span class="sd">        genes (DictList): A DictList of GenePro objects</span>
<span class="sd">        model (Model): A COBRApy Model object</span>


<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gem_name</span><span class="p">,</span> <span class="n">pdb_file_type</span><span class="o">=</span><span class="s1">&#39;cif&#39;</span><span class="p">,</span> <span class="n">create_dirs</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">root_dir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">genome_path</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">gem</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">gem_file_path</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">gem_file_type</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">genes_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">genes_and_sequences</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">description</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="n">Object</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">gem_name</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="n">description</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">pdb_file_type</span> <span class="o">=</span> <span class="n">pdb_file_type</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome_path</span> <span class="o">=</span> <span class="n">genome_path</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="c1"># Load a model object</span>
        <span class="k">if</span> <span class="n">gem</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_load_cobra_model</span><span class="p">(</span><span class="n">gem</span><span class="p">)</span>

        <span class="c1"># Or, load a GEM file</span>
        <span class="k">elif</span> <span class="n">gem_file_path</span> <span class="ow">and</span> <span class="n">gem_file_type</span><span class="p">:</span>
            <span class="n">gem_manual</span> <span class="o">=</span> <span class="n">ssbio</span><span class="o">.</span><span class="n">cobra</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">model_loader</span><span class="p">(</span><span class="n">gem_file_path</span><span class="p">,</span> <span class="n">gem_file_type</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_load_cobra_model</span><span class="p">(</span><span class="n">gem_manual</span><span class="p">)</span>

        <span class="c1"># Or, load a list of gene IDs</span>
        <span class="k">elif</span> <span class="n">genes_list</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="n">genes_list</span>

        <span class="c1"># Or, load a dictionary of genes and their sequences</span>
        <span class="k">elif</span> <span class="n">genes_and_sequences</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">genes_and_sequences</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

        <span class="c1"># If neither a model or genes are input, you can still add IDs with method add_genes_by_id later</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">log</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s1">&#39;No model or genes input&#39;</span><span class="p">)</span>

        <span class="c1"># Create directories</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_root_dir</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">create_dirs</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">root_dir</span><span class="p">:</span>
                <span class="n">root_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">root_dir</span> <span class="o">=</span> <span class="n">root_dir</span>
        <span class="k">if</span> <span class="n">genes_and_sequences</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">manual_seq_mapping</span><span class="p">(</span><span class="n">genes_and_sequences</span><span class="p">)</span>

        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: number of genes&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)))</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">root_dir</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;str: Directory where GEM-PRO project folder is located&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_root_dir</span>

    <span class="nd">@root_dir</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">root_dir</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">path</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: folder does not exist&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">path</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_root_dir</span> <span class="o">=</span> <span class="n">path</span>

        <span class="k">if</span> <span class="n">path</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">base_dir</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">model_dir</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes_dir</span><span class="p">]:</span>
                <span class="n">ssbio</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">make_dir</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>

            <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: GEM-PRO project location&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">base_dir</span><span class="p">))</span>

            <span class="c1"># Additionally set the gene folder paths</span>
            <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">:</span>
                <span class="n">g</span><span class="o">.</span><span class="n">root_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes_dir</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s1">&#39;Root directory not set&#39;</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">base_dir</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;str: GEM-PRO project folder&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">root_dir</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">root_dir</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s1">&#39;Root directory not set&#39;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">model_dir</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;str: Directory where original GEMs and GEM-related files are stored&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">root_dir</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">base_dir</span><span class="p">,</span> <span class="s1">&#39;model&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s1">&#39;Root directory not set&#39;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">data_dir</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;str: Directory where all data (dataframes and more) will be stored&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">root_dir</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">base_dir</span><span class="p">,</span> <span class="s1">&#39;data&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s1">&#39;Root directory not set&#39;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">genes_dir</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;str: Directory where all gene specific information is stored&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">root_dir</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">base_dir</span><span class="p">,</span> <span class="s1">&#39;genes&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s1">&#39;Root directory not set&#39;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">_load_cobra_model</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Load a COBRApy Model object into the GEM-PRO project.</span>

<span class="sd">        Args:</span>
<span class="sd">            model (Model): COBRApy Model object</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">=</span> <span class="n">model</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">genes</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: loaded model&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">id</span><span class="p">))</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: number of reactions&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">reactions</span><span class="p">)))</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: number of reactions linked to a gene&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">ssbio</span><span class="o">.</span><span class="n">cobra</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">true_num_reactions</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">)))</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: number of genes (excluding spontaneous)&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">ssbio</span><span class="o">.</span><span class="n">cobra</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">true_num_genes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">)))</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: number of metabolites&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">metabolites</span><span class="p">)))</span>
        <span class="n">log</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s1">&#39;IMPORTANT: All Gene objects have been transformed into GenePro objects, and will be for any new ones&#39;</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">genes_with_structures</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return a DictList of all genes with any available protein structures&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">DictList</span><span class="p">(</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">num_structures</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">genes_with_experimental_structures</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return a DictList of all genes that have at least one experimental structure&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">DictList</span><span class="p">(</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes_with_structures</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">num_structures_experimental</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">genes_with_homology_models</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return a DictList of all genes that have at least one homology model&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">DictList</span><span class="p">(</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes_with_structures</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">num_structures_homology</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">genes_with_a_representative_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return a DictList of all genes with a representative sequence&quot;&quot;&quot;</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">DictList</span><span class="p">(</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">DictList</span><span class="p">(</span><span class="n">y</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">tmp</span> <span class="k">if</span> <span class="n">y</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="o">.</span><span class="n">seq_record</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">genes_with_a_representative_structure</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return a DictList of all genes with a representative protein structure&quot;&quot;&quot;</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">DictList</span><span class="p">(</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_structure</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">DictList</span><span class="p">(</span><span class="n">y</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">tmp</span> <span class="k">if</span> <span class="n">y</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_structure</span><span class="o">.</span><span class="n">structure_file</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">genes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genes</span>

    <span class="nd">@genes</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">genes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genes_list</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Set the genes attribute to be a DictList of GenePro objects.</span>

<span class="sd">        A &quot;protein&quot; attribute will be added to each Gene.</span>

<span class="sd">        Args:</span>
<span class="sd">            genes_list: DictList of COBRApy Gene objects, or list of gene IDs</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">genes_list</span><span class="p">,</span> <span class="n">DictList</span><span class="p">):</span>
            <span class="n">tmp_list</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">genes_list</span><span class="p">)):</span>
                <span class="n">x</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
                <span class="n">new_gene</span> <span class="o">=</span> <span class="n">GenePro</span><span class="p">(</span><span class="nb">id</span><span class="o">=</span><span class="n">x</span><span class="p">,</span> <span class="n">pdb_file_type</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">pdb_file_type</span><span class="p">)</span>
                <span class="n">tmp_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_gene</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_genes</span> <span class="o">=</span> <span class="n">DictList</span><span class="p">(</span><span class="n">tmp_list</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_genes</span> <span class="o">=</span> <span class="n">genes_list</span>

    <span class="k">def</span> <span class="nf">add_genes_by_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genes_list</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Add gene IDs manually into our GEM-PRO project.</span>

<span class="sd">        Args:</span>
<span class="sd">            genes_list (list): List of gene IDs as strings.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">new_genes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">genes_list</span><span class="p">)):</span>
            <span class="n">new_gene</span> <span class="o">=</span> <span class="n">GenePro</span><span class="p">(</span><span class="nb">id</span><span class="o">=</span><span class="n">x</span><span class="p">,</span> <span class="n">pdb_file_type</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">pdb_file_type</span><span class="p">)</span>
            <span class="n">new_genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_gene</span><span class="p">)</span>

        <span class="c1"># Add unique genes only</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">new_genes</span><span class="p">)</span>

    <span class="c1">####################################################################################################################</span>
    <span class="c1">### SEQUENCE RELATED METHODS ###</span>
<div class="viewcode-block" id="GEMPRO.kegg_mapping_and_metadata"><a class="viewcode-back" href="../../../notebooks/GEM-PRO - SBML Model (iNJ661).html#ssbio.pipeline.gempro.GEMPRO.kegg_mapping_and_metadata">[docs]</a>    <span class="k">def</span> <span class="nf">kegg_mapping_and_metadata</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">kegg_organism_code</span><span class="p">,</span> <span class="n">custom_gene_mapping</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                  <span class="n">set_as_representative</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">force_rerun</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Map all genes in the model to KEGG IDs using the KEGG service.</span>

<span class="sd">        This function does these things:</span>
<span class="sd">            1. Download all metadata and sequence files in the sequences directory</span>
<span class="sd">            2. Creates a KEGGProp object in the protein.sequences attribute</span>
<span class="sd">            3. Returns a Pandas DataFrame of mapping results</span>

<span class="sd">        Args:</span>
<span class="sd">            kegg_organism_code (str): The three letter KEGG code of your organism</span>
<span class="sd">            custom_gene_mapping (dict): If your model genes differ from the gene IDs you want to map,</span>
<span class="sd">                custom_gene_mapping allows you to input a dictionary which maps model gene IDs to new ones.</span>
<span class="sd">                Dictionary keys must match model gene IDs.</span>
<span class="sd">            outdir (str): Path to output directory of downloaded files, must be set if GEM-PRO directories</span>
<span class="sd">                were not created initially</span>
<span class="sd">            set_as_representative (bool): If mapped KEGG IDs should be set as representative sequences</span>
<span class="sd">            force_rerun (bool): If you want to overwrite any existing mappings and files</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># First map all of the organism&#39;s KEGG genes to UniProt</span>
        <span class="n">kegg_to_uniprot</span> <span class="o">=</span> <span class="n">ssbio</span><span class="o">.</span><span class="n">databases</span><span class="o">.</span><span class="n">kegg</span><span class="o">.</span><span class="n">map_kegg_all_genes</span><span class="p">(</span><span class="n">organism_code</span><span class="o">=</span><span class="n">kegg_organism_code</span><span class="p">,</span> <span class="n">target_db</span><span class="o">=</span><span class="s1">&#39;uniprot&#39;</span><span class="p">)</span>

        <span class="n">successfully_mapped_counter</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">custom_gene_mapping</span><span class="p">:</span>
                <span class="n">kegg_g</span> <span class="o">=</span> <span class="n">custom_gene_mapping</span><span class="p">[</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">kegg_g</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">id</span>

            <span class="c1"># Download both FASTA and KEGG metadata files</span>
            <span class="n">kegg_prop</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">load_kegg</span><span class="p">(</span><span class="n">kegg_id</span><span class="o">=</span><span class="n">kegg_g</span><span class="p">,</span> <span class="n">kegg_organism_code</span><span class="o">=</span><span class="n">kegg_organism_code</span><span class="p">,</span>
                                            <span class="n">download</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="n">outdir</span><span class="p">,</span> <span class="n">set_as_representative</span><span class="o">=</span><span class="n">set_as_representative</span><span class="p">,</span>
                                            <span class="n">force_rerun</span><span class="o">=</span><span class="n">force_rerun</span><span class="p">)</span>

            <span class="c1"># Update potentially old UniProt ID</span>
            <span class="k">if</span> <span class="n">kegg_g</span> <span class="ow">in</span> <span class="n">kegg_to_uniprot</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">kegg_prop</span><span class="o">.</span><span class="n">uniprot</span> <span class="o">=</span> <span class="n">kegg_to_uniprot</span><span class="p">[</span><span class="n">kegg_g</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="o">.</span><span class="n">kegg</span> <span class="o">==</span> <span class="n">kegg_prop</span><span class="o">.</span><span class="n">kegg</span><span class="p">:</span>
                        <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="o">.</span><span class="n">uniprot</span> <span class="o">=</span> <span class="n">kegg_to_uniprot</span><span class="p">[</span><span class="n">kegg_g</span><span class="p">]</span>

            <span class="c1"># Keep track of missing mappings - missing is defined by no available sequence</span>
            <span class="k">if</span> <span class="n">kegg_prop</span><span class="o">.</span><span class="n">sequence_file</span><span class="p">:</span>
                <span class="n">successfully_mapped_counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: loaded KEGG information for gene&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">))</span>

        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">/</span><span class="si">{}</span><span class="s1">: number of genes mapped to KEGG&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">successfully_mapped_counter</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)))</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Completed ID mapping --&gt; KEGG. See the &quot;df_kegg_metadata&quot; attribute for a summary dataframe.&#39;</span><span class="p">)</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">df_kegg_metadata</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">kegg_pre_df</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">df_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">,</span> <span class="s1">&#39;kegg&#39;</span><span class="p">,</span> <span class="s1">&#39;refseq&#39;</span><span class="p">,</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span> <span class="s1">&#39;num_pdbs&#39;</span><span class="p">,</span> <span class="s1">&#39;pdbs&#39;</span><span class="p">,</span> <span class="s1">&#39;seq_len&#39;</span><span class="p">,</span> <span class="s1">&#39;sequence_file&#39;</span><span class="p">,</span> <span class="s1">&#39;metadata_file&#39;</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">:</span>
            <span class="n">kegg_mappings</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">filter_sequences</span><span class="p">(</span><span class="n">KEGGProp</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">kegg_prop</span> <span class="ow">in</span> <span class="n">kegg_mappings</span><span class="p">:</span>
                <span class="n">kegg_dict</span> <span class="o">=</span> <span class="n">kegg_prop</span><span class="o">.</span><span class="n">get_dict</span><span class="p">(</span><span class="n">df_format</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">only_keys</span><span class="o">=</span><span class="n">df_cols</span><span class="p">)</span>
                <span class="n">kegg_dict</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">id</span>
                <span class="n">kegg_pre_df</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">kegg_dict</span><span class="p">)</span>

        <span class="c1"># Save a dataframe of the file mapping info</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">kegg_pre_df</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">df_cols</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;gene&#39;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">ssbio</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">clean_df</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">missing_kegg_mapping</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">kegg_missing</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">:</span>
            <span class="n">keggs</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">filter_sequences</span><span class="p">(</span><span class="n">KEGGProp</span><span class="p">)</span>
            <span class="n">no_sequence_file_available</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">keggs</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">k</span><span class="o">.</span><span class="n">sequence_file</span><span class="p">:</span>
                    <span class="n">no_sequence_file_available</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">break</span>
            <span class="k">if</span> <span class="n">no_sequence_file_available</span><span class="p">:</span>
                <span class="n">kegg_missing</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">kegg_missing</span><span class="p">))</span>

<div class="viewcode-block" id="GEMPRO.uniprot_mapping_and_metadata"><a class="viewcode-back" href="../../../notebooks/GEM-PRO - SBML Model (iNJ661).html#ssbio.pipeline.gempro.GEMPRO.uniprot_mapping_and_metadata">[docs]</a>    <span class="k">def</span> <span class="nf">uniprot_mapping_and_metadata</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model_gene_source</span><span class="p">,</span> <span class="n">custom_gene_mapping</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                     <span class="n">set_as_representative</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">force_rerun</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Map all genes in the model to UniProt IDs using the UniProt mapping service. Also download all metadata and sequences.</span>

<span class="sd">        Args:</span>
<span class="sd">            model_gene_source (str): the database source of your model gene IDs</span>
<span class="sd">                See: http://www.uniprot.org/help/programmatic_access</span>
<span class="sd">                Common model gene sources are:</span>
<span class="sd">                Ensembl Genomes - ENSEMBLGENOME_ID (i.e. E. coli b-numbers)</span>
<span class="sd">                Entrez Gene (GeneID) - P_ENTREZGENEID</span>
<span class="sd">                RefSeq Protein - P_REFSEQ_AC</span>
<span class="sd">            custom_gene_mapping (dict): If your model genes differ from the gene IDs you want to map,</span>
<span class="sd">                custom_gene_mapping allows you to input a dictionary which maps model gene IDs to new ones.</span>
<span class="sd">                Dictionary keys must match model genes.</span>
<span class="sd">            outdir (str): Path to output directory of downloaded files, must be set if GEM-PRO directories</span>
<span class="sd">                were not created initially</span>
<span class="sd">            set_as_representative (bool): If mapped UniProt IDs should be set as representative sequences</span>
<span class="sd">            force_rerun (bool): If you want to overwrite any existing mappings and files</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Allow model gene --&gt; custom ID mapping ({&#39;TM_1012&#39;:&#39;TM1012&#39;})</span>
        <span class="k">if</span> <span class="n">custom_gene_mapping</span><span class="p">:</span>
            <span class="n">genes_to_map</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">custom_gene_mapping</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">genes_to_map</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">]</span>

        <span class="c1"># Map all IDs first to available UniProts</span>
        <span class="n">genes_to_uniprots</span> <span class="o">=</span> <span class="n">bs_unip</span><span class="o">.</span><span class="n">mapping</span><span class="p">(</span><span class="n">fr</span><span class="o">=</span><span class="n">model_gene_source</span><span class="p">,</span> <span class="n">to</span><span class="o">=</span><span class="s1">&#39;ACC&#39;</span><span class="p">,</span> <span class="n">query</span><span class="o">=</span><span class="n">genes_to_map</span><span class="p">)</span>

        <span class="n">successfully_mapped_counter</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">custom_gene_mapping</span> <span class="ow">and</span> <span class="n">g</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">custom_gene_mapping</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">uniprot_gene</span> <span class="o">=</span> <span class="n">custom_gene_mapping</span><span class="p">[</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">uniprot_gene</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">id</span>

            <span class="k">if</span> <span class="n">uniprot_gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">genes_to_uniprots</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
                <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: unable to map to UniProt&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">mapped_uniprot</span> <span class="ow">in</span> <span class="n">genes_to_uniprots</span><span class="p">[</span><span class="n">uniprot_gene</span><span class="p">]:</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">uniprot_prop</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">load_uniprot</span><span class="p">(</span><span class="n">uniprot_id</span><span class="o">=</span><span class="n">mapped_uniprot</span><span class="p">,</span> <span class="n">download</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="n">outdir</span><span class="p">,</span>
                                                              <span class="n">set_as_representative</span><span class="o">=</span><span class="n">set_as_representative</span><span class="p">,</span>
                                                              <span class="n">force_rerun</span><span class="o">=</span><span class="n">force_rerun</span><span class="p">)</span>
                    <span class="k">except</span> <span class="n">requests</span><span class="o">.</span><span class="n">HTTPError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                        <span class="nb">print</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
                        <span class="n">log</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">, </span><span class="si">{}</span><span class="s1">: unable to complete web request&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">mapped_uniprot</span><span class="p">))</span>
                        <span class="k">continue</span>

                    <span class="k">if</span> <span class="n">uniprot_prop</span><span class="o">.</span><span class="n">sequence_file</span><span class="p">:</span>
                        <span class="n">successfully_mapped_counter</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">/</span><span class="si">{}</span><span class="s1">: number of genes mapped to UniProt&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">successfully_mapped_counter</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)))</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Completed ID mapping --&gt; UniProt. See the &quot;df_uniprot_metadata&quot; attribute for a summary dataframe.&#39;</span><span class="p">)</span></div>

<div class="viewcode-block" id="GEMPRO.manual_uniprot_mapping"><a class="viewcode-back" href="../../../notebooks/GEM-PRO - SBML Model (iNJ661).html#ssbio.pipeline.gempro.GEMPRO.manual_uniprot_mapping">[docs]</a>    <span class="k">def</span> <span class="nf">manual_uniprot_mapping</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_to_uniprot_dict</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">set_as_representative</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Read a manual dictionary of model gene IDs --&gt; UniProt IDs. By default sets them as representative.</span>

<span class="sd">        This allows for mapping of the missing genes, or overriding of automatic mappings.</span>

<span class="sd">        Args:</span>
<span class="sd">            gene_to_uniprot_dict: Dictionary of mappings with key:value pairs:</span>
<span class="sd">                &lt;gene_id1&gt;:&lt;uniprot_id1&gt;,</span>
<span class="sd">                &lt;gene_id2&gt;:&lt;uniprot_id2&gt;,</span>
<span class="sd">                ...</span>
<span class="sd">            outdir (str): Path to output directory of downloaded files, must be set if GEM-PRO directories</span>
<span class="sd">                were not created initially</span>
<span class="sd">            set_as_representative (bool): If mapped UniProt IDs should be set as representative sequences</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">g</span><span class="p">,</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">gene_to_uniprot_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">g</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">g</span><span class="p">)</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">g</span><span class="p">)</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="n">uniprot_prop</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">load_uniprot</span><span class="p">(</span><span class="n">uniprot_id</span><span class="o">=</span><span class="n">u</span><span class="p">,</span>
                                                         <span class="n">outdir</span><span class="o">=</span><span class="n">outdir</span><span class="p">,</span> <span class="n">download</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                                         <span class="n">set_as_representative</span><span class="o">=</span><span class="n">set_as_representative</span><span class="p">)</span>
            <span class="k">except</span> <span class="n">requests</span><span class="o">.</span><span class="n">HTTPError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
                <span class="n">log</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">, </span><span class="si">{}</span><span class="s1">: unable to complete web request&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="n">u</span><span class="p">))</span>
                <span class="k">continue</span>

        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Completed manual ID mapping --&gt; UniProt. See the &quot;df_uniprot_metadata&quot; attribute for a summary dataframe.&#39;</span><span class="p">)</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">df_uniprot_metadata</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">uniprot_pre_df</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">df_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">,</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span> <span class="s1">&#39;reviewed&#39;</span><span class="p">,</span> <span class="s1">&#39;gene_name&#39;</span><span class="p">,</span> <span class="s1">&#39;kegg&#39;</span><span class="p">,</span> <span class="s1">&#39;refseq&#39;</span><span class="p">,</span> <span class="s1">&#39;num_pdbs&#39;</span><span class="p">,</span> <span class="s1">&#39;pdbs&#39;</span><span class="p">,</span> <span class="s1">&#39;ec_number&#39;</span><span class="p">,</span>
                   <span class="s1">&#39;pfam&#39;</span><span class="p">,</span> <span class="s1">&#39;seq_len&#39;</span><span class="p">,</span> <span class="s1">&#39;description&#39;</span><span class="p">,</span> <span class="s1">&#39;entry_version&#39;</span><span class="p">,</span> <span class="s1">&#39;seq_version&#39;</span><span class="p">,</span> <span class="s1">&#39;sequence_file&#39;</span><span class="p">,</span> <span class="s1">&#39;metadata_file&#39;</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">:</span>
            <span class="n">uniprot_mappings</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">filter_sequences</span><span class="p">(</span><span class="n">UniProtProp</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">uniprot_prop</span> <span class="ow">in</span> <span class="n">uniprot_mappings</span><span class="p">:</span>
                <span class="n">uniprot_dict</span> <span class="o">=</span> <span class="n">uniprot_prop</span><span class="o">.</span><span class="n">get_dict</span><span class="p">(</span><span class="n">df_format</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">only_keys</span><span class="o">=</span><span class="n">df_cols</span><span class="p">)</span>
                <span class="n">uniprot_dict</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">g</span>
                <span class="n">uniprot_pre_df</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">uniprot_dict</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">uniprot_pre_df</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">df_cols</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;gene&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">ssbio</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">clean_df</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">missing_uniprot_mapping</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">uniprot_missing</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">:</span>
            <span class="n">ups</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">filter_sequences</span><span class="p">(</span><span class="n">UniProtProp</span><span class="p">)</span>

            <span class="n">no_sequence_file_available</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">ups</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">u</span><span class="o">.</span><span class="n">sequence_file</span><span class="p">:</span>
                    <span class="n">no_sequence_file_available</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">break</span>
            <span class="k">if</span> <span class="n">no_sequence_file_available</span><span class="p">:</span>
                <span class="n">uniprot_missing</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>

        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">uniprot_missing</span><span class="p">))</span>

    <span class="c1"># TODO: should also have a seq --&gt; uniprot id function (has to be 100% match) (also needs organism)</span>
<div class="viewcode-block" id="GEMPRO.manual_seq_mapping"><a class="viewcode-back" href="../../../python_api.html#ssbio.pipeline.gempro.GEMPRO.manual_seq_mapping">[docs]</a>    <span class="k">def</span> <span class="nf">manual_seq_mapping</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_to_seq_dict</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">set_as_representative</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Read a manual input dictionary of model gene IDs --&gt; protein sequences. By default sets them as representative.</span>

<span class="sd">        Args:</span>
<span class="sd">            gene_to_seq_dict (dict): Mapping of gene IDs to their protein sequence strings</span>
<span class="sd">            outdir (str): Path to output directory of downloaded files, must be set if GEM-PRO directories</span>
<span class="sd">                were not created initially</span>
<span class="sd">            set_as_representative (bool): If mapped sequences should be set as representative</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">outdir</span><span class="p">:</span>
            <span class="n">outdir_set</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">outdir_set</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="c1"># Save the sequence information in individual FASTA files</span>
        <span class="k">for</span> <span class="n">g</span><span class="p">,</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">gene_to_seq_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">g</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">g</span><span class="p">)</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">get_by_id</span><span class="p">(</span><span class="n">g</span><span class="p">)</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">outdir_set</span><span class="p">:</span>
                <span class="n">outdir</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">sequence_dir</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">outdir</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Output directory must be specified&#39;</span><span class="p">)</span>

            <span class="n">manual_info</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">load_manual_sequence</span><span class="p">(</span><span class="n">ident</span><span class="o">=</span><span class="n">g</span><span class="p">,</span> <span class="n">seq</span><span class="o">=</span><span class="n">s</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="n">outdir</span><span class="p">,</span>
                                                            <span class="n">write_fasta_file</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                                            <span class="n">set_as_representative</span><span class="o">=</span><span class="n">set_as_representative</span><span class="p">)</span>
            <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: loaded manually defined sequence information&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="p">))</span>

        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Loaded in </span><span class="si">{}</span><span class="s1"> sequences&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_to_seq_dict</span><span class="p">)))</span></div>

    <span class="k">def</span> <span class="nf">set_representative_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Combine information from KEGG, UniProt, and manual mappings. Saves a DataFrame of results.</span>

<span class="sd">            Manual mappings override all existing mappings. UniProt mappings override KEGG mappings except</span>
<span class="sd">            when KEGG mappings have PDBs associated with them and UniProt doesn&#39;t.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># TODO: rethink use of multiple database sources - may lead to inconsistency with genome sources</span>

        <span class="n">sequence_missing</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">successfully_mapped_counter</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">):</span>
            <span class="n">repseq</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">set_representative_sequence</span><span class="p">()</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">repseq</span><span class="p">:</span>
                <span class="n">sequence_missing</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
            <span class="k">elif</span> <span class="ow">not</span> <span class="n">repseq</span><span class="o">.</span><span class="n">sequence_path</span><span class="p">:</span>
                <span class="n">sequence_missing</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">successfully_mapped_counter</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">/</span><span class="si">{}</span><span class="s1">: number of genes with a representative sequence&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes_with_a_representative_sequence</span><span class="p">),</span>
                                                                                <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)))</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;See the &quot;df_representative_sequences&quot; attribute for a summary dataframe.&#39;</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">df_representative_sequences</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">seq_mapping_pre_df</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">df_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">,</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span> <span class="s1">&#39;kegg&#39;</span><span class="p">,</span> <span class="s1">&#39;num_pdbs&#39;</span><span class="p">,</span> <span class="s1">&#39;pdbs&#39;</span><span class="p">,</span> <span class="s1">&#39;seq_len&#39;</span><span class="p">,</span> <span class="s1">&#39;sequence_file&#39;</span><span class="p">,</span> <span class="s1">&#39;metadata_file&#39;</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes_with_a_representative_sequence</span><span class="p">:</span>
            <span class="n">gene_dict</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="o">.</span><span class="n">get_dict</span><span class="p">(</span><span class="n">df_format</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">only_keys</span><span class="o">=</span><span class="n">df_cols</span><span class="p">)</span>
            <span class="n">gene_dict</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">id</span>
            <span class="n">seq_mapping_pre_df</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene_dict</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">seq_mapping_pre_df</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">df_cols</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;gene&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">ssbio</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">clean_df</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">missing_representative_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span> <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes_with_a_representative_sequence</span><span class="o">.</span><span class="n">has_id</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">id</span><span class="p">)]</span>

    <span class="k">def</span> <span class="nf">write_representative_sequences_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">outname</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">set_ids_from_model</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Write all the model&#39;s sequences as a single FASTA file. By default, sets IDs to model gene IDs.</span>

<span class="sd">        Args:</span>
<span class="sd">            outname (str): Name of the output FASTA file without the extension</span>
<span class="sd">            outdir (str): Path to output directory of downloaded files, must be set if GEM-PRO directories</span>
<span class="sd">                were not created initially</span>
<span class="sd">            set_ids_from_model (bool): If the gene ID source should be the model gene IDs, not the original sequence ID</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">outdir</span><span class="p">:</span>
            <span class="n">outdir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">outdir</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Output directory must be specified&#39;</span><span class="p">)</span>

        <span class="n">outfile</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="n">outname</span> <span class="o">+</span> <span class="s1">&#39;.faa&#39;</span><span class="p">)</span>

        <span class="n">tmp</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes_with_a_representative_sequence</span><span class="p">:</span>
            <span class="n">repseq</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span>
            <span class="n">copied_seq_record</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="n">repseq</span><span class="o">.</span><span class="n">seq_record</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">set_ids_from_model</span><span class="p">:</span>
                <span class="n">copied_seq_record</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">id</span>
            <span class="n">tmp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">copied_seq_record</span><span class="p">)</span>

        <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">tmp</span><span class="p">,</span> <span class="n">outfile</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: wrote all representative sequences to file&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">outfile</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome_path</span> <span class="o">=</span> <span class="n">outfile</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome_path</span>

    <span class="k">def</span> <span class="nf">get_sequence_properties</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Run Biopython ProteinAnalysis and EMBOSS pepstats to summarize basic statistics of the protein sequences.</span>

<span class="sd">        Annotations are stored in the gene protein&#39;s representative sequence at .seq_record.annotations</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes_with_a_representative_sequence</span><span class="p">):</span>
            <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="o">.</span><span class="n">get_biopython_pepstats</span><span class="p">()</span>
            <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="o">.</span><span class="n">get_emboss_pepstats</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">get_scratch_predictions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">path_to_scratch</span><span class="p">,</span> <span class="n">results_dir</span><span class="p">,</span> <span class="n">scratch_basename</span><span class="o">=</span><span class="s1">&#39;scratch&#39;</span><span class="p">,</span> <span class="n">num_cores</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                                <span class="n">exposed_buried_cutoff</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">custom_gene_mapping</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Run and parse SCRATCH results to predict secondary structure and solvent accessibility.</span>

<span class="sd">            Annotations are stored in the gene protein&#39;s representative sequence at:</span>
<span class="sd">                .seq_record.annotations and</span>
<span class="sd">                .seq_record.letter_annotations</span>

<span class="sd">        Args:</span>
<span class="sd">            path_to_scratch (str): Path to SCRATCH executable</span>
<span class="sd">            results_dir (str): Path to SCRATCH results folder, which will have the files (scratch.ss, scratch.ss8,</span>
<span class="sd">            scratch.acc, scratch.acc20)</span>
<span class="sd">            scratch_basename (str): Basename of the SCRATCH results (&#39;scratch&#39; is default)</span>
<span class="sd">            exposed_buried_cutoff (int): Cutoff of exposed/buried for the acc20 predictions</span>
<span class="sd">            custom_gene_mapping (dict): Default parsing of SCRATCH output files is to look for the model gene IDs. If</span>
<span class="sd">                your output files contain IDs which differ from the model gene IDs, use this dictionary to map model</span>
<span class="sd">                gene IDs to result file IDs. Dictionary keys must match model genes.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome_path</span><span class="p">:</span>
            <span class="c1"># Write all sequences as one file</span>
            <span class="n">all_seqs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">write_representative_sequences_file</span><span class="p">(</span><span class="n">outname</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">_cds&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">))</span>

        <span class="c1"># Runs SCRATCH or loads existing results in results_dir</span>
        <span class="n">scratch</span> <span class="o">=</span> <span class="n">SCRATCH</span><span class="p">(</span><span class="n">project_name</span><span class="o">=</span><span class="n">scratch_basename</span><span class="p">,</span> <span class="n">seq_file</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">genome_path</span><span class="p">)</span>
        <span class="n">scratch</span><span class="o">.</span><span class="n">run_scratch</span><span class="p">(</span><span class="n">path_to_scratch</span><span class="o">=</span><span class="n">path_to_scratch</span><span class="p">,</span> <span class="n">num_cores</span><span class="o">=</span><span class="n">num_cores</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="n">results_dir</span><span class="p">)</span>

        <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="c1"># Adding the scratch annotations to the representative_sequences letter_annotations</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes_with_a_representative_sequence</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">custom_gene_mapping</span><span class="p">:</span>
                <span class="n">g_id</span> <span class="o">=</span> <span class="n">custom_gene_mapping</span><span class="p">[</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">g_id</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">id</span>

            <span class="k">if</span> <span class="n">g_id</span> <span class="ow">in</span> <span class="n">scratch</span><span class="o">.</span><span class="n">sspro_summary</span><span class="p">():</span>
                <span class="c1"># Secondary structure</span>
                <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="o">.</span><span class="n">seq_record</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">scratch</span><span class="o">.</span><span class="n">sspro_summary</span><span class="p">()[</span><span class="n">g_id</span><span class="p">])</span>
                <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="o">.</span><span class="n">seq_record</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">scratch</span><span class="o">.</span><span class="n">sspro8_summary</span><span class="p">()[</span><span class="n">g_id</span><span class="p">])</span>
                <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="o">.</span><span class="n">load_letter_annotations</span><span class="p">(</span><span class="s1">&#39;SS-sspro&#39;</span><span class="p">,</span> <span class="n">scratch</span><span class="o">.</span><span class="n">sspro_results</span><span class="p">()[</span><span class="n">g_id</span><span class="p">])</span>
                <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="o">.</span><span class="n">load_letter_annotations</span><span class="p">(</span><span class="s1">&#39;SS-sspro8&#39;</span><span class="p">,</span> <span class="n">scratch</span><span class="o">.</span><span class="n">sspro8_results</span><span class="p">()[</span><span class="n">g_id</span><span class="p">])</span>

                <span class="c1"># Solvent accessibility</span>
                <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="o">.</span><span class="n">seq_record</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">scratch</span><span class="o">.</span><span class="n">accpro_summary</span><span class="p">()[</span><span class="n">g_id</span><span class="p">])</span>
                <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="o">.</span><span class="n">seq_record</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">scratch</span><span class="o">.</span><span class="n">accpro20_summary</span><span class="p">(</span><span class="n">exposed_buried_cutoff</span><span class="p">)[</span><span class="n">g_id</span><span class="p">])</span>
                <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="o">.</span><span class="n">load_letter_annotations</span><span class="p">(</span><span class="s1">&#39;RSA-accpro&#39;</span><span class="p">,</span> <span class="n">scratch</span><span class="o">.</span><span class="n">accpro_results</span><span class="p">()[</span><span class="n">g_id</span><span class="p">])</span>
                <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="o">.</span><span class="n">load_letter_annotations</span><span class="p">(</span><span class="s1">&#39;RSA-accpro20&#39;</span><span class="p">,</span> <span class="n">scratch</span><span class="o">.</span><span class="n">accpro20_results</span><span class="p">()[</span><span class="n">g_id</span><span class="p">])</span>

                <span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">log</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: missing SCRATCH results&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">))</span>

        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">/</span><span class="si">{}</span><span class="s1">: number of genes with SCRATCH predictions loaded&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">counter</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)))</span>

    <span class="k">def</span> <span class="nf">get_tmhmm_predictions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">tmhmm_results</span><span class="p">,</span> <span class="n">custom_gene_mapping</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Parse TMHMM results and store in the representative sequences</span>

<span class="sd">        Args:</span>
<span class="sd">            tmhmm_results (str): Path to TMHMM results (long format)</span>
<span class="sd">            custom_gene_mapping (dict): Default parsing of TMHMM output is to look for the model gene IDs. If</span>
<span class="sd">                your output file contains IDs which differ from the model gene IDs, use this dictionary to map model</span>
<span class="sd">                gene IDs to result file IDs. Dictionary keys must match model genes.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">tmhmm_dict</span> <span class="o">=</span> <span class="n">ssbio</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">properties</span><span class="o">.</span><span class="n">tmhmm</span><span class="o">.</span><span class="n">parse_tmhmm_long</span><span class="p">(</span><span class="n">tmhmm_results</span><span class="p">)</span>

        <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes_with_a_representative_sequence</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">custom_gene_mapping</span><span class="p">:</span>
                <span class="n">g_id</span> <span class="o">=</span> <span class="n">custom_gene_mapping</span><span class="p">[</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">g_id</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">id</span>

            <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">tmhmm_dict</span><span class="p">:</span>
                <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: loading TMHMM results&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">))</span>
                <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="o">.</span><span class="n">seq_record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s1">&#39;num_tm_helix-tmhmm&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tmhmm_dict</span><span class="p">[</span><span class="n">g_id</span><span class="p">][</span><span class="s1">&#39;num_tm_helices&#39;</span><span class="p">]</span>
                <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="o">.</span><span class="n">load_letter_annotations</span><span class="p">(</span><span class="s1">&#39;TM-tmhmm&#39;</span><span class="p">,</span> <span class="n">tmhmm_dict</span><span class="p">[</span><span class="n">g_id</span><span class="p">][</span><span class="s1">&#39;sequence&#39;</span><span class="p">])</span>
                <span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">log</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">: missing TMHMM results&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">))</span>

        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;[}/</span><span class="si">{}</span><span class="s1">: number of genes with TMHMM predictions loaded&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">counter</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)))</span>

    <span class="c1">### END SEQUENCE RELATED METHODS ###</span>
    <span class="c1">####################################################################################################################</span>

    <span class="c1">####################################################################################################################</span>
    <span class="c1">### STRUCTURE RELATED METHODS ###</span>
    <span class="k">def</span> <span class="nf">map_uniprot_to_pdb</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seq_ident_cutoff</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">force_rerun</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Map UniProt IDs to a ranked list of PDB structures available.</span>

<span class="sd">            Uses the &quot;Best structures&quot; API availble from https://www.ebi.ac.uk/pdbe/api/doc/sifts.html</span>
<span class="sd">            The list of PDB structures mapping to a UniProt accession sorted by coverage of the protein and,</span>
<span class="sd">            if the same, resolution. Also creates a summary dataframe accessible by the attribute &quot;df_pdb_ranking&quot;.</span>

<span class="sd">        Args:</span>
<span class="sd">            seq_ident_cutoff (float): Cutoff results based on percent coverage (in decimal form)</span>
<span class="sd">            outdir (str): Path to output directory of downloaded files, must be set if GEM-PRO directories</span>
<span class="sd">                were not created initially</span>
<span class="sd">            force_rerun (bool): Obtain best structures mapping ignoring previously downloaded results</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># First get all UniProt IDs and check if they have PDBs</span>
        <span class="n">all_representative_uniprots</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes_with_a_representative_sequence</span><span class="p">:</span>
            <span class="n">uniprot_id</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="o">.</span><span class="n">uniprot</span>
            <span class="k">if</span> <span class="n">uniprot_id</span><span class="p">:</span>
                <span class="n">all_representative_uniprots</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">uniprot_id</span><span class="p">)</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Mapping UniProt IDs --&gt; PDB IDs...&#39;</span><span class="p">)</span>
        <span class="n">uniprots_to_pdbs</span> <span class="o">=</span> <span class="n">bs_unip</span><span class="o">.</span><span class="n">mapping</span><span class="p">(</span><span class="n">fr</span><span class="o">=</span><span class="s1">&#39;ACC&#39;</span><span class="p">,</span> <span class="n">to</span><span class="o">=</span><span class="s1">&#39;PDB_ID&#39;</span><span class="p">,</span> <span class="n">query</span><span class="o">=</span><span class="n">all_representative_uniprots</span><span class="p">)</span>

        <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="c1"># Now run the best_structures API for all genes</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes_with_a_representative_sequence</span><span class="p">):</span>
            <span class="n">uniprot_id</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span><span class="o">.</span><span class="n">uniprot</span>
            <span class="k">if</span> <span class="n">uniprot_id</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">uniprot_id</span> <span class="ow">in</span> <span class="n">uniprots_to_pdbs</span><span class="p">:</span>
                    <span class="n">best_structures</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">map_uniprot_to_pdb</span><span class="p">(</span><span class="n">seq_ident_cutoff</span><span class="o">=</span><span class="n">seq_ident_cutoff</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="n">outdir</span><span class="p">,</span> <span class="n">force_rerun</span><span class="o">=</span><span class="n">force_rerun</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">best_structures</span><span class="p">:</span>
                        <span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: </span><span class="si">{}</span><span class="s1"> PDBs mapped&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">best_structures</span><span class="p">)))</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">, </span><span class="si">{}</span><span class="s1">: no PDBs available&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">uniprot_id</span><span class="p">))</span>

        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">/</span><span class="si">{}</span><span class="s1">: number of genes with at least one experimental structure&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes_with_experimental_structures</span><span class="p">),</span>
                                                                                          <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)))</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Completed UniProt --&gt; best PDB mapping. See the &quot;df_pdb_ranking&quot; attribute for a summary dataframe.&#39;</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">df_pdb_ranking</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes_with_experimental_structures</span><span class="p">:</span>
            <span class="n">protein_df</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">df_pdb_ranking</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">protein_df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                <span class="n">protein_df</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">id</span>
                <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_df</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">ssbio</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">clean_df</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;gene&#39;</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">blast_seqs_to_pdb</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seq_ident_cutoff</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">evalue</span><span class="o">=</span><span class="mf">0.0001</span><span class="p">,</span> <span class="n">all_genes</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">display_link</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                          <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">force_rerun</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;BLAST each gene sequence to the PDB. Saves raw BLAST results (XML files).</span>

<span class="sd">            Also creates a summary dataframe accessible by the attribute &quot;df_pdb_blast&quot;.</span>

<span class="sd">        Args:</span>
<span class="sd">            seq_ident_cutoff (float, optional): Cutoff results based on percent coverage (in decimal form)</span>
<span class="sd">            evalue (float, optional): Cutoff for the E-value - filters for significant hits. 0.001 is liberal, 0.0001 is stringent (default).</span>
<span class="sd">            all_genes (bool): If all genes should be BLASTed, or only those without any structures currently mapped</span>
<span class="sd">            display_link (bool, optional): Set to True if links to the HTML results should be displayed</span>
<span class="sd">            outdir (str): Path to output directory of downloaded files, must be set if GEM-PRO directories</span>
<span class="sd">                were not created initially</span>
<span class="sd">            force_rerun (bool, optional): If existing BLAST results should not be used, set to True. Default is False</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">):</span>
            <span class="c1"># If all_genes=False, BLAST only genes without a uniprot -&gt; pdb mapping</span>
            <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">num_structures</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">all_genes</span><span class="p">:</span>
                <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: skipping BLAST, </span><span class="si">{}</span><span class="s1"> structures already mapped and all_genes flag is False&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                                                                                                                <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">num_structures</span><span class="p">))</span>
                <span class="k">continue</span>

            <span class="c1"># BLAST the sequence to the PDB</span>
            <span class="n">new_pdbs</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">blast_representative_sequence_to_pdb</span><span class="p">(</span><span class="n">seq_ident_cutoff</span><span class="o">=</span><span class="n">seq_ident_cutoff</span><span class="p">,</span>
                                                                   <span class="n">evalue</span><span class="o">=</span><span class="n">evalue</span><span class="p">,</span>
                                                                   <span class="n">display_link</span><span class="o">=</span><span class="n">display_link</span><span class="p">,</span>
                                                                   <span class="n">outdir</span><span class="o">=</span><span class="n">outdir</span><span class="p">,</span>
                                                                   <span class="n">force_rerun</span><span class="o">=</span><span class="n">force_rerun</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">new_pdbs</span><span class="p">:</span>
                <span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: </span><span class="si">{}</span><span class="s1"> PDBs BLASTed&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_pdbs</span><span class="p">)))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: no BLAST results&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">))</span>

        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Completed sequence --&gt; PDB BLAST. See the &quot;df_pdb_blast&quot; attribute for a summary dataframe.&#39;</span><span class="p">)</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: number of genes with additional structures added from BLAST&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">counter</span><span class="p">))</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">df_pdb_blast</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes_with_experimental_structures</span><span class="p">:</span>
            <span class="n">protein_df</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">df_pdb_blast</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">protein_df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                <span class="n">protein_df</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">id</span>
                <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_df</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">ssbio</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">clean_df</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;gene&#39;</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">manual_homology_models</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">input_dict</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">clean</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">force_rerun</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Copy homology models to the GEM-PRO project.</span>

<span class="sd">        Args:</span>
<span class="sd">            input_dict: Dictionary of dictionaries of gene names to homology model IDs and information. Input a dict of:</span>
<span class="sd">                {model_gene: {homology_model_id1: {&#39;model_file&#39;: &#39;/path/to/homology/model&#39;,</span>
<span class="sd">                                                   &#39;file_type&#39;: &#39;pdb&#39;},</span>
<span class="sd">                              homology_model_id2: {&#39;model_file&#39;: &#39;/path/to/homology/model&#39;</span>
<span class="sd">                                                   &#39;file_type&#39;: &#39;pdb&#39;}</span>
<span class="sd">                             },</span>
<span class="sd">                }</span>
<span class="sd">            outdir (str): Path to output directory of downloaded files, must be set if GEM-PRO directories</span>
<span class="sd">                were not created initially</span>
<span class="sd">            clean (bool): If homology files should be cleaned and saved as a new PDB file</span>
<span class="sd">            force_rerun (bool): If homology files should be copied again even if they exist in the GEM-PRO directory</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">outdir</span><span class="p">:</span>
            <span class="n">outdir_set</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">outdir_set</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">input_dict</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">outdir_set</span><span class="p">:</span>
                <span class="n">outdir</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">structure_dir</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">outdir</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Output directory must be specified&#39;</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">hid</span><span class="p">,</span> <span class="n">hdict</span> <span class="ow">in</span> <span class="n">input_dict</span><span class="p">[</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="s1">&#39;model_file&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">hdict</span> <span class="ow">or</span> <span class="s1">&#39;file_type&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">hdict</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="s1">&#39;&quot;model_file&quot; and &quot;file_type&quot; must be keys in the manual input dictionary.&#39;</span><span class="p">)</span>

                <span class="n">new_homology</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">load_generic_structure</span><span class="p">(</span><span class="n">hid</span><span class="p">,</span> <span class="n">hdict</span><span class="p">[</span><span class="s1">&#39;model_file&#39;</span><span class="p">])</span>

                <span class="k">if</span> <span class="n">clean</span><span class="p">:</span>
                    <span class="n">new_homology</span><span class="o">.</span><span class="n">load_structure_path</span><span class="p">(</span><span class="n">new_homology</span><span class="o">.</span><span class="n">clean_structure</span><span class="p">(</span><span class="n">outdir</span><span class="o">=</span><span class="n">outdir</span><span class="p">,</span> <span class="n">force_rerun</span><span class="o">=</span><span class="n">force_rerun</span><span class="p">),</span>
                                                     <span class="n">hdict</span><span class="p">[</span><span class="s1">&#39;file_type&#39;</span><span class="p">])</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">ssbio</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">force_rerun</span><span class="p">(</span><span class="n">force_rerun</span><span class="p">,</span> <span class="n">op</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">hdict</span><span class="p">[</span><span class="s1">&#39;model_file&#39;</span><span class="p">])):</span>
                        <span class="c1"># Just copy the file to the structure directory and store the file name</span>
                        <span class="n">shutil</span><span class="o">.</span><span class="n">copy2</span><span class="p">(</span><span class="n">hdict</span><span class="p">[</span><span class="s1">&#39;model_file&#39;</span><span class="p">],</span> <span class="n">outdir</span><span class="p">)</span>
                        <span class="n">new_homology</span><span class="o">.</span><span class="n">load_structure_path</span><span class="p">(</span><span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="n">hdict</span><span class="p">[</span><span class="s1">&#39;model_file&#39;</span><span class="p">]),</span> <span class="n">hdict</span><span class="p">[</span><span class="s1">&#39;file_type&#39;</span><span class="p">])</span>

                <span class="c1"># TODO: need to handle other info in the provided dictionary, if any</span>
                <span class="n">new_homology</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">hdict</span><span class="p">)</span>

                <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: updated homology model information and copied model file.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">))</span>
            <span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Updated homology model information for </span><span class="si">{}</span><span class="s1"> genes.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">counter</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">get_itasser_models</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">homology_raw_dir</span><span class="p">,</span> <span class="n">custom_itasser_name_mapping</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">force_rerun</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Copy generated I-TASSER models from a directory to the GEM-PRO directory.</span>

<span class="sd">        Args:</span>
<span class="sd">            homology_raw_dir (str): Root directory of I-TASSER folders.</span>
<span class="sd">            custom_itasser_name_mapping (dict): Use this if your I-TASSER folder names differ from your model gene names.</span>
<span class="sd">                Input a dict of {model_gene: ITASSER_folder}.</span>
<span class="sd">            outdir (str): Path to output directory of downloaded files, must be set if GEM-PRO directories</span>
<span class="sd">                were not created initially</span>
<span class="sd">            force_rerun (bool): If homology files should be copied again even if they exist in the GEM-PRO directory</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">custom_itasser_name_mapping</span> <span class="ow">and</span> <span class="n">g</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">custom_itasser_name_mapping</span><span class="p">:</span>
                <span class="n">hom_id</span> <span class="o">=</span> <span class="n">custom_itasser_name_mapping</span><span class="p">[</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">]</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">homology_raw_dir</span><span class="p">,</span> <span class="n">hom_id</span><span class="p">)):</span>
                    <span class="n">hom_id</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">id</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">hom_id</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">id</span>

            <span class="c1"># The name of the actual pdb file will be $GENEID_model1.pdb</span>
            <span class="n">new_itasser_name</span> <span class="o">=</span> <span class="n">hom_id</span> <span class="o">+</span> <span class="s1">&#39;_model1&#39;</span>
            <span class="n">orig_itasser_dir</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">homology_raw_dir</span><span class="p">,</span> <span class="n">hom_id</span><span class="p">)</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="n">itasser_prop</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">load_itasser_folder</span><span class="p">(</span><span class="n">ident</span><span class="o">=</span><span class="n">hom_id</span><span class="p">,</span> <span class="n">itasser_folder</span><span class="o">=</span><span class="n">orig_itasser_dir</span><span class="p">,</span>
                                                             <span class="n">organize</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="n">outdir</span><span class="p">,</span> <span class="n">organize_name</span><span class="o">=</span><span class="n">new_itasser_name</span><span class="p">,</span>
                                                             <span class="n">create_dfs</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">force_rerun</span><span class="o">=</span><span class="n">force_rerun</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">OSError</span><span class="p">:</span>
                <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: homology model folder unavailable&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">))</span>
                <span class="k">continue</span>

            <span class="k">if</span> <span class="n">itasser_prop</span><span class="o">.</span><span class="n">model_file</span><span class="p">:</span>
                <span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: homology model file unavailable, perhaps modelling did not finish&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">))</span>

        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Completed copying of </span><span class="si">{}</span><span class="s1"> I-TASSER models to GEM-PRO directory. See the &quot;df_homology_models&quot; attribute for a summary dataframe.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">counter</span><span class="p">))</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">df_homology_models</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes_with_homology_models</span><span class="p">:</span>
            <span class="n">protein_df</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">df_homology_models</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">protein_df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                <span class="n">protein_df</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">id</span>
                <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_df</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">ssbio</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">clean_df</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;gene&#39;</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">set_representative_structure</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seq_outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">struct_outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pdb_file_type</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                     <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;needle&#39;</span><span class="p">,</span> <span class="n">always_use_homology</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                     <span class="n">seq_ident_cutoff</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">allow_missing_on_termini</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
                                     <span class="n">allow_mutants</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">allow_deletions</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                     <span class="n">allow_insertions</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">allow_unresolved</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                     <span class="n">force_rerun</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Set the representative protein structure for a gene.</span>

<span class="sd">        Each gene can have a combination of the following, which will be analyzed to set a representative structure.</span>
<span class="sd">            - Homology model(s)</span>
<span class="sd">            - Ranked PDBs</span>
<span class="sd">            - BLASTed PDBs</span>

<span class="sd">        If the always_use_homology flag is true, homology models are always set as representative when they exist.</span>
<span class="sd">            If there are multiple homology models, we rank by the percent sequence coverage.</span>

<span class="sd">        Args:</span>
<span class="sd">            seq_outdir (str): Path to output directory of sequence alignment files, must be set if GEM-PRO directories</span>
<span class="sd">                were not created initially</span>
<span class="sd">            struct_outdir (str): Path to output directory of structure files, must be set if GEM-PRO directories</span>
<span class="sd">                were not created initially</span>
<span class="sd">            pdb_file_type (str): pdb, pdb.gz, mmcif, cif, cif.gz, xml.gz, mmtf, mmtf.gz - PDB structure file type that</span>
<span class="sd">                should be downloaded</span>
<span class="sd">            engine (str): &quot;needle&quot; or &quot;biopython&quot; - which pairwise sequence alignment engine should be used</span>
<span class="sd">                needle is the standard EMBOSS tool to run pairwise alignments</span>
<span class="sd">                biopython is Biopython&#39;s implementation of needle. Results can differ!</span>
<span class="sd">            always_use_homology (bool): If homology models should always be set as the representative structure</span>
<span class="sd">            seq_ident_cutoff (float): Percent sequence identity cutoff, in decimal form</span>
<span class="sd">            allow_missing_on_termini (float): Percentage of the total length of the reference sequence which will be ignored</span>
<span class="sd">                when checking for modifications. Example: if 0.1, and reference sequence is 100 AA, then only residues</span>
<span class="sd">                5 to 95 will be checked for modifications.</span>
<span class="sd">            allow_mutants (bool): If mutations should be allowed or checked for</span>
<span class="sd">            allow_deletions (bool): If deletions should be allowed or checked for</span>
<span class="sd">            allow_insertions (bool): If insertions should be allowed or checked for</span>
<span class="sd">            allow_unresolved (bool): If unresolved residues should be allowed or checked for</span>
<span class="sd">            force_rerun (bool): If sequence to structure alignment should be rerun</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">):</span>
            <span class="n">repstruct</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">set_representative_structure</span><span class="p">(</span><span class="n">seq_outdir</span><span class="o">=</span><span class="n">seq_outdir</span><span class="p">,</span>
                                                               <span class="n">struct_outdir</span><span class="o">=</span><span class="n">struct_outdir</span><span class="p">,</span>
                                                               <span class="n">pdb_file_type</span><span class="o">=</span><span class="n">pdb_file_type</span><span class="p">,</span>
                                                               <span class="n">engine</span><span class="o">=</span><span class="n">engine</span><span class="p">,</span>
                                                               <span class="n">seq_ident_cutoff</span><span class="o">=</span><span class="n">seq_ident_cutoff</span><span class="p">,</span>
                                                               <span class="n">always_use_homology</span><span class="o">=</span><span class="n">always_use_homology</span><span class="p">,</span>
                                                               <span class="n">allow_missing_on_termini</span><span class="o">=</span><span class="n">allow_missing_on_termini</span><span class="p">,</span>
                                                               <span class="n">allow_mutants</span><span class="o">=</span><span class="n">allow_mutants</span><span class="p">,</span>
                                                               <span class="n">allow_deletions</span><span class="o">=</span><span class="n">allow_deletions</span><span class="p">,</span>
                                                               <span class="n">allow_insertions</span><span class="o">=</span><span class="n">allow_insertions</span><span class="p">,</span>
                                                               <span class="n">allow_unresolved</span><span class="o">=</span><span class="n">allow_unresolved</span><span class="p">,</span>
                                                               <span class="n">force_rerun</span><span class="o">=</span><span class="n">force_rerun</span><span class="p">)</span>

        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">/</span><span class="si">{}</span><span class="s1">: number of genes with a representative structure&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes_with_a_representative_structure</span><span class="p">),</span>
                                                                                 <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)))</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;See the &quot;df_representative_structures&quot; attribute for a summary dataframe.&#39;</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">df_representative_structures</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">rep_struct_pre_df</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">df_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">,</span> <span class="s1">&#39;id&#39;</span><span class="p">,</span> <span class="s1">&#39;is_experimental&#39;</span><span class="p">,</span> <span class="s1">&#39;reference_seq&#39;</span><span class="p">,</span> <span class="s1">&#39;reference_seq_top_coverage&#39;</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes_with_a_representative_structure</span><span class="p">:</span>
            <span class="n">repdict</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_structure</span><span class="o">.</span><span class="n">get_dict</span><span class="p">(</span><span class="n">df_format</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">only_keys</span><span class="o">=</span><span class="n">df_cols</span><span class="p">)</span>
            <span class="n">repdict</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">id</span>
            <span class="n">rep_struct_pre_df</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">repdict</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">rep_struct_pre_df</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">df_cols</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;gene&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">ssbio</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">clean_df</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">missing_representative_structure</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span> <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes_with_a_representative_structure</span><span class="o">.</span><span class="n">has_id</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">id</span><span class="p">)]</span>

    <span class="k">def</span> <span class="nf">prep_itasser_models</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">itasser_installation</span><span class="p">,</span> <span class="n">itlib_folder</span><span class="p">,</span> <span class="n">runtype</span><span class="p">,</span> <span class="n">create_in_dir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                            <span class="n">execute_from_dir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">all_genes</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">print_exec</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Prepare to run I-TASSER homology modeling for genes without structures, or all genes.</span>

<span class="sd">        Args:</span>
<span class="sd">            itasser_installation (str): Path to I-TASSER folder, i.e. &#39;~/software/I-TASSER4.4&#39;</span>
<span class="sd">            itlib_folder (str): Path to ITLIB folder, i.e. &#39;~/software/ITLIB&#39;</span>
<span class="sd">            runtype: How you will be running I-TASSER - local, slurm, or torque</span>
<span class="sd">            create_in_dir (str): Local directory where folders will be created</span>
<span class="sd">            execute_from_dir (str): Optional path to execution directory - use this if you are copying the homology</span>
<span class="sd">                models to another location such as a supercomputer for running</span>
<span class="sd">            all_genes (bool): If all genes should be prepped, or only those without any mapped structures</span>
<span class="sd">            print_exec (bool): If the execution statement should be printed to run modelling</span>
<span class="sd">            **kwargs: TODO - extra options for SLURM or Torque execution</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># TODO: kwargs for slurm/torque options</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">create_in_dir</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Output directory must be specified&#39;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">homology_models_dir</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span><span class="p">,</span> <span class="s1">&#39;homology_models&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">homology_models_dir</span> <span class="o">=</span> <span class="n">create_in_dir</span>

        <span class="n">ssbio</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">make_dir</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">homology_models_dir</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">execute_from_dir</span><span class="p">:</span>
            <span class="n">execute_from_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">homology_models_dir</span>

        <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes_with_a_representative_sequence</span><span class="p">:</span>
            <span class="n">repseq</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_sequence</span>

            <span class="c1"># TODO: refactor to Protein class</span>
            <span class="c1"># Skip homology modeling if there is a structure set</span>
            <span class="n">repstruct</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_structure</span>
            <span class="k">if</span> <span class="n">repstruct</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">all_genes</span><span class="p">:</span>
                <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: representative structure set, skipping homology modeling&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">))</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="s1">&#39;walltime&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
                <span class="n">ITASSERPrep</span><span class="p">(</span><span class="n">ident</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">seq_str</span><span class="o">=</span><span class="n">repseq</span><span class="o">.</span><span class="n">seq_str</span><span class="p">,</span> <span class="n">root_dir</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">homology_models_dir</span><span class="p">,</span>
                            <span class="n">itasser_path</span><span class="o">=</span><span class="n">itasser_installation</span><span class="p">,</span> <span class="n">itlib_path</span><span class="o">=</span><span class="n">itlib_folder</span><span class="p">,</span>
                            <span class="n">runtype</span><span class="o">=</span><span class="n">runtype</span><span class="p">,</span> <span class="n">print_exec</span><span class="o">=</span><span class="n">print_exec</span><span class="p">,</span> <span class="n">execute_dir</span><span class="o">=</span><span class="n">execute_from_dir</span><span class="p">,</span>
                            <span class="n">walltime</span><span class="o">=</span><span class="n">kwargs</span><span class="p">[</span><span class="s1">&#39;walltime&#39;</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">ITASSERPrep</span><span class="p">(</span><span class="n">ident</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">seq_str</span><span class="o">=</span><span class="n">repseq</span><span class="o">.</span><span class="n">seq_str</span><span class="p">,</span> <span class="n">root_dir</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">homology_models_dir</span><span class="p">,</span>
                            <span class="n">itasser_path</span><span class="o">=</span><span class="n">itasser_installation</span><span class="p">,</span> <span class="n">itlib_path</span><span class="o">=</span><span class="n">itlib_folder</span><span class="p">,</span>
                            <span class="n">runtype</span><span class="o">=</span><span class="n">runtype</span><span class="p">,</span> <span class="n">print_exec</span><span class="o">=</span><span class="n">print_exec</span><span class="p">,</span> <span class="n">execute_dir</span><span class="o">=</span><span class="n">execute_from_dir</span><span class="p">)</span>
            <span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Prepared I-TASSER modeling folders for </span><span class="si">{}</span><span class="s1"> genes in folder </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">counter</span><span class="p">,</span>
                                                                                       <span class="bp">self</span><span class="o">.</span><span class="n">homology_models_dir</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">pdb_downloader_and_metadata</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pdb_file_type</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">force_rerun</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Download all structures which have been mapped to our genes.</span>

<span class="sd">        Obtains the PDB file and mmCIF header and creates a metadata table &quot;df_pdb_metadata&quot;.</span>

<span class="sd">        Args:</span>
<span class="sd">            outdir (str): Path to output directory of downloaded files, must be set if GEM-PRO directories</span>
<span class="sd">                were not created initially</span>
<span class="sd">            pdb_file_type (str): pdb, pdb.gz, mmcif, cif, cif.gz, xml.gz, mmtf, mmtf.gz - PDB structure file type that</span>
<span class="sd">                should be downloaded</span>
<span class="sd">            force_rerun (bool): If you want to overwrite any existing mappings and files</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">pdb_file_type</span><span class="p">:</span>
            <span class="n">pdb_file_type</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pdb_file_type</span>

        <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">):</span>
            <span class="n">pdbs</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">pdb_downloader_and_metadata</span><span class="p">(</span><span class="n">outdir</span><span class="o">=</span><span class="n">outdir</span><span class="p">,</span> <span class="n">pdb_file_type</span><span class="o">=</span><span class="n">pdb_file_type</span><span class="p">,</span> <span class="n">force_rerun</span><span class="o">=</span><span class="n">force_rerun</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">pdbs</span><span class="p">:</span>
                <span class="n">counter</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">pdbs</span><span class="p">)</span>

        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Updated PDB metadata dataframe. See the &quot;df_pdb_metadata&quot; attribute for a summary dataframe.&#39;</span><span class="p">)</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Saved </span><span class="si">{}</span><span class="s1"> structures total&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">counter</span><span class="p">))</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">df_pdb_metadata</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes_with_experimental_structures</span><span class="p">:</span>
            <span class="c1"># Get per protein DataFrame</span>
            <span class="n">protein_df</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">df_pdb_metadata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
            <span class="n">protein_df</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">id</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">protein_df</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">ssbio</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">clean_df</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;gene&#39;</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">get_dssp_annotations</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Run DSSP on all representative structures and store calculations.</span>

<span class="sd">        Stored in:</span>
<span class="sd">            g.protein.representative_structure.representative_chain.seq_record.letter_annotations[&#39;*-dssp&#39;]</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_structure</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_structure</span><span class="o">.</span><span class="n">get_dssp_annotations</span><span class="p">(</span><span class="n">outdir</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">structure_dir</span><span class="p">)</span>
                <span class="k">except</span> <span class="n">PDBException</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
                    <span class="n">log</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: Biopython error, issue matching sequences with </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_structure</span><span class="p">))</span>
                <span class="k">except</span> <span class="ne">TypeError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
                    <span class="n">log</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: Biopython error, DSSP SeqRecord length mismatch with </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_structure</span><span class="p">))</span>
                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
                    <span class="n">log</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: DSSP failed to run on </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_structure</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">get_msms_annotations</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Run MSMS on all representative structures and store calculations.</span>

<span class="sd">        Stored in:</span>
<span class="sd">            g.protein.representative_structure.representative_chain.seq_record.letter_annotations[&#39;*-msms&#39;]</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_structure</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_structure</span><span class="o">.</span><span class="n">get_residue_depths</span><span class="p">(</span><span class="n">outdir</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">structure_dir</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">TypeError</span><span class="p">:</span>
                    <span class="n">log</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: MSMS SeqRecord length mismatch with </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_structure</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">get_disulfide_bridges</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Run Biopython&#39;s disulfide bridge finder and store calculations.</span>

<span class="sd">        Stored in:</span>
<span class="sd">            g.protein.representative_structure.representative_chain.seq_record.annotations[&#39;SSBOND-biopython&#39;]</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_structure</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_structure</span><span class="o">.</span><span class="n">get_disulfide_bridges</span><span class="p">()</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="n">log</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: unable to run disulfide bridge finder on </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">g</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">representative_structure</span><span class="p">))</span>

    <span class="c1">### END STRUCTURE RELATED METHODS ###</span>
    <span class="c1">####################################################################################################################</span>

    <span class="c1"># def run_pipeline(self, sequence_mapping_engine=&#39;&#39;, structure_mapping_engine=&#39;&#39;, **kwargs):</span>
    <span class="c1">#     &quot;&quot;&quot;Run the entire GEM-PRO pipeline.</span>
    <span class="c1">#</span>
    <span class="c1">#     Options include:</span>
    <span class="c1">#     ...</span>
    <span class="c1">#</span>
    <span class="c1">#     Returns:</span>
    <span class="c1">#</span>
    <span class="c1">#     &quot;&quot;&quot;</span>
    <span class="c1">#     current_sequence_mapping_engines = [&#39;kegg&#39;, &#39;uniprot&#39;, &#39;all&#39;]</span>
    <span class="c1">#     if sequence_mapping_engine not in current_sequence_mapping_engines:</span>
    <span class="c1">#         raise ValueError(&#39;Sequence mapping engine not available&#39;)</span>
    <span class="c1">#</span>
    <span class="c1">#     if sequence_mapping_engine == &#39;kegg&#39; or sequence_mapping_engine == &#39;all&#39;:</span>
    <span class="c1">#         if &#39;kegg_organism_code&#39; not in kwargs:</span>
    <span class="c1">#             raise TypeError(&#39;kegg_organism_code needed&#39;)</span>
    <span class="c1">#</span>
    <span class="c1">#     if sequence_mapping_engine == &#39;uniprot&#39; or sequence_mapping_engine == &#39;all&#39;:</span>
    <span class="c1">#         if &#39;model_gene_source&#39; not in kwargs:</span>
    <span class="c1">#             raise TypeError(&#39;UniProt model_gene_source needed&#39;)</span>
    <span class="c1">#</span>
    <span class="c1">#     print(&#39;Running KEGG mapping...&#39;)</span>
    <span class="c1">#     # TODO: better way of passing optional arguments to these things - yaml? There are too many if statements to check for things</span>
    <span class="c1">#     if &#39;custom_gene_mapping&#39; in kwargs:</span>
    <span class="c1">#         self.kegg_mapping_and_metadata(kegg_organism_code=kwargs[&#39;kegg_organism_code&#39;], custom_gene_mapping=kwargs[&#39;custom_gene_mapping&#39;])</span>
    <span class="c1">#     else:</span>
    <span class="c1">#         self.kegg_mapping_and_metadata(kegg_organism_code=kwargs[&#39;kegg_organism_code&#39;])</span>
    <span class="c1">#</span>
    <span class="c1">#     print(&#39;Running UniProt mapping...&#39;)</span>
    <span class="c1">#     self.uniprot_mapping_and_metadata(model_gene_source=kwargs[&#39;model_gene_source&#39;])</span>
    <span class="c1">#</span>
    <span class="c1">#     print(&#39;Setting representative sequences...&#39;)</span>
    <span class="c1">#     self.set_representative_sequence()</span>
    <span class="c1">#</span>
    <span class="c1">#     current_structure_mapping_engines = [&#39;uniprot&#39;, &#39;itasser&#39;, &#39;all&#39;, &#39;none&#39;] # &#39;blast&#39;, &#39;manual_homology&#39;</span>
    <span class="c1">#     if structure_mapping_engine not in current_structure_mapping_engines:</span>
    <span class="c1">#         raise ValueError(&#39;Structure mapping engine not available&#39;)</span>
    <span class="c1">#</span>
    <span class="c1">#     if structure_mapping_engine == &#39;itasser&#39; or structure_mapping_engine == &#39;all&#39;:</span>
    <span class="c1">#         if &#39;homology_raw_dir&#39; not in kwargs:</span>
    <span class="c1">#             raise TypeError(&#39;Path to homology models needed as homology_raw_dir&#39;)</span>
    <span class="c1">#</span>
    <span class="c1">#     if not structure_mapping_engine == &#39;none&#39;:</span>
    <span class="c1">#         if &#39;always_use_homology&#39; not in kwargs:</span>
    <span class="c1">#             raise TypeError(&#39;Flag always_use_homology needed&#39;)</span>
    <span class="c1">#</span>
    <span class="c1">#     if structure_mapping_engine == &#39;uniprot&#39; or structure_mapping_engine == &#39;all&#39;:</span>
    <span class="c1">#         print(&#39;Mapping UniProt IDs to the PDB...&#39;)</span>
    <span class="c1">#         self.map_uniprot_to_pdb()</span>
    <span class="c1">#</span>
    <span class="c1">#     if structure_mapping_engine == &#39;itasser&#39; or structure_mapping_engine == &#39;all&#39;:</span>
    <span class="c1">#         print(&#39;Copying I-TASSER models...&#39;)</span>
    <span class="c1">#         self.get_itasser_models(homology_raw_dir=kwargs[&#39;homology_raw_dir&#39;],</span>
    <span class="c1">#                                 custom_itasser_name_mapping=kwargs[&#39;custom_itasser_name_mapping&#39;])</span>
    <span class="c1">#</span>
    <span class="c1">#     if not structure_mapping_engine == &#39;none&#39;:</span>
    <span class="c1">#         print(&#39;Setting representative structures...&#39;)</span>
    <span class="c1">#         self.set_representative_structure(always_use_homology=kwargs[&#39;always_use_homology&#39;])</span>

    <span class="k">def</span> <span class="nf">__json_encode__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">to_return</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="c1"># Don&#39;t save properties, methods in the JSON</span>
        <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="n">a</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="nb">dir</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">a</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;__&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">),</span> <span class="n">a</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span> <span class="nb">property</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">callable</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">a</span><span class="p">))]:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="ow">and</span> <span class="n">x</span> <span class="o">==</span> <span class="s1">&#39;genes&#39;</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">to_return</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">x</span><span class="p">:</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">)})</span>
        <span class="k">return</span> <span class="n">to_return</span>

    <span class="k">def</span> <span class="nf">__json_decode__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">attrs</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">attrs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="n">DictList</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">genes</span>


<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="k">pass</span>
    <span class="c1"># TODO: Run the GEM-PRO pipeline!</span>
</pre></div>

           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Nathan Mih.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'1.0.0a0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>