{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# README\n",
    "\n",
    "This notebook gives an example of how to run the GEM-PRO pipeline on a **list of gene IDs** in Python 2.\n",
    "\n",
    "### Installation\n",
    "\n",
    "See: https://github.com/nmih/ssbio/blob/master/README.md\n",
    "- If something isn't working, make sure to update the repository before you do anything (git pull)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Import the GEM-PRO class\n",
    "from ssbio.pipeline.gempro import GEMPRO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Create logger\n",
    "import logging\n",
    "logger = logging.getLogger()\n",
    "\n",
    "############# SET YOUR LOGGING LEVEL HERE #############\n",
    "# - CRITICAL\n",
    "#     - Only really important messages shown\n",
    "# - ERROR\n",
    "#     - Major errors\n",
    "# - WARNING\n",
    "#     - Warnings that don't affect running of the pipeline\n",
    "# - INFO\n",
    "#     - Info such as the number of structures mapped per gene\n",
    "# - DEBUG\n",
    "#     - Really detailed information that will print out a lot of stuff\n",
    "logger.setLevel(logging.INFO)\n",
    "#######################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initialization of the project\n",
    "\n",
    "Set these three things:\n",
    "\n",
    "- GEM_NAME\n",
    "    - Your project name\n",
    "- ROOT_DIR\n",
    "    - The directory where the GEM_NAME folder will be created\n",
    "- GENES\n",
    "    - Your list of gene IDs\n",
    "    \n",
    "A directory will be created in ROOT_DIR named after your GEM_NAME. The folders are:\n",
    "```\n",
    "    .\n",
    "    ├── data  # where dataframes are stored\n",
    "    ├── figures  # where figures are stored\n",
    "    ├── model  # where SBML and GEM-PRO models are stored\n",
    "    ├── notebooks  # location of any ipython notebooks for analyses\n",
    "    ├── sequences  # sequences are stored here, in gene specific folders\n",
    "    │   ├── <gene_id1>\n",
    "    │   │   └── sequence.fasta\n",
    "    │   ├── <gene_id2>\n",
    "    │   └── ...\n",
    "    └── structures\n",
    "        ├── by_gene  # structures are stored here, in gene specific folders\n",
    "        │   ├── <gene_id1>\n",
    "        │   │   └── 1abc.pdb\n",
    "        │   ├── <gene_id2>\n",
    "        │   └── ...\n",
    "        └── by_complex  # complexes for reactions are stored here (in progress)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "GEM_NAME = 'py2test'\n",
    "\n",
    "ROOT_DIR = '/home/nathan/projects_unsynced/'\n",
    "\n",
    "GENES = ['b0761','b0889','b0995','b1013','b1014','b1040','b1130','b1187','b1221','b1299','b1323']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:ssbio.pipeline.gempro:Number of genes: 11\n"
     ]
    }
   ],
   "source": [
    "# Create the GEM-PRO project\n",
    "my_gempro = GEMPRO(gem_name=GEM_NAME, root_dir=ROOT_DIR, genes_list=GENES)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mapping gene ID -> sequence\n",
    "\n",
    "**You only need to map gene IDs using one service.** However you can try both if some genes don't map in one service and map in another!\n",
    "\n",
    "### kegg_mapping_and_metadata\n",
    "- kegg_organism_code\n",
    "    - See the full list of organisms here: http://www.genome.jp/kegg/catalog/org_list.html\n",
    "    - E. coli MG1655 is \"eco\"\n",
    "\n",
    "### uniprot_mapping_and_metadata\n",
    "You can try mapping your genes using the actual service here:\n",
    "http://www.uniprot.org/uploadlists/\n",
    "\n",
    "- model_gene_source\n",
    "    - Here is a list of the gene IDs that can be mapped to UniProt IDs: http://www.uniprot.org/help/programmatic_access#id_mapping_examples\n",
    "    - E. coli b-numbers in this example are of the source \"ENSEMBLGENOME_ID\"\n",
    "\n",
    "### set_representative_sequence\n",
    "This function allows you to consolidate sources of ID mapping (currently KEGG and UniProt).\n",
    "\n",
    "-------------------\n",
    "\n",
    "## Saved information:\n",
    "\n",
    "### Gene annotations\n",
    "Check out the gene annotations saved directly into the gene. These are COBRApy Gene objects!\n",
    "    - my_gempro.genes.get_by_id('b1187').annotation\n",
    "\n",
    "### DataFrames\n",
    "Check out the dataframes to see a summary of results.\n",
    "    - my_gempro.df_uniprot_metadata\n",
    "    - my_gempro.df_kegg_metadata\n",
    "    - my_gempro.df_sequence_mapping  # Summary of the representative sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'sequence': {'kegg': {'kegg_id': None,\n",
       "   'metadata_file': None,\n",
       "   'pdbs': [],\n",
       "   'seq_file': None,\n",
       "   'seq_len': 0,\n",
       "   'uniprot_acc': None},\n",
       "  'representative': {'kegg_id': None,\n",
       "   'metadata_file': None,\n",
       "   'pdbs': [],\n",
       "   'seq_file': None,\n",
       "   'seq_len': 0,\n",
       "   'uniprot_acc': None},\n",
       "  'uniprot': {}},\n",
       " 'structure': {'homology': {},\n",
       "  'pdb': OrderedDict(),\n",
       "  'representative': {'clean_pdb_file': None,\n",
       "   'original_pdb_file': None,\n",
       "   'seq_coverage': 0,\n",
       "   'structure_id': None}}}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Looking at the Gene object - empty for now\n",
    "example_gene = my_gempro.genes[0]\n",
    "example_gene.annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:getUserAgent: Begin\n",
      "INFO:root:getUserAgent: user_agent: EBI-Sample-Client/ (services.pyc; Python 2.7.12; Linux) Python-requests/2.9.1\n",
      "INFO:root:getUserAgent: End\n",
      "100%|██████████| 11/11 [00:00<00:00, 2010.96it/s]\n",
      "INFO:ssbio.pipeline.gempro:Created UniProt metadata dataframe. See the \"df_uniprot_metadata\" attribute.\n"
     ]
    }
   ],
   "source": [
    "# UniProt mapping of gene_id -> uniprot_id\n",
    "my_gempro.uniprot_mapping_and_metadata(model_gene_source='ENSEMBLGENOME_ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 11/11 [00:00<00:00, 618.29it/s]\n",
      "INFO:ssbio.pipeline.gempro:Created KEGG metadata dataframe. See the \"df_kegg_metadata\" attribute.\n"
     ]
    }
   ],
   "source": [
    "# KEGG mapping of gene ids\n",
    "my_gempro.kegg_mapping_and_metadata(kegg_organism_code='eco')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:ssbio.pipeline.gempro:Created sequence mapping dataframe. See the \"df_sequence_mapping\" attribute.\n"
     ]
    }
   ],
   "source": [
    "# Consolidate mappings sources\n",
    "my_gempro.set_representative_sequence()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>uniprot_acc</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>seq_len</th>\n",
       "      <th>pdbs</th>\n",
       "      <th>seq_file</th>\n",
       "      <th>metadata_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b1323</td>\n",
       "      <td>P07604</td>\n",
       "      <td>eco:b1323</td>\n",
       "      <td>513</td>\n",
       "      <td>[2JHE]</td>\n",
       "      <td>eco-b1323.faa</td>\n",
       "      <td>eco-b1323.kegg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b0995</td>\n",
       "      <td>P38684</td>\n",
       "      <td>eco:b0995</td>\n",
       "      <td>230</td>\n",
       "      <td>[1ZGZ]</td>\n",
       "      <td>eco-b0995.faa</td>\n",
       "      <td>eco-b0995.kegg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b0761</td>\n",
       "      <td>P0A9G8</td>\n",
       "      <td>eco:b0761</td>\n",
       "      <td>262</td>\n",
       "      <td>[1B9M, 1H9S, 1B9N, 1O7L, 1H9R]</td>\n",
       "      <td>eco-b0761.faa</td>\n",
       "      <td>eco-b0761.kegg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b1221</td>\n",
       "      <td>P0AF28</td>\n",
       "      <td>eco:b1221</td>\n",
       "      <td>216</td>\n",
       "      <td>[1JE8, 1A04, 1ZG1, 1ZG5, 1RNL]</td>\n",
       "      <td>eco-b1221.faa</td>\n",
       "      <td>eco-b1221.kegg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b1299</td>\n",
       "      <td>P0A9U6</td>\n",
       "      <td>eco:b1299</td>\n",
       "      <td>185</td>\n",
       "      <td>NaN</td>\n",
       "      <td>eco-b1299.faa</td>\n",
       "      <td>eco-b1299.kegg</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene uniprot_acc    kegg_id  seq_len                            pdbs  \\\n",
       "0  b1323      P07604  eco:b1323      513                          [2JHE]   \n",
       "1  b0995      P38684  eco:b0995      230                          [1ZGZ]   \n",
       "2  b0761      P0A9G8  eco:b0761      262  [1B9M, 1H9S, 1B9N, 1O7L, 1H9R]   \n",
       "3  b1221      P0AF28  eco:b1221      216  [1JE8, 1A04, 1ZG1, 1ZG5, 1RNL]   \n",
       "4  b1299      P0A9U6  eco:b1299      185                             NaN   \n",
       "\n",
       "        seq_file   metadata_file  \n",
       "0  eco-b1323.faa  eco-b1323.kegg  \n",
       "1  eco-b0995.faa  eco-b0995.kegg  \n",
       "2  eco-b0761.faa  eco-b0761.kegg  \n",
       "3  eco-b1221.faa  eco-b1221.kegg  \n",
       "4  eco-b1299.faa  eco-b1299.kegg  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Looking at the KEGG dataframe\n",
    "my_gempro.df_kegg_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'kegg': {'kegg_id': 'eco:b1187',\n",
       "  'metadata_file': 'eco-b1187.kegg',\n",
       "  'pdbs': ['1HW1', '1H9T', '1HW2', '1H9G', '1E2X'],\n",
       "  'seq_file': 'eco-b1187.faa',\n",
       "  'seq_len': 239,\n",
       "  'uniprot_acc': 'P0A8V6'},\n",
       " 'representative': {'kegg_id': ['ecj:JW1176', 'eco:b1187'],\n",
       "  'metadata_file': 'P0A8V6.txt',\n",
       "  'pdbs': ['1H9T', '1HW2', '1H9G', '1E2X', '1HW1'],\n",
       "  'seq_file': 'P0A8V6.fasta',\n",
       "  'seq_len': 239,\n",
       "  'uniprot_acc': 'P0A8V6'},\n",
       " 'uniprot': {'P0A8V6': {'description': ['Fatty acid metabolism regulator protein {ECO:0000255|HAMAP-Rule:MF_00696}'],\n",
       "   'entry_version': '2016-10-05',\n",
       "   'gene': 'b1187',\n",
       "   'gene_name': 'fadR',\n",
       "   'kegg_id': ['ecj:JW1176', 'eco:b1187'],\n",
       "   'metadata_file': 'P0A8V6.txt',\n",
       "   'pdbs': ['1H9T', '1HW2', '1H9G', '1E2X', '1HW1'],\n",
       "   'pfam': ['PF07840', 'PF00392'],\n",
       "   'refseq': ['NP_415705.1', 'NC_000913.3', 'WP_000234823.1', 'NZ_LN832404.1'],\n",
       "   'reviewed': True,\n",
       "   'seq_file': 'P0A8V6.fasta',\n",
       "   'seq_len': 239,\n",
       "   'seq_version': '2007-01-23',\n",
       "   'uniprot_acc': 'P0A8V6'}}}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Looking at information saved per gene\n",
    "my_gempro.genes.get_by_id('b1187').annotation['sequence']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mapping sequence -> structure\n",
    "\n",
    "There are two ways to map sequence to structure:\n",
    "1. Simply use the UniProt ID and their automatic mappings to the PDB\n",
    "2. BLAST the sequence to the PDB\n",
    "\n",
    "### map_uniprot_to_pdb\n",
    "This uses a service from the PDB to return a rank ordered list of PDBs that match a UniProt ID. \n",
    "- seq_ident_cutoff (from 0 to 1)\n",
    "    - Provide the seq_ident_cutoff as a percentage to filter for structures with only a percent identity above the cutoff.\n",
    "    - **Warning:** if you set the seq_ident_cutoff too high you risk filtering out PDBs that do match the sequence, but are just missing large portions of it.\n",
    "\n",
    "### blast_seqs_to_pdb\n",
    "This will BLAST the representative sequence against the entire PDB, and return significant hits. This will however return hits in other organisms, which may not be ideal.\n",
    "- all_genes\n",
    "    - Set to True if you want all genes and their sequences BLASTed\n",
    "    - Set to False if you only want to BLAST sequences that did not have any PDBs from the function map_uniprot_to_pdb\n",
    "- seq_ident_cutoff\n",
    "    - Same as above\n",
    "- evalue\n",
    "    - Significance of BLAST results\n",
    "\n",
    "-------------------\n",
    "\n",
    "## Saved information:\n",
    "\n",
    "### Gene annotations\n",
    "    - my_gempro.genes.get_by_id('b1187').annotation['structure']['pdb']\n",
    "\n",
    "### DataFrames\n",
    "    - my_gempro.df_pdb_ranking\n",
    "    - my_gempro.df_pdb_blast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 11/11 [00:03<00:00,  2.91it/s]\n",
      "INFO:ssbio.pipeline.gempro:Completed UniProt -> best PDB mapping. See the \"df_pdb_ranking\" attribute.\n"
     ]
    }
   ],
   "source": [
    "# Mapping using the PDBe best_structures service\n",
    "my_gempro.map_uniprot_to_pdb(seq_ident_cutoff=.3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/11 [00:00<?, ?it/s]INFO:ssbio.pipeline.gempro:b1013: Adding 2 PDBs from BLAST results.\n",
      "100%|██████████| 11/11 [00:00<00:00, 151.57it/s]\n",
      "INFO:ssbio.pipeline.gempro:Completed sequence --> PDB BLAST. See the \"df_pdb_blast\" attribute.\n"
     ]
    }
   ],
   "source": [
    "# Mapping using BLAST\n",
    "my_gempro.blast_seqs_to_pdb(all_genes=True, seq_ident_cutoff=.9, evalue=0.00001)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>uniprot_acc</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>pdb_chain_id</th>\n",
       "      <th>experimental_method</th>\n",
       "      <th>resolution</th>\n",
       "      <th>seq_coverage</th>\n",
       "      <th>release_date</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>pdb_start</th>\n",
       "      <th>pdb_end</th>\n",
       "      <th>unp_start</th>\n",
       "      <th>unp_end</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b1323</td>\n",
       "      <td>P07604</td>\n",
       "      <td>2jhe</td>\n",
       "      <td>A</td>\n",
       "      <td>X-ray diffraction</td>\n",
       "      <td>2.3</td>\n",
       "      <td>0.37</td>\n",
       "      <td>2008-06-24</td>\n",
       "      <td>83333</td>\n",
       "      <td>1</td>\n",
       "      <td>190</td>\n",
       "      <td>1</td>\n",
       "      <td>190</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b1323</td>\n",
       "      <td>P07604</td>\n",
       "      <td>2jhe</td>\n",
       "      <td>B</td>\n",
       "      <td>X-ray diffraction</td>\n",
       "      <td>2.3</td>\n",
       "      <td>0.37</td>\n",
       "      <td>2008-06-24</td>\n",
       "      <td>83333</td>\n",
       "      <td>1</td>\n",
       "      <td>190</td>\n",
       "      <td>1</td>\n",
       "      <td>190</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b1323</td>\n",
       "      <td>P07604</td>\n",
       "      <td>2jhe</td>\n",
       "      <td>C</td>\n",
       "      <td>X-ray diffraction</td>\n",
       "      <td>2.3</td>\n",
       "      <td>0.37</td>\n",
       "      <td>2008-06-24</td>\n",
       "      <td>83333</td>\n",
       "      <td>1</td>\n",
       "      <td>190</td>\n",
       "      <td>1</td>\n",
       "      <td>190</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b1323</td>\n",
       "      <td>P07604</td>\n",
       "      <td>2jhe</td>\n",
       "      <td>D</td>\n",
       "      <td>X-ray diffraction</td>\n",
       "      <td>2.3</td>\n",
       "      <td>0.37</td>\n",
       "      <td>2008-06-24</td>\n",
       "      <td>83333</td>\n",
       "      <td>1</td>\n",
       "      <td>190</td>\n",
       "      <td>1</td>\n",
       "      <td>190</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b0995</td>\n",
       "      <td>P38684</td>\n",
       "      <td>1zgz</td>\n",
       "      <td>A</td>\n",
       "      <td>X-ray diffraction</td>\n",
       "      <td>1.8</td>\n",
       "      <td>0.53</td>\n",
       "      <td>2005-12-13</td>\n",
       "      <td>562</td>\n",
       "      <td>1</td>\n",
       "      <td>122</td>\n",
       "      <td>1</td>\n",
       "      <td>122</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene uniprot_acc pdb_id pdb_chain_id experimental_method  resolution  \\\n",
       "0  b1323      P07604   2jhe            A   X-ray diffraction         2.3   \n",
       "1  b1323      P07604   2jhe            B   X-ray diffraction         2.3   \n",
       "2  b1323      P07604   2jhe            C   X-ray diffraction         2.3   \n",
       "3  b1323      P07604   2jhe            D   X-ray diffraction         2.3   \n",
       "4  b0995      P38684   1zgz            A   X-ray diffraction         1.8   \n",
       "\n",
       "   seq_coverage release_date  taxonomy_id  pdb_start  pdb_end  unp_start  \\\n",
       "0          0.37   2008-06-24        83333          1      190          1   \n",
       "1          0.37   2008-06-24        83333          1      190          1   \n",
       "2          0.37   2008-06-24        83333          1      190          1   \n",
       "3          0.37   2008-06-24        83333          1      190          1   \n",
       "4          0.53   2005-12-13          562          1      122          1   \n",
       "\n",
       "   unp_end  rank  \n",
       "0      190     1  \n",
       "1      190     2  \n",
       "2      190     3  \n",
       "3      190     4  \n",
       "4      122     1  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Looking at the PDB ranking dataframe\n",
    "my_gempro.df_pdb_ranking.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>pdb_chain_id</th>\n",
       "      <th>resolution</th>\n",
       "      <th>release_date</th>\n",
       "      <th>blast_score</th>\n",
       "      <th>blast_evalue</th>\n",
       "      <th>seq_coverage</th>\n",
       "      <th>seq_similar</th>\n",
       "      <th>seq_num_coverage</th>\n",
       "      <th>seq_num_similar</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b0761</td>\n",
       "      <td>1b9n</td>\n",
       "      <td>A</td>\n",
       "      <td>2.09</td>\n",
       "      <td>2000-03-15</td>\n",
       "      <td>1091.0</td>\n",
       "      <td>4.797990e-119</td>\n",
       "      <td>0.931298</td>\n",
       "      <td>0.931298</td>\n",
       "      <td>244</td>\n",
       "      <td>244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b0761</td>\n",
       "      <td>1b9n</td>\n",
       "      <td>B</td>\n",
       "      <td>2.09</td>\n",
       "      <td>2000-03-15</td>\n",
       "      <td>1091.0</td>\n",
       "      <td>4.797990e-119</td>\n",
       "      <td>0.931298</td>\n",
       "      <td>0.931298</td>\n",
       "      <td>244</td>\n",
       "      <td>244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b0761</td>\n",
       "      <td>1b9m</td>\n",
       "      <td>A</td>\n",
       "      <td>1.75</td>\n",
       "      <td>2000-03-15</td>\n",
       "      <td>1091.0</td>\n",
       "      <td>4.797990e-119</td>\n",
       "      <td>0.931298</td>\n",
       "      <td>0.931298</td>\n",
       "      <td>244</td>\n",
       "      <td>244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b0761</td>\n",
       "      <td>1b9m</td>\n",
       "      <td>B</td>\n",
       "      <td>1.75</td>\n",
       "      <td>2000-03-15</td>\n",
       "      <td>1091.0</td>\n",
       "      <td>4.797990e-119</td>\n",
       "      <td>0.931298</td>\n",
       "      <td>0.931298</td>\n",
       "      <td>244</td>\n",
       "      <td>244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b0761</td>\n",
       "      <td>1o7l</td>\n",
       "      <td>A</td>\n",
       "      <td>2.75</td>\n",
       "      <td>2003-02-20</td>\n",
       "      <td>1089.0</td>\n",
       "      <td>9.510390e-119</td>\n",
       "      <td>0.931298</td>\n",
       "      <td>0.931298</td>\n",
       "      <td>244</td>\n",
       "      <td>244</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene pdb_id pdb_chain_id  resolution release_date  blast_score  \\\n",
       "0  b0761   1b9n            A        2.09   2000-03-15       1091.0   \n",
       "1  b0761   1b9n            B        2.09   2000-03-15       1091.0   \n",
       "2  b0761   1b9m            A        1.75   2000-03-15       1091.0   \n",
       "3  b0761   1b9m            B        1.75   2000-03-15       1091.0   \n",
       "4  b0761   1o7l            A        2.75   2003-02-20       1089.0   \n",
       "\n",
       "    blast_evalue  seq_coverage  seq_similar  seq_num_coverage  seq_num_similar  \n",
       "0  4.797990e-119      0.931298     0.931298               244              244  \n",
       "1  4.797990e-119      0.931298     0.931298               244              244  \n",
       "2  4.797990e-119      0.931298     0.931298               244              244  \n",
       "3  4.797990e-119      0.931298     0.931298               244              244  \n",
       "4  9.510390e-119      0.931298     0.931298               244              244  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Looking at the BLAST results\n",
    "my_gempro.df_pdb_blast.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "OrderedDict([(('1hw1', 'A'),\n",
       "              {'experimental_method': u'X-ray diffraction',\n",
       "               'pdb_chain_id': 'A',\n",
       "               'pdb_end': 239,\n",
       "               'pdb_id': '1hw1',\n",
       "               'pdb_start': 1,\n",
       "               'rank': 1,\n",
       "               'release_date': '2001-01-24',\n",
       "               'resolution': 1.5,\n",
       "               'seq_coverage': 1,\n",
       "               'taxonomy_id': 562,\n",
       "               'uniprot_acc': 'P0A8V6',\n",
       "               'unp_end': 239,\n",
       "               'unp_start': 1}),\n",
       "             (('1hw1', 'B'),\n",
       "              {'experimental_method': u'X-ray diffraction',\n",
       "               'pdb_chain_id': 'B',\n",
       "               'pdb_end': 239,\n",
       "               'pdb_id': '1hw1',\n",
       "               'pdb_start': 1,\n",
       "               'rank': 2,\n",
       "               'release_date': '2001-01-24',\n",
       "               'resolution': 1.5,\n",
       "               'seq_coverage': 1,\n",
       "               'taxonomy_id': 562,\n",
       "               'uniprot_acc': 'P0A8V6',\n",
       "               'unp_end': 239,\n",
       "               'unp_start': 1}),\n",
       "             (('1e2x', 'A'),\n",
       "              {'experimental_method': u'X-ray diffraction',\n",
       "               'pdb_chain_id': 'A',\n",
       "               'pdb_end': 243,\n",
       "               'pdb_id': '1e2x',\n",
       "               'pdb_start': 5,\n",
       "               'rank': 3,\n",
       "               'release_date': '2000-12-03',\n",
       "               'resolution': 2,\n",
       "               'seq_coverage': 1,\n",
       "               'taxonomy_id': 562,\n",
       "               'uniprot_acc': 'P0A8V6',\n",
       "               'unp_end': 239,\n",
       "               'unp_start': 1}),\n",
       "             (('1h9g', 'A'),\n",
       "              {'experimental_method': u'X-ray diffraction',\n",
       "               'pdb_chain_id': 'A',\n",
       "               'pdb_end': 243,\n",
       "               'pdb_id': '1h9g',\n",
       "               'pdb_start': 5,\n",
       "               'rank': 4,\n",
       "               'release_date': '2001-03-09',\n",
       "               'resolution': 2.1,\n",
       "               'seq_coverage': 1,\n",
       "               'taxonomy_id': 562,\n",
       "               'uniprot_acc': 'P0A8V6',\n",
       "               'unp_end': 239,\n",
       "               'unp_start': 1}),\n",
       "             (('1h9t', 'A'),\n",
       "              {'experimental_method': u'X-ray diffraction',\n",
       "               'pdb_chain_id': 'A',\n",
       "               'pdb_end': 243,\n",
       "               'pdb_id': '1h9t',\n",
       "               'pdb_start': 5,\n",
       "               'rank': 5,\n",
       "               'release_date': '2001-04-04',\n",
       "               'resolution': 3.25,\n",
       "               'seq_coverage': 1,\n",
       "               'taxonomy_id': 562,\n",
       "               'uniprot_acc': 'P0A8V6',\n",
       "               'unp_end': 239,\n",
       "               'unp_start': 1}),\n",
       "             (('1h9t', 'B'),\n",
       "              {'experimental_method': u'X-ray diffraction',\n",
       "               'pdb_chain_id': 'B',\n",
       "               'pdb_end': 243,\n",
       "               'pdb_id': '1h9t',\n",
       "               'pdb_start': 5,\n",
       "               'rank': 6,\n",
       "               'release_date': '2001-04-04',\n",
       "               'resolution': 3.25,\n",
       "               'seq_coverage': 1,\n",
       "               'taxonomy_id': 562,\n",
       "               'uniprot_acc': 'P0A8V6',\n",
       "               'unp_end': 239,\n",
       "               'unp_start': 1}),\n",
       "             (('1hw2', 'A'),\n",
       "              {'experimental_method': u'X-ray diffraction',\n",
       "               'pdb_chain_id': 'A',\n",
       "               'pdb_end': 239,\n",
       "               'pdb_id': '1hw2',\n",
       "               'pdb_start': 1,\n",
       "               'rank': 7,\n",
       "               'release_date': '2001-01-22',\n",
       "               'resolution': 3.25,\n",
       "               'seq_coverage': 1,\n",
       "               'taxonomy_id': 562,\n",
       "               'uniprot_acc': 'P0A8V6',\n",
       "               'unp_end': 239,\n",
       "               'unp_start': 1}),\n",
       "             (('1hw2', 'B'),\n",
       "              {'experimental_method': u'X-ray diffraction',\n",
       "               'pdb_chain_id': 'B',\n",
       "               'pdb_end': 239,\n",
       "               'pdb_id': '1hw2',\n",
       "               'pdb_start': 1,\n",
       "               'rank': 8,\n",
       "               'release_date': '2001-01-22',\n",
       "               'resolution': 3.25,\n",
       "               'seq_coverage': 1,\n",
       "               'taxonomy_id': 562,\n",
       "               'uniprot_acc': 'P0A8V6',\n",
       "               'unp_end': 239,\n",
       "               'unp_start': 1})])"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Looking at the information saved per gene\n",
    "my_gempro.genes.get_by_id('b1187').annotation['structure']['pdb']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ranking and downloading structures\n",
    "\n",
    "### set_representative_structure\n",
    "Rank available structures, run QC/QA, download and clean the final structure\n",
    "### pdb_downloader_and_metadata\n",
    "Download all structures per gene. This also adds some additional metadata to the annotation for each PDB.\n",
    "    \n",
    "-------------------\n",
    "\n",
    "## Saved information:\n",
    "\n",
    "### Gene annotations\n",
    "    - my_gempro.genes.get_by_id('b1187').annotation['structure']['representative']\n",
    "\n",
    "### DataFrames\n",
    "    - my_gempro.df_pdb_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 11/11 [00:08<00:00,  1.15s/it]\n",
      "INFO:ssbio.pipeline.gempro:Created PDB metadata dataframe.\n"
     ]
    }
   ],
   "source": [
    "# Download all mapped PDBs and gather the metadata\n",
    "my_gempro.pdb_downloader_and_metadata()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>pdb_chain_id</th>\n",
       "      <th>taxonomy_name</th>\n",
       "      <th>experimental_method</th>\n",
       "      <th>resolution</th>\n",
       "      <th>seq_coverage</th>\n",
       "      <th>chemicals</th>\n",
       "      <th>rank</th>\n",
       "      <th>release_date</th>\n",
       "      <th>pdb_file</th>\n",
       "      <th>mmcif_header</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b1323</td>\n",
       "      <td>2jhe</td>\n",
       "      <td>A</td>\n",
       "      <td>ESCHERICHIA COLI</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>2.3</td>\n",
       "      <td>0.37</td>\n",
       "      <td>AE3;PG4;SO4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2008-06-24</td>\n",
       "      <td>2jhe.pdb</td>\n",
       "      <td>2jhe.header.cif</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b1323</td>\n",
       "      <td>2jhe</td>\n",
       "      <td>B</td>\n",
       "      <td>ESCHERICHIA COLI</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>2.3</td>\n",
       "      <td>0.37</td>\n",
       "      <td>AE3;PG4;SO4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2008-06-24</td>\n",
       "      <td>2jhe.pdb</td>\n",
       "      <td>2jhe.header.cif</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b1323</td>\n",
       "      <td>2jhe</td>\n",
       "      <td>C</td>\n",
       "      <td>ESCHERICHIA COLI</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>2.3</td>\n",
       "      <td>0.37</td>\n",
       "      <td>AE3;PG4;SO4</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2008-06-24</td>\n",
       "      <td>2jhe.pdb</td>\n",
       "      <td>2jhe.header.cif</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b1323</td>\n",
       "      <td>2jhe</td>\n",
       "      <td>D</td>\n",
       "      <td>ESCHERICHIA COLI</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>2.3</td>\n",
       "      <td>0.37</td>\n",
       "      <td>AE3;PG4;SO4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2008-06-24</td>\n",
       "      <td>2jhe.pdb</td>\n",
       "      <td>2jhe.header.cif</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b0995</td>\n",
       "      <td>1zgz</td>\n",
       "      <td>A</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>1.8</td>\n",
       "      <td>0.53</td>\n",
       "      <td>SO4;GOL</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2005-12-13</td>\n",
       "      <td>1zgz.pdb</td>\n",
       "      <td>1zgz.header.cif</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene pdb_id pdb_chain_id     taxonomy_name experimental_method  \\\n",
       "0  b1323   2jhe            A  ESCHERICHIA COLI   X-RAY DIFFRACTION   \n",
       "1  b1323   2jhe            B  ESCHERICHIA COLI   X-RAY DIFFRACTION   \n",
       "2  b1323   2jhe            C  ESCHERICHIA COLI   X-RAY DIFFRACTION   \n",
       "3  b1323   2jhe            D  ESCHERICHIA COLI   X-RAY DIFFRACTION   \n",
       "4  b0995   1zgz            A  Escherichia coli   X-RAY DIFFRACTION   \n",
       "\n",
       "   resolution  seq_coverage    chemicals  rank release_date  pdb_file  \\\n",
       "0         2.3          0.37  AE3;PG4;SO4   1.0   2008-06-24  2jhe.pdb   \n",
       "1         2.3          0.37  AE3;PG4;SO4   2.0   2008-06-24  2jhe.pdb   \n",
       "2         2.3          0.37  AE3;PG4;SO4   3.0   2008-06-24  2jhe.pdb   \n",
       "3         2.3          0.37  AE3;PG4;SO4   4.0   2008-06-24  2jhe.pdb   \n",
       "4         1.8          0.53      SO4;GOL   1.0   2005-12-13  1zgz.pdb   \n",
       "\n",
       "      mmcif_header  \n",
       "0  2jhe.header.cif  \n",
       "1  2jhe.header.cif  \n",
       "2  2jhe.header.cif  \n",
       "3  2jhe.header.cif  \n",
       "4  1zgz.header.cif  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Look at the summary of structures\n",
    "my_gempro.df_pdb_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "OrderedDict([(('1hw1', 'A'),\n",
       "              {'chemicals': ['SO4', 'ZN'],\n",
       "               'experimental_method': 'X-RAY DIFFRACTION',\n",
       "               'mmcif_header': '1hw1.header.cif',\n",
       "               'pdb_chain_id': 'A',\n",
       "               'pdb_end': 239,\n",
       "               'pdb_file': '1hw1.pdb',\n",
       "               'pdb_id': '1hw1',\n",
       "               'pdb_start': 1,\n",
       "               'rank': 1,\n",
       "               'release_date': '2001-01-24',\n",
       "               'resolution': 1.5,\n",
       "               'seq_coverage': 1,\n",
       "               'taxonomy_id': 562,\n",
       "               'taxonomy_name': 'Escherichia coli',\n",
       "               'uniprot_acc': 'P0A8V6',\n",
       "               'unp_end': 239,\n",
       "               'unp_start': 1}),\n",
       "             (('1hw1', 'B'),\n",
       "              {'chemicals': ['SO4', 'ZN'],\n",
       "               'experimental_method': 'X-RAY DIFFRACTION',\n",
       "               'mmcif_header': '1hw1.header.cif',\n",
       "               'pdb_chain_id': 'B',\n",
       "               'pdb_end': 239,\n",
       "               'pdb_file': '1hw1.pdb',\n",
       "               'pdb_id': '1hw1',\n",
       "               'pdb_start': 1,\n",
       "               'rank': 2,\n",
       "               'release_date': '2001-01-24',\n",
       "               'resolution': 1.5,\n",
       "               'seq_coverage': 1,\n",
       "               'taxonomy_id': 562,\n",
       "               'taxonomy_name': 'Escherichia coli',\n",
       "               'uniprot_acc': 'P0A8V6',\n",
       "               'unp_end': 239,\n",
       "               'unp_start': 1}),\n",
       "             (('1e2x', 'A'),\n",
       "              {'chemicals': ['SO4'],\n",
       "               'experimental_method': 'X-RAY DIFFRACTION',\n",
       "               'mmcif_header': '1e2x.header.cif',\n",
       "               'pdb_chain_id': 'A',\n",
       "               'pdb_end': 243,\n",
       "               'pdb_file': '1e2x.pdb',\n",
       "               'pdb_id': '1e2x',\n",
       "               'pdb_start': 5,\n",
       "               'rank': 3,\n",
       "               'release_date': '2000-12-03',\n",
       "               'resolution': 2.0,\n",
       "               'seq_coverage': 1,\n",
       "               'taxonomy_id': 562,\n",
       "               'taxonomy_name': 'ESCHERICHIA COLI',\n",
       "               'uniprot_acc': 'P0A8V6',\n",
       "               'unp_end': 239,\n",
       "               'unp_start': 1}),\n",
       "             (('1h9g', 'A'),\n",
       "              {'chemicals': ['COA', 'MYR'],\n",
       "               'experimental_method': 'X-RAY DIFFRACTION',\n",
       "               'mmcif_header': '1h9g.header.cif',\n",
       "               'pdb_chain_id': 'A',\n",
       "               'pdb_end': 243,\n",
       "               'pdb_file': '1h9g.pdb',\n",
       "               'pdb_id': '1h9g',\n",
       "               'pdb_start': 5,\n",
       "               'rank': 4,\n",
       "               'release_date': '2001-03-09',\n",
       "               'resolution': 2.1,\n",
       "               'seq_coverage': 1,\n",
       "               'taxonomy_id': 562,\n",
       "               'taxonomy_name': 'ESCHERICHIA COLI',\n",
       "               'uniprot_acc': 'P0A8V6',\n",
       "               'unp_end': 239,\n",
       "               'unp_start': 1}),\n",
       "             (('1h9t', 'A'),\n",
       "              {'chemicals': ['DC', 'DA', 'DT', 'DG', 'AU', 'CL'],\n",
       "               'experimental_method': 'X-RAY DIFFRACTION',\n",
       "               'mmcif_header': '1h9t.header.cif',\n",
       "               'pdb_chain_id': 'A',\n",
       "               'pdb_end': 243,\n",
       "               'pdb_file': '1h9t.pdb',\n",
       "               'pdb_id': '1h9t',\n",
       "               'pdb_start': 5,\n",
       "               'rank': 5,\n",
       "               'release_date': '2001-04-04',\n",
       "               'resolution': 3.25,\n",
       "               'seq_coverage': 1,\n",
       "               'taxonomy_id': 562,\n",
       "               'taxonomy_name': 'ESCHERICHIA COLI',\n",
       "               'uniprot_acc': 'P0A8V6',\n",
       "               'unp_end': 239,\n",
       "               'unp_start': 1}),\n",
       "             (('1h9t', 'B'),\n",
       "              {'chemicals': ['DC', 'DA', 'DT', 'DG', 'AU', 'CL'],\n",
       "               'experimental_method': 'X-RAY DIFFRACTION',\n",
       "               'mmcif_header': '1h9t.header.cif',\n",
       "               'pdb_chain_id': 'B',\n",
       "               'pdb_end': 243,\n",
       "               'pdb_file': '1h9t.pdb',\n",
       "               'pdb_id': '1h9t',\n",
       "               'pdb_start': 5,\n",
       "               'rank': 6,\n",
       "               'release_date': '2001-04-04',\n",
       "               'resolution': 3.25,\n",
       "               'seq_coverage': 1,\n",
       "               'taxonomy_id': 562,\n",
       "               'taxonomy_name': 'ESCHERICHIA COLI',\n",
       "               'uniprot_acc': 'P0A8V6',\n",
       "               'unp_end': 239,\n",
       "               'unp_start': 1}),\n",
       "             (('1hw2', 'A'),\n",
       "              {'chemicals': ['DC', 'DG', 'DA', 'DT', 'MG'],\n",
       "               'experimental_method': 'X-RAY DIFFRACTION',\n",
       "               'mmcif_header': '1hw2.header.cif',\n",
       "               'pdb_chain_id': 'A',\n",
       "               'pdb_end': 239,\n",
       "               'pdb_file': '1hw2.pdb',\n",
       "               'pdb_id': '1hw2',\n",
       "               'pdb_start': 1,\n",
       "               'rank': 7,\n",
       "               'release_date': '2001-01-22',\n",
       "               'resolution': 3.25,\n",
       "               'seq_coverage': 1,\n",
       "               'taxonomy_id': 562,\n",
       "               'taxonomy_name': 'Escherichia coli',\n",
       "               'uniprot_acc': 'P0A8V6',\n",
       "               'unp_end': 239,\n",
       "               'unp_start': 1}),\n",
       "             (('1hw2', 'B'),\n",
       "              {'chemicals': ['DC', 'DG', 'DA', 'DT', 'MG'],\n",
       "               'experimental_method': 'X-RAY DIFFRACTION',\n",
       "               'mmcif_header': '1hw2.header.cif',\n",
       "               'pdb_chain_id': 'B',\n",
       "               'pdb_end': 239,\n",
       "               'pdb_file': '1hw2.pdb',\n",
       "               'pdb_id': '1hw2',\n",
       "               'pdb_start': 1,\n",
       "               'rank': 8,\n",
       "               'release_date': '2001-01-22',\n",
       "               'resolution': 3.25,\n",
       "               'seq_coverage': 1,\n",
       "               'taxonomy_id': 562,\n",
       "               'taxonomy_name': 'Escherichia coli',\n",
       "               'uniprot_acc': 'P0A8V6',\n",
       "               'unp_end': 239,\n",
       "               'unp_start': 1})])"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Looking at information saved per gene\n",
    "my_gempro.genes.get_by_id('b1187').annotation['structure']['pdb']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'sequence': {'kegg': {'kegg_id': 'eco:b1187',\n",
       "   'metadata_file': 'eco-b1187.kegg',\n",
       "   'pdbs': ['1HW1', '1H9T', '1HW2', '1H9G', '1E2X'],\n",
       "   'seq_file': 'eco-b1187.faa',\n",
       "   'seq_len': 239,\n",
       "   'uniprot_acc': 'P0A8V6'},\n",
       "  'representative': {'kegg_id': ['ecj:JW1176', 'eco:b1187'],\n",
       "   'metadata_file': 'P0A8V6.txt',\n",
       "   'pdbs': ['1H9T', '1HW2', '1H9G', '1E2X', '1HW1'],\n",
       "   'seq_file': 'P0A8V6.fasta',\n",
       "   'seq_len': 239,\n",
       "   'uniprot_acc': 'P0A8V6'},\n",
       "  'uniprot': {'P0A8V6': {'description': ['Fatty acid metabolism regulator protein {ECO:0000255|HAMAP-Rule:MF_00696}'],\n",
       "    'entry_version': '2016-10-05',\n",
       "    'gene': 'b1187',\n",
       "    'gene_name': 'fadR',\n",
       "    'kegg_id': ['ecj:JW1176', 'eco:b1187'],\n",
       "    'metadata_file': 'P0A8V6.txt',\n",
       "    'pdbs': ['1H9T', '1HW2', '1H9G', '1E2X', '1HW1'],\n",
       "    'pfam': ['PF07840', 'PF00392'],\n",
       "    'refseq': ['NP_415705.1',\n",
       "     'NC_000913.3',\n",
       "     'WP_000234823.1',\n",
       "     'NZ_LN832404.1'],\n",
       "    'reviewed': True,\n",
       "    'seq_file': 'P0A8V6.fasta',\n",
       "    'seq_len': 239,\n",
       "    'seq_version': '2007-01-23',\n",
       "    'uniprot_acc': 'P0A8V6'}}},\n",
       " 'structure': {'homology': {},\n",
       "  'pdb': OrderedDict([(('1hw1', 'A'),\n",
       "                {'chemicals': ['SO4', 'ZN'],\n",
       "                 'experimental_method': 'X-RAY DIFFRACTION',\n",
       "                 'mmcif_header': '1hw1.header.cif',\n",
       "                 'pdb_chain_id': 'A',\n",
       "                 'pdb_end': 239,\n",
       "                 'pdb_file': '1hw1.pdb',\n",
       "                 'pdb_id': '1hw1',\n",
       "                 'pdb_start': 1,\n",
       "                 'rank': 1,\n",
       "                 'release_date': '2001-01-24',\n",
       "                 'resolution': 1.5,\n",
       "                 'seq_coverage': 1,\n",
       "                 'taxonomy_id': 562,\n",
       "                 'taxonomy_name': 'Escherichia coli',\n",
       "                 'uniprot_acc': 'P0A8V6',\n",
       "                 'unp_end': 239,\n",
       "                 'unp_start': 1}),\n",
       "               (('1hw1', 'B'),\n",
       "                {'chemicals': ['SO4', 'ZN'],\n",
       "                 'experimental_method': 'X-RAY DIFFRACTION',\n",
       "                 'mmcif_header': '1hw1.header.cif',\n",
       "                 'pdb_chain_id': 'B',\n",
       "                 'pdb_end': 239,\n",
       "                 'pdb_file': '1hw1.pdb',\n",
       "                 'pdb_id': '1hw1',\n",
       "                 'pdb_start': 1,\n",
       "                 'rank': 2,\n",
       "                 'release_date': '2001-01-24',\n",
       "                 'resolution': 1.5,\n",
       "                 'seq_coverage': 1,\n",
       "                 'taxonomy_id': 562,\n",
       "                 'taxonomy_name': 'Escherichia coli',\n",
       "                 'uniprot_acc': 'P0A8V6',\n",
       "                 'unp_end': 239,\n",
       "                 'unp_start': 1}),\n",
       "               (('1e2x', 'A'),\n",
       "                {'chemicals': ['SO4'],\n",
       "                 'experimental_method': 'X-RAY DIFFRACTION',\n",
       "                 'mmcif_header': '1e2x.header.cif',\n",
       "                 'pdb_chain_id': 'A',\n",
       "                 'pdb_end': 243,\n",
       "                 'pdb_file': '1e2x.pdb',\n",
       "                 'pdb_id': '1e2x',\n",
       "                 'pdb_start': 5,\n",
       "                 'rank': 3,\n",
       "                 'release_date': '2000-12-03',\n",
       "                 'resolution': 2.0,\n",
       "                 'seq_coverage': 1,\n",
       "                 'taxonomy_id': 562,\n",
       "                 'taxonomy_name': 'ESCHERICHIA COLI',\n",
       "                 'uniprot_acc': 'P0A8V6',\n",
       "                 'unp_end': 239,\n",
       "                 'unp_start': 1}),\n",
       "               (('1h9g', 'A'),\n",
       "                {'chemicals': ['COA', 'MYR'],\n",
       "                 'experimental_method': 'X-RAY DIFFRACTION',\n",
       "                 'mmcif_header': '1h9g.header.cif',\n",
       "                 'pdb_chain_id': 'A',\n",
       "                 'pdb_end': 243,\n",
       "                 'pdb_file': '1h9g.pdb',\n",
       "                 'pdb_id': '1h9g',\n",
       "                 'pdb_start': 5,\n",
       "                 'rank': 4,\n",
       "                 'release_date': '2001-03-09',\n",
       "                 'resolution': 2.1,\n",
       "                 'seq_coverage': 1,\n",
       "                 'taxonomy_id': 562,\n",
       "                 'taxonomy_name': 'ESCHERICHIA COLI',\n",
       "                 'uniprot_acc': 'P0A8V6',\n",
       "                 'unp_end': 239,\n",
       "                 'unp_start': 1}),\n",
       "               (('1h9t', 'A'),\n",
       "                {'chemicals': ['DC', 'DA', 'DT', 'DG', 'AU', 'CL'],\n",
       "                 'experimental_method': 'X-RAY DIFFRACTION',\n",
       "                 'mmcif_header': '1h9t.header.cif',\n",
       "                 'pdb_chain_id': 'A',\n",
       "                 'pdb_end': 243,\n",
       "                 'pdb_file': '1h9t.pdb',\n",
       "                 'pdb_id': '1h9t',\n",
       "                 'pdb_start': 5,\n",
       "                 'rank': 5,\n",
       "                 'release_date': '2001-04-04',\n",
       "                 'resolution': 3.25,\n",
       "                 'seq_coverage': 1,\n",
       "                 'taxonomy_id': 562,\n",
       "                 'taxonomy_name': 'ESCHERICHIA COLI',\n",
       "                 'uniprot_acc': 'P0A8V6',\n",
       "                 'unp_end': 239,\n",
       "                 'unp_start': 1}),\n",
       "               (('1h9t', 'B'),\n",
       "                {'chemicals': ['DC', 'DA', 'DT', 'DG', 'AU', 'CL'],\n",
       "                 'experimental_method': 'X-RAY DIFFRACTION',\n",
       "                 'mmcif_header': '1h9t.header.cif',\n",
       "                 'pdb_chain_id': 'B',\n",
       "                 'pdb_end': 243,\n",
       "                 'pdb_file': '1h9t.pdb',\n",
       "                 'pdb_id': '1h9t',\n",
       "                 'pdb_start': 5,\n",
       "                 'rank': 6,\n",
       "                 'release_date': '2001-04-04',\n",
       "                 'resolution': 3.25,\n",
       "                 'seq_coverage': 1,\n",
       "                 'taxonomy_id': 562,\n",
       "                 'taxonomy_name': 'ESCHERICHIA COLI',\n",
       "                 'uniprot_acc': 'P0A8V6',\n",
       "                 'unp_end': 239,\n",
       "                 'unp_start': 1}),\n",
       "               (('1hw2', 'A'),\n",
       "                {'chemicals': ['DC', 'DG', 'DA', 'DT', 'MG'],\n",
       "                 'experimental_method': 'X-RAY DIFFRACTION',\n",
       "                 'mmcif_header': '1hw2.header.cif',\n",
       "                 'pdb_chain_id': 'A',\n",
       "                 'pdb_end': 239,\n",
       "                 'pdb_file': '1hw2.pdb',\n",
       "                 'pdb_id': '1hw2',\n",
       "                 'pdb_start': 1,\n",
       "                 'rank': 7,\n",
       "                 'release_date': '2001-01-22',\n",
       "                 'resolution': 3.25,\n",
       "                 'seq_coverage': 1,\n",
       "                 'taxonomy_id': 562,\n",
       "                 'taxonomy_name': 'Escherichia coli',\n",
       "                 'uniprot_acc': 'P0A8V6',\n",
       "                 'unp_end': 239,\n",
       "                 'unp_start': 1}),\n",
       "               (('1hw2', 'B'),\n",
       "                {'chemicals': ['DC', 'DG', 'DA', 'DT', 'MG'],\n",
       "                 'experimental_method': 'X-RAY DIFFRACTION',\n",
       "                 'mmcif_header': '1hw2.header.cif',\n",
       "                 'pdb_chain_id': 'B',\n",
       "                 'pdb_end': 239,\n",
       "                 'pdb_file': '1hw2.pdb',\n",
       "                 'pdb_id': '1hw2',\n",
       "                 'pdb_start': 1,\n",
       "                 'rank': 8,\n",
       "                 'release_date': '2001-01-22',\n",
       "                 'resolution': 3.25,\n",
       "                 'seq_coverage': 1,\n",
       "                 'taxonomy_id': 562,\n",
       "                 'taxonomy_name': 'Escherichia coli',\n",
       "                 'uniprot_acc': 'P0A8V6',\n",
       "                 'unp_end': 239,\n",
       "                 'unp_start': 1})]),\n",
       "  'representative': {'clean_pdb_file': None,\n",
       "   'original_pdb_file': None,\n",
       "   'seq_coverage': 0,\n",
       "   'structure_id': None}}}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_gempro.genes.get_by_id('b1187').annotation"
   ]
  }
 ],
 "metadata": {
  "_draft": {
   "nbviewer_url": "https://gist.github.com/342e44766d1dda637e975a4ef61dae19"
  },
  "anaconda-cloud": {},
  "gist": {
   "data": {
    "description": "GEM-PRO pipeline example for a list of gene IDs",
    "public": false
   },
   "id": "342e44766d1dda637e975a4ef61dae19"
  },
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
