<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>ssbio.gempro.pipeline &#8212; ssbio 0.1 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="top" title="ssbio 0.1 documentation" href="../../../index.html" />
    <link rel="up" title="Module code" href="../../index.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for ssbio.gempro.pipeline</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">os.path</span> <span class="k">as</span> <span class="nn">op</span>
<span class="kn">import</span> <span class="nn">shutil</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">import</span> <span class="nn">glob</span>

<span class="kn">from</span> <span class="nn">cobra.io</span> <span class="k">import</span> <span class="n">load_matlab_model</span>
<span class="kn">from</span> <span class="nn">cobra.io.sbml</span> <span class="k">import</span> <span class="n">create_cobra_model_from_sbml_file</span>

<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">SeqIO</span>

<span class="kn">from</span> <span class="nn">bioservices.uniprot</span> <span class="k">import</span> <span class="n">UniProt</span>

<span class="n">bsup</span> <span class="o">=</span> <span class="n">UniProt</span><span class="p">()</span>
<span class="kn">from</span> <span class="nn">bidict</span> <span class="k">import</span> <span class="n">bidict</span>

<span class="kn">from</span> <span class="nn">ssbio</span> <span class="k">import</span> <span class="n">utils</span>
<span class="kn">from</span> <span class="nn">ssbio.cobra.utils</span> <span class="k">import</span> <span class="n">is_spontaneous</span><span class="p">,</span> <span class="n">true_num_reactions</span><span class="p">,</span> <span class="n">true_num_genes</span>

<span class="n">date</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">Date</span><span class="p">()</span>

<span class="kn">import</span> <span class="nn">ssbio.databases.kegg</span>
<span class="kn">import</span> <span class="nn">ssbio.databases.uniprot</span>
<span class="kn">import</span> <span class="nn">ssbio.databases.pdb</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">from</span> <span class="nn">tqdm</span> <span class="k">import</span> <span class="n">tqdm</span>
<span class="kn">import</span> <span class="nn">requests</span>

<span class="kn">from</span> <span class="nn">ssbio.databases.pdb</span> <span class="k">import</span> <span class="n">top_pdb_blast_hit</span>

<span class="kn">from</span> <span class="nn">bioservices</span> <span class="k">import</span> <span class="n">KEGG</span>

<span class="n">bs_kegg</span> <span class="o">=</span> <span class="n">KEGG</span><span class="p">()</span>


<div class="viewcode-block" id="GEMPRO"><a class="viewcode-back" href="../../../python_api.html#ssbio.gempro.pipeline.GEMPRO">[docs]</a><span class="k">class</span> <span class="nc">GEMPRO</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Generic class to represent all information of a GEM-PRO for a GEM</span>

<span class="sd">    Main steps are:</span>
<span class="sd">    1. Preparation of folders</span>
<span class="sd">    2. Automatic ID mapping</span>
<span class="sd">    3. QC/QA</span>
<span class="sd">    4. Homology modeling</span>
<span class="sd">    5. Representative structures</span>

<span class="sd">    Each step may generate a report and also request additional files if mappings are incomplete</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gem_name</span><span class="p">,</span> <span class="n">root_dir</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">root_dir</span> <span class="o">=</span> <span class="n">root_dir</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">model_dir</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">root_dir</span><span class="p">,</span> <span class="n">gem_name</span><span class="p">)</span>
        <span class="c1"># model_files - directory where original gems and gem-related files are stored</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">model_files</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_dir</span><span class="p">,</span> <span class="s1">&#39;model_files&#39;</span><span class="p">)</span>

<div class="viewcode-block" id="GEMPRO.prep_folders"><a class="viewcode-back" href="../../../python_api.html#ssbio.gempro.pipeline.GEMPRO.prep_folders">[docs]</a>    <span class="k">def</span> <span class="nf">prep_folders</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Prepare all folders for a GEM-PRO project</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># data - directory where all data will be stored</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_dir</span><span class="p">,</span> <span class="s1">&#39;data&#39;</span><span class="p">)</span>

        <span class="c1"># notebooks - directory where ipython notebooks will be stored for manual analyses</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">notebooks</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_dir</span><span class="p">,</span> <span class="s1">&#39;notebooks&#39;</span><span class="p">)</span>

        <span class="c1"># # figures - directory where all figures will be stored</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">figures</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_dir</span><span class="p">,</span> <span class="s1">&#39;figures&#39;</span><span class="p">)</span>

        <span class="c1"># missing - directory where missing reports go</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">missing</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_dir</span><span class="p">,</span> <span class="s1">&#39;missing&#39;</span><span class="p">)</span>

        <span class="c1"># struct_files - directory where structure related files will be downloaded/are located</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">struct_files</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_dir</span><span class="p">,</span> <span class="s1">&#39;structure_files&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">struct_single_chain</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">struct_files</span><span class="p">,</span> <span class="s1">&#39;by_gene&#39;</span><span class="p">)</span>

        <span class="c1"># seq_files - sequence related files are stored here</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seq_files</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_dir</span><span class="p">,</span> <span class="s1">&#39;sequence_files&#39;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">directory</span> <span class="ow">in</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">model_dir</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">notebooks</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">figures</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">missing</span><span class="p">,</span>
                          <span class="bp">self</span><span class="o">.</span><span class="n">model_files</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">struct_files</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">struct_single_chain</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_files</span><span class="p">]:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">directory</span><span class="p">):</span>
                <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">directory</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[PREP] Created directory: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">directory</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[PREP] Directory already exists: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">directory</span><span class="p">))</span></div>

<div class="viewcode-block" id="GEMPRO.load_model"><a class="viewcode-back" href="../../../python_api.html#ssbio.gempro.pipeline.GEMPRO.load_model">[docs]</a>    <span class="k">def</span> <span class="nf">load_model</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gem_file</span><span class="p">,</span> <span class="n">file_type</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Load the GEM using COBRApy. Accept SBML or MAT files as input</span>

<span class="sd">        Args:</span>
<span class="sd">            file_type: if your model is in &quot;SBML&quot; (or &quot;XML&quot;), or &quot;Matlab&quot; formats</span>

<span class="sd">        Returns: None - loads the model in the GEMPRO object instance</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">extension</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">gem_file</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">file_type</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s1">&#39;xml&#39;</span> <span class="ow">or</span> <span class="n">file_type</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s1">&#39;sbml&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">=</span> <span class="n">create_cobra_model_from_sbml_file</span><span class="p">(</span><span class="n">gem_file</span><span class="p">,</span> <span class="n">print_time</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[PREP] Loaded model: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gem_file</span><span class="p">))</span>
        <span class="k">elif</span> <span class="n">extension</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s1">&#39;mat&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">=</span> <span class="n">load_matlab_model</span><span class="p">(</span><span class="n">gem_file</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[PREP] Loaded model: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gem_file</span><span class="p">))</span>

        <span class="c1"># place a copy of the current used model in model_files</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_files</span><span class="p">,</span> <span class="n">op</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">gem_file</span><span class="p">))):</span>
            <span class="n">shutil</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">gem_file</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">model_files</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[PREP] Copied model file to model_files&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gem_file</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_files</span><span class="p">,</span> <span class="n">op</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">gem_file</span><span class="p">))</span>

        <span class="c1"># obtain list of all gene ids, excluding spontaneous things</span>
        <span class="c1"># TODO: account for isoforms</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genes_unfiltered</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">genes</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">y</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes_unfiltered</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">is_spontaneous</span><span class="p">(</span><span class="n">y</span><span class="p">)]</span>

        <span class="c1"># print information on the number of things</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[MODEL] Number of reactions: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">reactions</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[MODEL] Number of reactions linked to a gene: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">true_num_reactions</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[MODEL] Number of genes (excluding spontaneous): </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">true_num_genes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[MODEL] Number of metabolites: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">metabolites</span><span class="p">)))</span></div>

<div class="viewcode-block" id="GEMPRO.kegg_mapping_and_metadata"><a class="viewcode-back" href="../../../python_api.html#ssbio.gempro.pipeline.GEMPRO.kegg_mapping_and_metadata">[docs]</a>    <span class="k">def</span> <span class="nf">kegg_mapping_and_metadata</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">kegg_organism_code</span><span class="p">,</span> <span class="n">custom_gene_mapping</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">force_rerun</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Map all genes in the model to UniProt IDs using the KEGG service. Also download all metadata and sequences</span>

<span class="sd">        Args:</span>
<span class="sd">            kegg_organism_code (str): The three letter KEGG code of your organism</span>
<span class="sd">            custom_gene_mapping (dict): If your model genes differ from the gene IDs you want to map,</span>
<span class="sd">                custom_gene_mapping allows you to input a dictionary which maps model gene IDs to new ones.</span>
<span class="sd">                Dictionary keys must match model genes.</span>
<span class="sd">            force_rerun (bool): Flag to indicate if function should be rerun (True) or if</span>
<span class="sd">                cached files that match the format *_kegg_df.csv should be used instead (default, False).</span>

<span class="sd">        Returns: path to metadata/mapping dataframe</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># all data will be stored in a dataframe</span>
        <span class="n">kegg_pre_df</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># first map all of the organism&#39;s kegg genes to UniProt</span>
        <span class="n">kegg_to_uniprot</span> <span class="o">=</span> <span class="n">ssbio</span><span class="o">.</span><span class="n">databases</span><span class="o">.</span><span class="n">kegg</span><span class="o">.</span><span class="n">map_kegg_all_genes</span><span class="p">(</span><span class="n">organism_code</span><span class="o">=</span><span class="n">kegg_organism_code</span><span class="p">,</span> <span class="n">target_db</span><span class="o">=</span><span class="s1">&#39;uniprot&#39;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">kegg_missing_genes</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">kegg_df_outfile</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">-kegg_df.csv&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">date</span><span class="o">.</span><span class="n">short_date</span><span class="p">))</span>
        <span class="n">err_file</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">missing</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">-kegg_mapping.err&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">date</span><span class="o">.</span><span class="n">short_date</span><span class="p">))</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">force_rerun</span><span class="p">:</span>
            <span class="c1"># check for already saved mappings</span>
            <span class="c1"># load the latest mapping if there are multiple</span>
            <span class="n">saved_dfs</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">rank_dated_files</span><span class="p">(</span><span class="s1">&#39;*-kegg_df.csv&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
            <span class="n">saved_err</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">rank_dated_files</span><span class="p">(</span><span class="s1">&#39;*-kegg_mapping.err&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">missing</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">saved_err</span><span class="p">:</span>
                <span class="n">to_load_err</span> <span class="o">=</span> <span class="n">saved_err</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">err_date</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">to_load_err</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">to_load_err</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;rt&#39;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">error_file</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">kegg_missing_genes</span> <span class="o">=</span> <span class="n">error_file</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span>
                    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[INFO] Loaded KEGG mapping errors dated </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">err_date</span><span class="p">))</span>

            <span class="k">if</span> <span class="n">saved_dfs</span><span class="p">:</span>
                <span class="n">to_load_df</span> <span class="o">=</span> <span class="n">saved_dfs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">df_date</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">to_load_df</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">kegg_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">to_load_df</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[INFO] Loaded KEGG mapping dataframe dated </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">df_date</span><span class="p">))</span>
                <span class="k">return</span>

        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">):</span>

            <span class="k">if</span> <span class="n">custom_gene_mapping</span><span class="p">:</span>
                <span class="n">kegg_g</span> <span class="o">=</span> <span class="n">custom_gene_mapping</span><span class="p">[</span><span class="n">g</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">kegg_g</span> <span class="o">=</span> <span class="n">g</span>

            <span class="n">kegg_dict</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">kegg_dict</span><span class="p">[</span><span class="s1">&#39;m_gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">g</span>
            <span class="n">kegg_dict</span><span class="p">[</span><span class="s1">&#39;k_kegg_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kegg_g</span>

            <span class="c1"># make the gene specific folder under the sequence_files directory</span>
            <span class="n">gene_folder</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">seq_files</span><span class="p">,</span> <span class="n">g</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">):</span>
                <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">)</span>

            <span class="c1"># download kegg metadata</span>
            <span class="n">metadata_file</span> <span class="o">=</span> <span class="n">ssbio</span><span class="o">.</span><span class="n">databases</span><span class="o">.</span><span class="n">kegg</span><span class="o">.</span><span class="n">download_kegg_gene_metadata</span><span class="p">(</span><span class="n">organism_code</span><span class="o">=</span><span class="n">kegg_organism_code</span><span class="p">,</span>
                                                                             <span class="n">gene_id</span><span class="o">=</span><span class="n">kegg_g</span><span class="p">,</span>
                                                                             <span class="n">out_dir</span><span class="o">=</span><span class="n">gene_folder</span><span class="p">)</span>

            <span class="c1"># download kegg sequence</span>
            <span class="n">sequence_file</span> <span class="o">=</span> <span class="n">ssbio</span><span class="o">.</span><span class="n">databases</span><span class="o">.</span><span class="n">kegg</span><span class="o">.</span><span class="n">download_kegg_aa_seq</span><span class="p">(</span><span class="n">organism_code</span><span class="o">=</span><span class="n">kegg_organism_code</span><span class="p">,</span>
                                                                      <span class="n">gene_id</span><span class="o">=</span><span class="n">kegg_g</span><span class="p">,</span>
                                                                      <span class="n">out_dir</span><span class="o">=</span><span class="n">gene_folder</span><span class="p">)</span>

            <span class="c1"># if there is no FASTA for this gene, consider it missing and do not add it to the dataframe</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">sequence_file</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">kegg_missing_genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span><span class="p">)</span>
                <span class="k">continue</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">kegg_dict</span><span class="p">[</span><span class="s1">&#39;k_seq_file&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">sequence_file</span><span class="p">)</span>
                <span class="n">kegg_dict</span><span class="p">[</span><span class="s1">&#39;k_seq_len&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">sequence_file</span><span class="p">),</span> <span class="s2">&quot;fasta&quot;</span><span class="p">))</span>

            <span class="k">if</span> <span class="n">metadata_file</span><span class="p">:</span>
                <span class="n">kegg_dict</span><span class="p">[</span><span class="s1">&#39;k_metadata_file&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">metadata_file</span><span class="p">)</span>

                <span class="c1"># parse PDB mapping from metadata file</span>
                <span class="c1"># in some cases, KEGG IDs do not map to a UniProt ID - this is to ensure we get the PDB mapping</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">metadata_file</span><span class="p">)</span> <span class="k">as</span> <span class="n">mf</span><span class="p">:</span>
                    <span class="n">kegg_parsed</span> <span class="o">=</span> <span class="n">bs_kegg</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">mf</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
                <span class="k">if</span> <span class="s1">&#39;STRUCTURE&#39;</span> <span class="ow">in</span> <span class="n">kegg_parsed</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="n">kegg_dict</span><span class="p">[</span><span class="s1">&#39;k_pdb_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kegg_parsed</span><span class="p">[</span><span class="s1">&#39;STRUCTURE&#39;</span><span class="p">]</span>
                    <span class="c1"># TODO: there is a lot more you can get from the KEGG metadata file, examples below (consider saving it in the DF)</span>
                    <span class="c1"># &#39;DBLINKS&#39;: {&#39;NCBI-GeneID&#39;: &#39;100763844&#39;, &#39;NCBI-ProteinID&#39;: &#39;XP_003514445&#39;}</span>
                    <span class="c1"># &#39;ORTHOLOGY&#39;: {&#39;K00473&#39;: &#39;procollagen-lysine,2-oxoglutarate 5-dioxygenase 1 [EC:1.14.11.4]&#39;},</span>
                    <span class="c1"># &#39;PATHWAY&#39;: {&#39;cge00310&#39;: &#39;Lysine degradation&#39;},</span>

            <span class="k">if</span> <span class="n">kegg_g</span> <span class="ow">in</span> <span class="n">kegg_to_uniprot</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">kegg_dict</span><span class="p">[</span><span class="s1">&#39;k_uniprot_acc&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kegg_to_uniprot</span><span class="p">[</span><span class="n">kegg_g</span><span class="p">]</span>

            <span class="n">kegg_pre_df</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">kegg_dict</span><span class="p">)</span>

        <span class="c1"># save a log file for the genes that could not be mapped</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">kegg_missing_genes</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
                <span class="s1">&#39;</span><span class="si">{}</span><span class="s1"> genes were not able to be mapped to a KEGG entry. Please see </span><span class="si">{}</span><span class="s1"> for a list of these genes&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">kegg_missing_genes</span><span class="p">),</span>
                    <span class="n">err_file</span><span class="p">))</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">err_file</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;wt&#39;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">myfile</span><span class="p">:</span>
                <span class="n">myfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">kegg_missing_genes</span><span class="p">))</span>

        <span class="c1"># save a dataframe of the file mapping info</span>
        <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;m_gene&#39;</span><span class="p">,</span> <span class="s1">&#39;k_kegg_id&#39;</span><span class="p">,</span> <span class="s1">&#39;k_uniprot_acc&#39;</span><span class="p">,</span> <span class="s1">&#39;k_pdb_id&#39;</span><span class="p">,</span> <span class="s1">&#39;k_metadata_file&#39;</span><span class="p">,</span> <span class="s1">&#39;k_seq_file&#39;</span><span class="p">,</span>
                <span class="s1">&#39;k_seq_len&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">kegg_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">kegg_pre_df</span><span class="p">)[</span><span class="n">cols</span><span class="p">]</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;m_gene&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">kegg_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">kegg_df_outfile</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[INFO] Saved KEGG information at </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">kegg_df_outfile</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">kegg_df_outfile</span></div>

<div class="viewcode-block" id="GEMPRO.uniprot_mapping_and_metadata"><a class="viewcode-back" href="../../../python_api.html#ssbio.gempro.pipeline.GEMPRO.uniprot_mapping_and_metadata">[docs]</a>    <span class="k">def</span> <span class="nf">uniprot_mapping_and_metadata</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">model_gene_source</span><span class="p">,</span> <span class="n">custom_gene_mapping</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">force_rerun</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Map all genes in the model to UniProt IDs using the UniProt mapping service. Also download all metadata and sequences.</span>

<span class="sd">        Args:</span>
<span class="sd">            model_gene_source (str): the database source of your model gene IDs, see http://www.uniprot.org/help/programmatic_access</span>
<span class="sd">                Common model gene sources are:</span>
<span class="sd">                Ensembl Genomes - ENSEMBLGENOME_ID</span>
<span class="sd">                Entrez Gene (GeneID) - P_ENTREZGENEID</span>
<span class="sd">                RefSeq Protein - P_REFSEQ_AC</span>
<span class="sd">            custom_gene_mapping (dict): If your model genes differ from the gene IDs you want to map,</span>
<span class="sd">                custom_gene_mapping allows you to input a dictionary which maps model gene IDs to new ones.</span>
<span class="sd">                Dictionary keys must match model genes.</span>
<span class="sd">            force_rerun (bool): Flag to indicate if function should be rerun (True) or if</span>
<span class="sd">                cached files that match the format *_uniprot_df.csv should be used instead (default, False).</span>

<span class="sd">        Returns: path to metadata/mapping dataframe</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># allow model gene --&gt; custom ID mapping</span>
        <span class="k">if</span> <span class="n">custom_gene_mapping</span><span class="p">:</span>
            <span class="c1"># create a reverse mapping dictionary to map from custom ID -&gt; model gene later</span>
            <span class="n">rev_map</span> <span class="o">=</span> <span class="n">bidict</span><span class="p">(</span><span class="n">custom_gene_mapping</span><span class="p">)</span>
            <span class="n">genes_to_map</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">custom_gene_mapping</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">genes_to_map</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">force_rerun</span><span class="p">:</span>
            <span class="c1"># check for already saved mappings</span>
            <span class="c1"># load the latest mapping if there are multiple</span>
            <span class="n">saved_dfs</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">rank_dated_files</span><span class="p">(</span><span class="s1">&#39;*-uniprot_df.csv&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
            <span class="n">saved_err</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">rank_dated_files</span><span class="p">(</span><span class="s1">&#39;*-uniprot_mapping.err&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">missing</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">saved_err</span><span class="p">:</span>
                <span class="n">to_load_err</span> <span class="o">=</span> <span class="n">saved_err</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">err_date</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">to_load_err</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">to_load_err</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;rt&#39;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">error_file</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">uniprot_missing_genes</span> <span class="o">=</span> <span class="n">error_file</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span>
                    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[INFO] Loaded UniProt mapping errors dated </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">err_date</span><span class="p">))</span>

            <span class="k">if</span> <span class="n">saved_dfs</span><span class="p">:</span>
                <span class="n">to_load_df</span> <span class="o">=</span> <span class="n">saved_dfs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">df_date</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">to_load_df</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">uniprot_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">to_load_df</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[INFO] Loaded UniProt mapping dataframe dated </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">df_date</span><span class="p">))</span>
                <span class="k">return</span>

        <span class="n">genes_to_uniprots</span> <span class="o">=</span> <span class="n">bsup</span><span class="o">.</span><span class="n">mapping</span><span class="p">(</span><span class="n">fr</span><span class="o">=</span><span class="n">model_gene_source</span><span class="p">,</span> <span class="n">to</span><span class="o">=</span><span class="s1">&#39;ACC&#39;</span><span class="p">,</span> <span class="n">query</span><span class="o">=</span><span class="n">genes_to_map</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">uniprot_missing_genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">genes_to_map</span><span class="p">)</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span><span class="n">genes_to_uniprots</span><span class="o">.</span><span class="n">keys</span><span class="p">()))</span>

        <span class="c1"># save a log file for the genes that could not be mapped</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">uniprot_missing_genes</span><span class="p">:</span>
            <span class="n">err_file</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">missing</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">-uniprot_mapping.err&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">date</span><span class="o">.</span><span class="n">short_date</span><span class="p">))</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
                <span class="s1">&#39;[INFO] </span><span class="si">{}</span><span class="s1"> genes were not able to be mapped to a UniProt entry. Please see </span><span class="si">{}</span><span class="s1"> for a list of these genes&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">uniprot_missing_genes</span><span class="p">),</span>
                    <span class="n">err_file</span><span class="p">))</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">err_file</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;wt&#39;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">myfile</span><span class="p">:</span>
                <span class="n">myfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">uniprot_missing_genes</span><span class="p">))</span>

        <span class="c1"># all data will be stored in a dataframe</span>
        <span class="n">uniprot_pre_df</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">g</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">genes_to_uniprots</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
            <span class="c1"># TODO: reverse mapping to custom ids could probably be done in a better way</span>
            <span class="k">if</span> <span class="n">custom_gene_mapping</span><span class="p">:</span>
                <span class="n">original_m_gene</span> <span class="o">=</span> <span class="n">rev_map</span><span class="p">[</span><span class="n">g</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">original_m_gene</span> <span class="o">=</span> <span class="n">g</span>

            <span class="k">for</span> <span class="n">mapped_uniprot</span> <span class="ow">in</span> <span class="n">v</span><span class="p">:</span>
                <span class="n">uniprot_dict</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="n">uniprot_dict</span><span class="p">[</span><span class="s1">&#39;m_gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">original_m_gene</span>
                <span class="n">uniprot_dict</span><span class="p">[</span><span class="s1">&#39;u_uniprot_acc&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mapped_uniprot</span>

                <span class="c1"># make the gene specific folder under the sequence_files directory</span>
                <span class="n">gene_folder</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">seq_files</span><span class="p">,</span> <span class="n">g</span><span class="p">)</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">):</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">)</span>

                <span class="c1"># download uniprot metadata</span>
                <span class="n">metadata_file</span> <span class="o">=</span> <span class="n">ssbio</span><span class="o">.</span><span class="n">databases</span><span class="o">.</span><span class="n">uniprot</span><span class="o">.</span><span class="n">download_uniprot_file</span><span class="p">(</span><span class="n">uniprot_id</span><span class="o">=</span><span class="n">mapped_uniprot</span><span class="p">,</span> <span class="n">filetype</span><span class="o">=</span><span class="s1">&#39;txt&#39;</span><span class="p">,</span>
                                                                              <span class="n">outdir</span><span class="o">=</span><span class="n">gene_folder</span><span class="p">)</span>

                <span class="c1"># download uniprot sequence</span>
                <span class="n">sequence_file</span> <span class="o">=</span> <span class="n">ssbio</span><span class="o">.</span><span class="n">databases</span><span class="o">.</span><span class="n">uniprot</span><span class="o">.</span><span class="n">download_uniprot_file</span><span class="p">(</span><span class="n">uniprot_id</span><span class="o">=</span><span class="n">mapped_uniprot</span><span class="p">,</span>
                                                                              <span class="n">filetype</span><span class="o">=</span><span class="s1">&#39;fasta&#39;</span><span class="p">,</span>
                                                                              <span class="n">outdir</span><span class="o">=</span><span class="n">gene_folder</span><span class="p">)</span>

                <span class="n">uniprot_dict</span><span class="p">[</span><span class="s1">&#39;u_seq_file&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">sequence_file</span><span class="p">)</span>
                <span class="n">uniprot_dict</span><span class="p">[</span><span class="s1">&#39;u_metadata_file&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">metadata_file</span><span class="p">)</span>

                <span class="c1"># adding additional uniprot metadata</span>
                <span class="n">metadata</span> <span class="o">=</span> <span class="n">ssbio</span><span class="o">.</span><span class="n">databases</span><span class="o">.</span><span class="n">uniprot</span><span class="o">.</span><span class="n">parse_uniprot_txt_file</span><span class="p">(</span><span class="n">metadata_file</span><span class="p">)</span>
                <span class="n">uniprot_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">metadata</span><span class="p">)</span>

                <span class="n">uniprot_pre_df</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">uniprot_dict</span><span class="p">)</span>

        <span class="c1"># save a dataframe of the file mapping info</span>
        <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;m_gene&#39;</span><span class="p">,</span> <span class="s1">&#39;u_uniprot_acc&#39;</span><span class="p">,</span> <span class="s1">&#39;u_gene_name&#39;</span><span class="p">,</span> <span class="s1">&#39;u_reviewed&#39;</span><span class="p">,</span> <span class="s1">&#39;u_seq_len&#39;</span><span class="p">,</span> <span class="s1">&#39;u_kegg_id&#39;</span><span class="p">,</span> <span class="s1">&#39;u_refseq&#39;</span><span class="p">,</span>
                <span class="s1">&#39;u_ec_number&#39;</span><span class="p">,</span> <span class="s1">&#39;u_pfam&#39;</span><span class="p">,</span> <span class="s1">&#39;u_description&#39;</span><span class="p">,</span> <span class="s1">&#39;u_entry_version&#39;</span><span class="p">,</span> <span class="s1">&#39;u_seq_version&#39;</span><span class="p">,</span> <span class="s1">&#39;u_metadata_file&#39;</span><span class="p">,</span>
                <span class="s1">&#39;u_seq_file&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">uniprot_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">uniprot_pre_df</span><span class="p">)[</span><span class="n">cols</span><span class="p">]</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;m_gene&#39;</span><span class="p">)</span>
        <span class="n">uniprot_df_outfile</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">-uniprot_df.csv&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">date</span><span class="o">.</span><span class="n">short_date</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">uniprot_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">uniprot_df_outfile</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[INFO] Saved UniProt information at </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">uniprot_df_outfile</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">uniprot_df_outfile</span></div>

<div class="viewcode-block" id="GEMPRO.manual_id_mapping"><a class="viewcode-back" href="../../../python_api.html#ssbio.gempro.pipeline.GEMPRO.manual_id_mapping">[docs]</a>    <span class="k">def</span> <span class="nf">manual_id_mapping</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mapping_infile</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Read a manual input file of model gene IDs --&gt; KEGG or UniProt IDs.</span>

<span class="sd">        This allows for mapping of the missing genes, or overriding of automatic mappings.</span>

<span class="sd">        Args:</span>
<span class="sd">            mapping_infile: Path to .csv file with the headers:</span>
<span class="sd">                m_gene - model gene ID</span>


<span class="sd">        Returns:</span>

<span class="sd">        &quot;&quot;&quot;</span></div>

<div class="viewcode-block" id="GEMPRO.cosonlidate_mappings"><a class="viewcode-back" href="../../../python_api.html#ssbio.gempro.pipeline.GEMPRO.cosonlidate_mappings">[docs]</a>    <span class="k">def</span> <span class="nf">cosonlidate_mappings</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Combine information from KEGG, UniProt, and manual mappings.</span>

<span class="sd">        Returns:</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># compare gene -&gt; uniprot mappings</span>

        <span class="c1"># compare sequence lengths</span>

        <span class="c1"># if different, compare sequences?</span>

        <span class="c1"># sometimes KEGG won&#39;t map to a uniprot ID, but will map to a PDB - this needs to be saved somewhere</span>


        <span class="k">pass</span></div>

<div class="viewcode-block" id="GEMPRO.blast_seqs_to_pdb"><a class="viewcode-back" href="../../../python_api.html#ssbio.gempro.pipeline.GEMPRO.blast_seqs_to_pdb">[docs]</a>    <span class="k">def</span> <span class="nf">blast_seqs_to_pdb</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">custom_df</span><span class="p">,</span> <span class="n">evalue</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span> <span class="n">link</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;BLAST each gene sequence to the PDB. Results are saved as dataframes in the structure_files folder.</span>

<span class="sd">        Returns: Path to summary results dataframe which details best hit for each gene.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">top_blast_pre_df</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">custom_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()):</span>
            <span class="n">g</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;m_gene&#39;</span><span class="p">]</span>
            <span class="n">seq</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;u_seq&#39;</span><span class="p">]</span>

            <span class="c1"># make the gene specific folder under the structure_files directory</span>
            <span class="n">gene_folder</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">struct_single_chain</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">g</span><span class="p">))</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">):</span>
                <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">)</span>

            <span class="n">blast_df</span> <span class="o">=</span> <span class="n">ssbio</span><span class="o">.</span><span class="n">databases</span><span class="o">.</span><span class="n">pdb</span><span class="o">.</span><span class="n">blast_pdb</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">evalue</span><span class="o">=</span><span class="n">evalue</span><span class="p">,</span> <span class="n">link</span><span class="o">=</span><span class="n">link</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">blast_df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># save the blast_df</span>
                <span class="n">blast_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_folder</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">g</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;_blast_df.csv&#39;</span><span class="p">))</span>
                <span class="c1"># save the top blast hit</span>
                <span class="n">top_hit</span> <span class="o">=</span> <span class="n">blast_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
                <span class="n">top_hit</span><span class="p">[</span><span class="s1">&#39;m_gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">g</span>
                <span class="n">top_blast_pre_df</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">top_hit</span><span class="p">)</span>

        <span class="n">top_blast_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">top_blast_pre_df</span><span class="p">)</span>
        <span class="n">reorg</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;m_gene&#39;</span><span class="p">,</span> <span class="s1">&#39;hit_pdb&#39;</span><span class="p">,</span> <span class="s1">&#39;hit_pdb_chain&#39;</span><span class="p">,</span> <span class="s1">&#39;hit_evalue&#39;</span><span class="p">,</span> <span class="s1">&#39;hit_score&#39;</span><span class="p">,</span> <span class="s1">&#39;hit_num_ident&#39;</span><span class="p">,</span> <span class="s1">&#39;hit_percent_ident&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">top_blast_df</span> <span class="o">=</span> <span class="n">top_blast_df</span><span class="p">[</span><span class="n">reorg</span><span class="p">]</span>
        <span class="n">top_blast_df_outfile</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">-pdb_blast_df.csv&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">date</span><span class="o">.</span><span class="n">short_date</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">top_blast_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">top_blast_df_outfile</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">top_blast_df_outfile</span></div>

<div class="viewcode-block" id="GEMPRO.organize_itasser_models"><a class="viewcode-back" href="../../../python_api.html#ssbio.gempro.pipeline.GEMPRO.organize_itasser_models">[docs]</a>    <span class="k">def</span> <span class="nf">organize_itasser_models</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">raw_dir</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Reorganize the raw information from I-TASSER modeling.</span>

<span class="sd">        - Copies the model1.pdb file to the matching gene folder in structure_files/by_gene/</span>
<span class="sd">        - Parses modeling information such as C-scores, template, etc</span>
<span class="sd">        - Also will parse binding site (COACH) information if it exists</span>

<span class="sd">        TODO: what else to parse?</span>
<span class="sd">        TODO: finish summary df stuff</span>

<span class="sd">        Returns: path to summary results dataframe</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">hom_pre_df</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">mode11_counter</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">model1_bsites_counter</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">):</span>
            <span class="n">hom_dict</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">g_files</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">raw_dir</span><span class="p">,</span> <span class="n">g</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span><span class="p">))</span>

            <span class="c1">### homology model results</span>
            <span class="n">model_folder</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">raw_dir</span><span class="p">,</span> <span class="n">g</span><span class="p">)</span>
            <span class="c1"># best homology model is &quot;model1.pdb&quot;</span>
            <span class="n">model1_file</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">model_folder</span><span class="p">,</span> <span class="s1">&#39;model1.pdb&#39;</span><span class="p">)</span>
            <span class="n">init_dat</span> <span class="o">=</span> <span class="s1">&#39;init.dat&#39;</span>
            <span class="n">init_dat_path</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">model_folder</span><span class="p">,</span> <span class="n">init_dat</span><span class="p">)</span>

            <span class="c1">### coach (binding site prediction) results</span>
            <span class="n">model1_coach_folder</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">model_folder</span><span class="p">,</span> <span class="s1">&#39;model1&#39;</span><span class="p">,</span> <span class="s1">&#39;coach&#39;</span><span class="p">)</span>
            <span class="n">bsites_inf</span> <span class="o">=</span> <span class="s1">&#39;Bsites.inf&#39;</span>
            <span class="n">bsites_inf_path</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">model1_coach_folder</span><span class="p">,</span> <span class="n">bsites_inf</span><span class="p">)</span>
            <span class="n">ec</span> <span class="o">=</span> <span class="s1">&#39;EC.dat&#39;</span>
            <span class="n">ec_path</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">model1_coach_folder</span><span class="p">,</span> <span class="n">ec</span><span class="p">)</span>
            <span class="n">go_mf</span> <span class="o">=</span> <span class="s1">&#39;GO_MF.dat&#39;</span>
            <span class="n">go_mf_path</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">model1_coach_folder</span><span class="p">,</span> <span class="n">go_mf</span><span class="p">)</span>
            <span class="n">go_bp</span> <span class="o">=</span> <span class="s1">&#39;GO_BP.dat&#39;</span>
            <span class="n">go_bp_path</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">model1_coach_folder</span><span class="p">,</span> <span class="n">go_bp</span><span class="p">)</span>
            <span class="n">go_cc</span> <span class="o">=</span> <span class="s1">&#39;GO_CC.dat&#39;</span>
            <span class="n">go_cc_path</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">model1_coach_folder</span><span class="p">,</span> <span class="n">go_cc</span><span class="p">)</span>

            <span class="c1"># if the homology model has been completed</span>
            <span class="k">if</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">model1_file</span><span class="p">):</span>
                <span class="c1"># make the destination gene folder</span>
                <span class="n">dest_gene_dir</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">struct_single_chain</span><span class="p">,</span> <span class="n">g</span><span class="p">)</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">dest_gene_dir</span><span class="p">):</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">dest_gene_dir</span><span class="p">)</span>

                <span class="c1"># the new model file will be named by the gene</span>
                <span class="n">dest</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dest_gene_dir</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">.pdb&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="p">))</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">dest</span><span class="p">):</span>
                    <span class="n">shutil</span><span class="o">.</span><span class="n">copy2</span><span class="p">(</span><span class="n">model1_file</span><span class="p">,</span> <span class="n">dest</span><span class="p">)</span>
                <span class="n">hom_dict</span><span class="p">[</span><span class="s1">&#39;m_gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">g</span>
                <span class="n">hom_dict</span><span class="p">[</span><span class="s1">&#39;i_model&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">.pdb&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">g</span><span class="p">)</span>
                <span class="n">hom_pre_df</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">hom_dict</span><span class="p">)</span>
                <span class="n">mode11_counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="c1"># if the binding site predictions have been completed</span>
            <span class="k">if</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">model1_coach_folder</span><span class="p">):</span>
                <span class="c1"># make the destination folder</span>
                <span class="n">dest_coach_dir</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">struct_single_chain</span><span class="p">,</span> <span class="n">g</span><span class="p">,</span> <span class="s1">&#39;coach&#39;</span><span class="p">)</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">dest_coach_dir</span><span class="p">):</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">dest_coach_dir</span><span class="p">)</span>

                <span class="n">new_bsites_inf</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dest_coach_dir</span><span class="p">,</span> <span class="n">bsites_inf</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">bsites_inf_path</span><span class="p">):</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">new_bsites_inf</span><span class="p">):</span>
                        <span class="n">shutil</span><span class="o">.</span><span class="n">copy2</span><span class="p">(</span><span class="n">bsites_inf_path</span><span class="p">,</span> <span class="n">new_bsites_inf</span><span class="p">)</span>
                <span class="n">model1_bsites_counter</span> <span class="o">+=</span> <span class="mi">1</span>

                <span class="n">new_ec</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dest_coach_dir</span><span class="p">,</span> <span class="n">ec</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">ec_path</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">new_ec</span><span class="p">):</span>
                    <span class="n">shutil</span><span class="o">.</span><span class="n">copy2</span><span class="p">(</span><span class="n">ec_path</span><span class="p">,</span> <span class="n">new_ec</span><span class="p">)</span>

                <span class="n">new_go_mf</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dest_coach_dir</span><span class="p">,</span> <span class="n">go_mf</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">go_mf_path</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">new_go_mf</span><span class="p">):</span>
                    <span class="n">shutil</span><span class="o">.</span><span class="n">copy2</span><span class="p">(</span><span class="n">go_mf_path</span><span class="p">,</span> <span class="n">new_go_mf</span><span class="p">)</span>

                <span class="n">new_go_bp</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dest_coach_dir</span><span class="p">,</span> <span class="n">go_bp</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">go_bp_path</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">new_go_bp</span><span class="p">):</span>
                    <span class="n">shutil</span><span class="o">.</span><span class="n">copy2</span><span class="p">(</span><span class="n">go_bp_path</span><span class="p">,</span> <span class="n">new_go_bp</span><span class="p">)</span>

                <span class="n">new_go_cc</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dest_coach_dir</span><span class="p">,</span> <span class="n">go_cc</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">go_cc_path</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">op</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">new_go_cc</span><span class="p">):</span>
                    <span class="n">shutil</span><span class="o">.</span><span class="n">copy2</span><span class="p">(</span><span class="n">go_cc_path</span><span class="p">,</span> <span class="n">new_go_cc</span><span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[INFO] </span><span class="si">{}</span><span class="s1"> homology models completed and copied to </span><span class="si">{}</span><span class="s1">.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">mode11_counter</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">struct_single_chain</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">model1_bsites_counter</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[INFO] </span><span class="si">{}</span><span class="s1"> binding site predictions completed and copied.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">model1_bsites_counter</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">homology_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">hom_pre_df</span><span class="p">)[[</span><span class="s1">&#39;m_gene&#39;</span><span class="p">,</span> <span class="s1">&#39;i_model&#39;</span><span class="p">]]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">homology_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">-homology_models_df.csv&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">date</span><span class="o">.</span><span class="n">short_date</span><span class="p">)))</span>
        <span class="k">return</span> <span class="n">op</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">,</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">-homology_models_df.csv&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">date</span><span class="o">.</span><span class="n">short_date</span><span class="p">))</span></div>

<div class="viewcode-block" id="GEMPRO.run_pipeline"><a class="viewcode-back" href="../../../python_api.html#ssbio.gempro.pipeline.GEMPRO.run_pipeline">[docs]</a>    <span class="k">def</span> <span class="nf">run_pipeline</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Run the entire GEM-PRO pipel ine.</span>

<span class="sd">        Options include:</span>
<span class="sd">        ...</span>

<span class="sd">        Returns:</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">prep_folders</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">load_model</span><span class="p">()</span>
        <span class="c1"># map gene IDs</span>
        <span class="c1"># check for missing maps, request manual mapping file?</span>
        <span class="k">pass</span></div></div>


<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="c1"># run the GEM-PRO pipeline!</span>

    <span class="c1"># parse arguments</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s1">&#39;Runs the GEM-PRO pipeline&#39;</span><span class="p">)</span>
    <span class="n">p</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;gemfile&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Path to the GEM file&#39;</span><span class="p">)</span>
    <span class="n">p</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;gemname&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Name you would like to use to refer to this GEM&#39;</span><span class="p">)</span>
    <span class="n">p</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;rootdir&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Directory where GEM-PRO files should be stored&#39;</span><span class="p">)</span>

    <span class="n">args</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="n">my_gem_pro</span> <span class="o">=</span> <span class="n">GEMPRO</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">gemfile</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">gemname</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">rootdir</span><span class="p">)</span>
    <span class="n">my_gem_pro</span><span class="o">.</span><span class="n">run_pipeline</span><span class="p">()</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  <li><a href="../../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2016, Nathan Mih.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.4.8</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.9</a>
      
    </div>

    

    
  </body>
</html>