

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>4. Escher, COBRA, and COBRApy &mdash; ssbio 1.0.0a0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="_build/theme_overrides.css" type="text/css" />
  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="ssbio 1.0.0a0 documentation" href="index.html"/>
        <link rel="next" title="5. GEM-PRO - Genes &amp; Sequences" href="notebooks/GEM-PRO - Genes &amp; Sequences.html"/>
        <link rel="prev" title="3. ATLAS pipeline" href="atlas.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> ssbio
          

          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">1. Getting Started</a><ul>
<li class="toctree-l2"><a class="reference internal" href="getting_started.html#introduction">1.1. Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="getting_started.html#structural-biology">1.2. Structural biology</a><ul>
<li class="toctree-l3"><a class="reference internal" href="getting_started.html#protein-structures">1.2.1. Protein structures</a></li>
<li class="toctree-l3"><a class="reference internal" href="getting_started.html#biopython">1.2.2. Biopython</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="getting_started.html#systems-biology">1.3. Systems biology</a><ul>
<li class="toctree-l3"><a class="reference internal" href="getting_started.html#cobra-models-cobrapy">1.3.1. COBRA models &amp; COBRApy</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="getting_started.html#reading">1.4. Reading</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="gempro.html">2. GEM-PRO pipeline</a><ul>
<li class="toctree-l2"><a class="reference internal" href="gempro.html#introduction">2.1. Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="gempro.html#starting-with-an-sbml-model">2.2. Starting with an SBML model</a></li>
<li class="toctree-l2"><a class="reference internal" href="gempro.html#starting-with-a-list-of-gene-ids">2.3. Starting with a list of gene IDs</a></li>
<li class="toctree-l2"><a class="reference internal" href="gempro.html#starting-with-a-dictionary-of-gene-ids-and-their-sequences">2.4. Starting with a dictionary of gene IDs and their sequences</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="atlas.html">3. ATLAS pipeline</a><ul>
<li class="toctree-l2"><a class="reference internal" href="atlas.html#introduction">3.1. Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="atlas.html#starting-with-an-sbml-model">3.2. Starting with an SBML model</a></li>
<li class="toctree-l2"><a class="reference internal" href="atlas.html#starting-with-a-list-of-gene-ids">3.3. Starting with a list of gene IDs</a></li>
<li class="toctree-l2"><a class="reference internal" href="atlas.html#starting-with-a-dictionary-of-gene-ids-and-their-sequences">3.4. Starting with a dictionary of gene IDs and their sequences</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">4. Escher, COBRA, and COBRApy</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#maps-and-models">4.1. Maps and models</a></li>
<li class="toctree-l2"><a class="reference internal" href="#what-is-json-and-why-do-we-use-it">4.2. What is JSON and why do we use it?</a></li>
<li class="toctree-l2"><a class="reference internal" href="#escher-sbml-and-sbgn">4.3. Escher, SBML, and SBGN</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/GEM-PRO - Genes &amp; Sequences.html">5. GEM-PRO - Genes &amp; Sequences</a><ul>
<li class="toctree-l2"><a class="reference internal" href="notebooks/GEM-PRO - Genes &amp; Sequences.html#Imports">5.1. Imports</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/GEM-PRO - Genes &amp; Sequences.html#Logging">5.1.1. Logging</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/GEM-PRO - Genes &amp; Sequences.html#Initialization-of-the-project">5.2. Initialization of the project</a></li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/GEM-PRO - Genes &amp; Sequences.html#Mapping-sequence--&gt;-structure">5.3. Mapping sequence -&gt; structure</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/GEM-PRO - Genes &amp; Sequences.html#Methods">5.3.1. Methods</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - Genes &amp; Sequences.html#blast_seqs_to_pdb">5.3.1.1. <code class="docutils literal"><span class="pre">blast_seqs_to_pdb</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/GEM-PRO - Genes &amp; Sequences.html#Downloading-and-ranking-structures">5.4. Downloading and ranking structures</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/GEM-PRO - Genes &amp; Sequences.html#Methods">5.4.1. Methods</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - Genes &amp; Sequences.html#pdb_downloader_and_metadata">5.4.1.1. <code class="docutils literal"><span class="pre">pdb_downloader_and_metadata</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - Genes &amp; Sequences.html#set_representative_structure">5.4.1.2. <code class="docutils literal"><span class="pre">set_representative_structure</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/GEM-PRO - List of Gene IDs.html">6. GEM-PRO - List of Gene IDs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="notebooks/GEM-PRO - List of Gene IDs.html#Imports">6.1. Imports</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/GEM-PRO - List of Gene IDs.html#Logging">6.1.1. Logging</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/GEM-PRO - List of Gene IDs.html#Initialization-of-the-project">6.2. Initialization of the project</a></li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/GEM-PRO - List of Gene IDs.html#Mapping-gene-ID--&gt;-sequence">6.3. Mapping gene ID -&gt; sequence</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/GEM-PRO - List of Gene IDs.html#Methods">6.3.1. Methods</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - List of Gene IDs.html#kegg_mapping_and_metadata">6.3.1.1. <code class="docutils literal"><span class="pre">kegg_mapping_and_metadata</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - List of Gene IDs.html#uniprot_mapping_and_metadata">6.3.1.2. <code class="docutils literal"><span class="pre">uniprot_mapping_and_metadata</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - List of Gene IDs.html#set_representative_sequence">6.3.1.3. <code class="docutils literal"><span class="pre">set_representative_sequence</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/GEM-PRO - List of Gene IDs.html#Mapping-representative-sequence-to-structure">6.4. Mapping representative sequence to structure</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/GEM-PRO - List of Gene IDs.html#Methods">6.4.1. Methods</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - List of Gene IDs.html#map_uniprot_to_pdb">6.4.1.1. <code class="docutils literal"><span class="pre">map_uniprot_to_pdb</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - List of Gene IDs.html#blast_seqs_to_pdb">6.4.1.2. <code class="docutils literal"><span class="pre">blast_seqs_to_pdb</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/GEM-PRO - List of Gene IDs.html#Downloading-and-ranking-structures">6.5. Downloading and ranking structures</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/GEM-PRO - List of Gene IDs.html#Methods">6.5.1. Methods</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - List of Gene IDs.html#pdb_downloader_and_metadata">6.5.1.1. <code class="docutils literal"><span class="pre">pdb_downloader_and_metadata</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - List of Gene IDs.html#set_representative_structure">6.5.1.2. <code class="docutils literal"><span class="pre">set_representative_structure</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html">7. GEM-PRO - SBML Model (iNJ661)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#Imports">7.1. Imports</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#Logging">7.1.1. Logging</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#Initialization-of-the-project">7.2. Initialization of the project</a></li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#Mapping-gene-ID--&gt;-sequence">7.3. Mapping gene ID -&gt; sequence</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#Methods">7.3.1. Methods</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#manual_seq_mapping">7.3.1.1. <code class="docutils literal"><span class="pre">manual_seq_mapping</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#manual_uniprot_mapping">7.3.1.2. <code class="docutils literal"><span class="pre">manual_uniprot_mapping</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#kegg_mapping_and_metadata">7.3.1.3. <code class="docutils literal"><span class="pre">kegg_mapping_and_metadata</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#uniprot_mapping_and_metadata">7.3.1.4. <code class="docutils literal"><span class="pre">uniprot_mapping_and_metadata</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#set_representative_sequence">7.3.1.5. <code class="docutils literal"><span class="pre">set_representative_sequence</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#Mapping-representative-sequence-to-structure">7.4. Mapping representative sequence to structure</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#Methods">7.4.1. Methods</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#map_uniprot_to_pdb">7.4.1.1. <code class="docutils literal"><span class="pre">map_uniprot_to_pdb</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#blast_seqs_to_pdb">7.4.1.2. <code class="docutils literal"><span class="pre">blast_seqs_to_pdb</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#get_itasser_models">7.4.1.3. <code class="docutils literal"><span class="pre">get_itasser_models</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#manual_homology_models">7.4.1.4. <code class="docutils literal"><span class="pre">manual_homology_models</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#Downloading-and-ranking-structures">7.5. Downloading and ranking structures</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#Methods">7.5.1. Methods</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#pdb_downloader_and_metadata">7.5.1.1. <code class="docutils literal"><span class="pre">pdb_downloader_and_metadata</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#set_representative_structure">7.5.1.2. <code class="docutils literal"><span class="pre">set_representative_structure</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#Creating-homology-models">7.6. Creating homology models</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#Methods">7.6.1. Methods</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#prep_itasser_models">7.6.1.1. <code class="docutils literal"><span class="pre">prep_itasser_models</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/GEM-PRO - SBML Model (iNJ661).html#Saving-your-GEM-PRO">7.7. Saving your GEM-PRO</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html">8. Protein - Structure Mapping, Alignments, and Visualization</a><ul>
<li class="toctree-l2"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Imports">8.1. Imports</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Logging">8.1.1. Logging</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Initialization-of-the-project">8.2. Initialization of the project</a></li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Mapping-sequence--&gt;-structure">8.3. Mapping sequence -&gt; structure</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Methods">8.3.1. Methods</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#blast_representative_sequence_to_pdb">8.3.1.1. <code class="docutils literal"><span class="pre">blast_representative_sequence_to_pdb</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Downloading-and-ranking-structures">8.4. Downloading and ranking structures</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Methods">8.4.1. Methods</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#pdb_downloader_and_metadata">8.4.1.1. <code class="docutils literal"><span class="pre">pdb_downloader_and_metadata</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#set_representative_structure">8.4.1.2. <code class="docutils literal"><span class="pre">set_representative_structure</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Loading-and-aligning-new-sequences">8.5. Loading and aligning new sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Methods">8.5.1. Methods</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#load_manual_sequence">8.5.1.1. <code class="docutils literal"><span class="pre">load_manual_sequence</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#pairwise_align_sequences_to_representative">8.5.1.2. <code class="docutils literal"><span class="pre">pairwise_align_sequences_to_representative</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#representative_sequence.sequence_mutation_summary">8.5.1.3. <code class="docutils literal"><span class="pre">representative_sequence.sequence_mutation_summary</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Viewing-structures">8.6. Viewing structures</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Methods">8.6.1. Methods</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#representative_structure.view_structure">8.6.1.1. <code class="docutils literal"><span class="pre">representative_structure.view_structure</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#view_all_mutations">8.6.1.2. <code class="docutils literal"><span class="pre">view_all_mutations</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#representative_structure.view_structure_and_highlight_residues">8.6.1.3. <code class="docutils literal"><span class="pre">representative_structure.view_structure_and_highlight_residues</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Mapping-sequence-residue-numbers-to-structure-residue-numbers">8.7. Mapping sequence residue numbers to structure residue numbers</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Methods">8.7.1. Methods</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#representative_structure.map_repseq_resnums_to_structure_resnums">8.7.1.1. <code class="docutils literal"><span class="pre">representative_structure.map_repseq_resnums_to_structure_resnums</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="notebooks/Protein - Structure Mapping, Alignments, and Visualization.html#Saving">8.8. Saving</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="python_api.html">9. Python API</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">ssbio</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          

 



<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>4. Escher, COBRA, and COBRApy</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/tutorial.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast,
.nboutput.nblast {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast + .nbinput {
    margin-top: -19px;
}

/* nice headers on first paragraph of info/warning boxes */
.admonition .first {
    margin: -12px;
    padding: 6px 12px;
    margin-bottom: 12px;
    color: #fff;
    line-height: 1;
    display: block;
}
.admonition.warning .first {
    background: #f0b37e;
}
.admonition.note .first {
    background: #6ab0de;
}
.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}
</style>
<div class="section" id="escher-cobra-and-cobrapy">
<h1>4. Escher, COBRA, and COBRApy<a class="headerlink" href="#escher-cobra-and-cobrapy" title="Permalink to this headline">¶</a></h1>
<p>Escher can be used as an independent application, but it draws heavily on the
information in <a class="reference external" href="http://opencobra.github.io/">COBRA</a> models. A COBRA model is a collection of all the
reactions, metabolites, and genes known to exist in an organism. (In the
literature, these are generally called genome-scale models (GEMs), but we refer
to them as COBRA models here to emphasize that Escher interoperates with models
that are exported from <a class="reference external" href="http://opencobra.github.io/cobrapy/">COBRApy</a>.)</p>
<p>By loading a COBRA model into the Escher interface, you have access to every
reaction and metabolite in that model. You also have the <span class="xref std std-ref">gene reaction
rules</span> for the reactions in the network, which allow you
to connect gene data to reactions and metabolites.</p>
<p><a class="reference external" href="http://opencobra.github.io/cobrapy/">COBRApy</a> is a software package for COBRA modeling written in Python. The Escher
Python package uses COBRApy package for reading and writing COBRA models.</p>
<div class="section" id="maps-and-models">
<h2>4.1. Maps and models<a class="headerlink" href="#maps-and-models" title="Permalink to this headline">¶</a></h2>
<p>In Escher, you will see references to <em>maps</em> and <em>models</em>.</p>
<p>A map contains the reactions and metabolites that you see in the Escher builder,
including their locations, text annotations, and the canvas.</p>
<p>A model (a COBRA model) contains reactions and metabolites that you have not
drawn yet. Thus, you can load a COBRA model when you want to draw new reactions
on the map.</p>
</div>
<div class="section" id="what-is-json-and-why-do-we-use-it">
<h2>4.2. What is JSON and why do we use it?<a class="headerlink" href="#what-is-json-and-why-do-we-use-it" title="Permalink to this headline">¶</a></h2>
<p>Both Escher maps and COBRA models are stored as <a class="reference external" href="http://www.wikiwand.com/en/JSON">JSON</a> files. JSON is a useful,
plain-text format for storing nested data structures. We use JSON much like the
SBML community uses XML. You may notice that SBML files have a .xml extension,
and Escher maps and COBRA models have a .json extension.</p>
<p>You can use Python to explore a JSON file like this:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">json</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;map.json&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
<span class="n">map_object</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

<span class="nb">print</span> <span class="n">map_object</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="nb">print</span> <span class="n">map_object</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="s1">&#39;reactions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="section" id="escher-sbml-and-sbgn">
<h2>4.3. Escher, SBML, and SBGN<a class="headerlink" href="#escher-sbml-and-sbgn" title="Permalink to this headline">¶</a></h2>
<p>A tool has been developed for converting Escher maps to <a class="reference external" href="http://sbml.org/Community/Wiki/SBML_Level_3_Proposals/Layout">SBML Layout</a> and
<a class="reference external" href="http://www.sbgn.org/">SBGN</a>, and it will be released soon.</p>
<p>COBRA models can be converted to SBML using <a class="reference external" href="http://opencobra.github.io/cobrapy/">COBRApy</a>.</p>
</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="notebooks/GEM-PRO - Genes &amp; Sequences.html" class="btn btn-neutral float-right" title="5. GEM-PRO - Genes &amp; Sequences" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="atlas.html" class="btn btn-neutral" title="3. ATLAS pipeline" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Nathan Mih.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.0.0a0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>